% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ctwas_est_parameters.R
\name{est_param}
\alias{est_param}
\title{Estimates cTWAS parameters using EM}
\usage{
est_param(
  region_data,
  init_group_prior = NULL,
  init_group_prior_var = NULL,
  niter_prefit = 3,
  niter = 30,
  min_p_single_effect = 0.8,
  use_null_weight = TRUE,
  min_snps = 2,
  min_genes = 1,
  ncore = 1,
  logfile = NULL,
  verbose = FALSE,
  ...
)
}
\arguments{
\item{region_data}{a list object indexing regions, variants and genes.}

\item{init_group_prior}{a vector of initial values of prior inclusion probabilities for SNPs and genes.}

\item{init_group_prior_var}{a vector of initial values of prior variances for SNPs and gene effects.}

\item{niter_prefit}{the number of iterations of the E-M algorithm to perform
during the initial parameter estimation step}

\item{niter}{the number of iterations of the E-M algorithm to perform during
the complete parameter estimation step}

\item{min_p_single_effect}{Regions with probability >= \code{min_p_single_effect}
of having at most one causal effect will be selected for the final EM step.}

\item{use_null_weight}{If TRUE, allow for a probability of no effect in susie}

\item{min_snps}{minimum number of SNPs in a region.}

\item{min_genes}{minimum number of genes in a region.}

\item{ncore}{The number of cores used to parallelize computation over regions}

\item{logfile}{The log filename. If NULL, print log info on screen.}

\item{verbose}{If TRUE, print detail messages}

\item{...}{Additional arguments of \code{susie_rss}.}
}
\value{
a list with estimated parameters
}
\description{
Estimates cTWAS parameters using EM
}
