---
title: "Using cTWAS main function"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using cTWAS main function}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```



## Running cTWAS analysis with the main function

In this tutorial, we will show how to perform full cTWAS analysis using the main function. 

We require first running the data preprocessing and harmonization steps before running cTWAS main analysis. 

Please follow the "Preparing cTWAS input data" tutorial to prepare cTWAS input data.

Load the packages
```{r load_package, message=FALSE}
library(ctwas)
```


We use the function `ctwas_sumstats()` to run cTWAS analysis with LD. It takes as input preprocessed GWAS z-scores (`z_snp`), and prediction models of genes (`weights`). In addition, it requires the reference data including information about regions (`region_info`), reference variant list (`snp_map`) and reference LD (`LD_map`). The function performs the main steps: (i) computing Z-scores of genes; (i) estimating model parameters; (iii) screening regions potentially containing causal genes, and (iv) fine-mapping those regions to discover causal genes. 
Let's first load the preprocessed input data:

```{r load_ctwas_res}
z_snp <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.z_snp.RDS", package = "ctwas"))

weights <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.weights.RDS", package = "ctwas"))

region_info <- readRDS(system.file("extdata/sample_data", "LDL_example.region_info.RDS", package = "ctwas"))

snp_map <- readRDS(system.file("extdata/sample_data", "LDL_example.snp_map.RDS", package = "ctwas"))

LD_map <- readRDS(system.file("extdata/sample_data", "LDL_example.LD_map.RDS", package = "ctwas"))
```


Note: please update the paths to LD matrices in `LD_map` to your own paths before running the ctwas_sumstats() function in the tutorial.

Now, we can run cTWAS using the `ctwas_sumstats()` function: 
```{r ctwas_sumstats, eval=FALSE}
ctwas_res <- ctwas_sumstats(z_snp, 
                            weights, 
                            region_info, 
                            snp_map, 
                            LD_map,
                            z_gene = z_gene,
                            thin = 0.1,
                            maxSNP = 20000,
                            filter_L = TRUE,
                            filter_nonSNP_PIP = FALSE,
                            min_nonSNP_PIP = 0.5,
                            ncore = 6, 
                            save_cor = TRUE, 
                            cor_dir = "./cor_matrix")
```


Important arguments that control how this function is executed:

+ `z_gene`: computed Z-scores of the genes. If not provided, it will compute those using `z_snp` and `weights`.

+ `thin`: the fraction of SNPs used to estimate parameters and screen regions likely containing signals. By default, we use `thin = 0.1`, i.e. 10% of the SNPs. To use all SNPs in parameter estimation and screening regions, one could change the `thin` parameter to 1, but the running time will be much longer. 

+ `maxSNP` Maximum number of SNPs in a region (e.g. 20000). Default is Inf, no limit. This can be useful if there are many SNPs in a region and you don't have enough memory to run the program.

+ `filter_L`: If `filter_L = TRUE` (default), it will first estimate the number of causal signals (L) for each region using uniform prior and select regions with at least one causal signal (L >= 1).

+ `filter_nonSNP_PIP` and `min_nonSNP_PIP`: If `filter_nonSNP_PIP = TRUE`, it will compute the non-SNP PIP for each region (using estimated L and estimated priors) and select regions with non-SNP PIP > `min_nonSNP_PIP` (by default, `min_nonSNP_PIP = 0.5`). By default, we use the `filter_L` in the version with LD, and use `filter_nonSNP_PIP` in the "no-LD" version. 

+ `ncore`:  the number of cores to parallelize computing over regions.

+ `save_cor` and `cor_dir`: If `save_cor = TRUE`, we save the correlation matrices in the `cor_dir`directory, which will be used later for visualizing the results.


