---
title: "Preparing cTWAS input data"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Preparing cTWAS input data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```



This tutorial shows how to prepare for the input data for running cTWAS with summary statistics.

Load the package
```{r load_package}
library(ctwas)
```


We use `VariantAnnotation` and `gwasvcf` packages to read the VCF files, so please install those packages before running this tutorial. 

The inputs for the summary statistics version of cTWAS include GWAS summary statistics, 
prediction models (referred to as “weights” in the documents), and LD reference. Additionally, cTWAS partitions the genome into disjoint regions, assuming no LD across regions. So a user also needs to provide the definition of these regions.  

Note: In this tutorial, we assume we run full cTWAS using the LD reference. To run cTWAS without LD, see the "minimal tutorial of running cTWAS without LD":

In the tutorial, we use sample data provided in the package.

## GWAS Z-scores

We will use summary statistics from a GWAS of LDL cholesterol in the UK Biobank.  

We can download the VCF from the IEU Open GWAS Project. 
```{bash, eval=FALSE}
wget https://gwas.mrcieu.ac.uk/files/ukb-d-30780_irnt/ukb-d-30780_irnt.vcf.gz
```


cTWAS needs a data frame `z_snp` as input, with columns "id", "A1", "A2", "z". Each row has data of a variant.  We use rsIDs as the variant IDs. The naming of variant IDs should be consistent between summary statistics, LD reference and weights. `A1` is the alternate allele, and `A2` is the reference allele. `z` is the Z-scores of variants in GWAS. In the GWAS summary statistics VCF file we downloaded here, Z-scores are not available, so we need to compute them from the effect sizes and standard errors. 

Let us read the GWAS summary statistics and convert it to the `z_snp` data frame. We will also collect the sample size, which will be used later. 
 
```{r read_sumstats, eval=FALSE}
z_snp <- VariantAnnotation::readVcf("ukb-d-30780_irnt.vcf.gz")
gwas_name <- "ukb-d-30780_irnt"

z_snp <- as.data.frame(gwasvcf::vcf_to_tibble(z_snp))
z_snp$Z <- z_snp$ES/z_snp$SE

gwas_n <- as.numeric(names(sort(table(z_snp$SS),decreasing=TRUE)[1]))
cat("gwas_n=", gwas_n, "\n")


z_snp <- z_snp[,c("rsid", "ALT", "REF", "Z")]
colnames(z_snp) <- c("id", "A1", "A2", "z")
```


If your GWAS summary statistics is of other formats, please extract relevant columns (rsid, alternative allele, reference allele, Z-score) and convert it to the format shown above. 

## Prediction models

Prediction models can be specified in either PredictDB or FUSION format. 
PredictDB is the recommended format, as it has information about correlations among variants included in a model, and also has extra information about the genes in the model. In this user guide, we focus on the PredictDB format, but will provide information on using FUSION format. We also note that it is possible to run cTWAS with only a list of QTLs, which is often the case in practice. See the section "Creating prediction models from QTL lists" below. 

In terms of the choice of prediction models, cTWAS performs best when prediction models are sparse, i.e. they have relatively few variants per gene. Dense weights may lead to a “cross-region” problem. Basically, if the variants in the prediction model of a gene span two regions, it would be unclear to cTWAS what region the gene should be assigned to. The problem becomes worse with dense weights. If this happens, cTWAS will attempt to assign the gene to one of the two regions that contain most of the weights. Nevertheless, there will be a risk of cross-region LD which can lead to problems. Given this consideration, we recommend choosing sparse prediction models such as Lasso. If using dense prediction models, we recommend removing variants with weights below a threshold from the prediction models.

Note: we implemented an option to address the “cross-region” problem. It performs “region merging” as a post-processing step. If any gene has variants in the weights that span two regions, those regions will be merged, and cTWAS will rerun the analysis using the merged regions. See the tutorial "Post-processing cTWAS results" for details.   

Please check [PredictDB](http://predictdb.org/) for the format of PredictDB weights. To specify weights in PredictDB format, provide the path to the `.db` file when running cTWAS. 

```{r predictdb_weights}
weight_file <- system.file("extdata/sample_data", "expression_Liver.db", package = "ctwas")
```

Please check [FUSION/TWAS](http://gusevlab.org/projects/fusion/#compute-your-own-predictive-models) for the format of FUSION weights. To specify weights in FUSION format, provide the path to the folder that contains the `.wgt.RDat` files. The `.pos` file which points to the individual *.RDat weight files are assumed to be outside the folder. For more format details, check the FUSION website. 

### Creating prediction models from QTL lists

Often a researcher may have a list of significant eQTLs, but have not explicitly built prediction models. In such cases, it is possible to run cTWAS. One just needs to use the top eQTL per gene as the prediction models. Running cTWAS in this way would be similar to colocalization analysis. Nevertheless, it still has some advantages over colocalization: it estimates the important parameters from data, and it analyzes multiple genes in a region simultaneously. See the Supplementary Note of the cTWAS paper for details. 

We could convert top QTLs into PredictDB format weights using the function `make_predictdb_from_QTLs()`. It will create PredictDB format weights, and save to the `outputdir` directory with `outname` as the file name.

```{r make_predictdb_from_QTLs, eval=FALSE}
make_predictdb_from_QTLs(weight_table, extra_table, 
                         use_top_QTL = TRUE,
                         outputdir, outname)
```


`weight_table` is a data frame with genes, top QTLs and weights for each gene in the following example format:

```{r}
"
gene        rsid                    varID         ref_allele eff_allele weight
cg07570687  rs11190571  chr10_100472320_G_A_b38          G          A      1
cg16342193   rs7070776  chr10_100504915_A_G_b38          A          G      1
cg21542427   rs2495721  chr10_100614350_G_A_b38          G          A      1
cg26540559   rs4917904  chr10_100687535_G_A_b38          G          A      1
cg23069052 rs147983854 chr10_100717502_GC_G_b38         GC          G      1
cg24179445  rs12268880  chr10_100735528_G_C_b38          G          C      1
"
```


`extra_table` is an optional data frame with information of the genes ("gene", "genename", "gene_type", etc.), for example: 

```{r}
"
gene      genename      gene_type
cg19220149      A2M protein_coding
cg06705064   A4GALT protein_coding
cg25015779   A4GALT protein_coding
cg08572757     AACS protein_coding
cg06287003     AACS protein_coding
cg26295774  AADACL2 protein_coding
"
```


When `use_top_QTL=TRUE`, it is limited to only one top eQTL per gene. If `weight_table` has more than one SNP per gene, it will select the top SNP with the largest `abs(weight)`. It will also create a simple covariance table with covariance set to 1, as each gene only has one top eQTL.

If you want to use multiple eQTLs per gene, you can set `use_top_QTL=FALSE`. But we assume the weights of the eQTLs are learned from multiple regression (instead of marginal effect sizes). 

## Reference data

The choice of LD reference population is important for fine-mapping. Best practice for fine-mapping is to use an in-sample LD reference (LD computed using the subjects in the GWAS sample). If in-sample LD reference is not an option, the LD reference should be as representative of the population in the GWAS sample as possible.

It is critical that the genome build (e.g. hg38) of the LD reference matches the genome build used to train the prediction models. The genome build of the GWAS summary statistics does not matter because variant positions are determined by the LD reference.

### Defining regions

cTWAS assumes that the genome is partitioned into approximately independent LD regions. 

We included in the package predefined regions based on European (EUR), Asian (ASN), or African (AFR) populations, using either genome build b38 or b37. 
These regions were previously generated using [LDetect](https://github.com/endrebak/ldetect). 

Here we use the b38 European LDetect blocks, which is again included in the R package.

```{r region_met}
region_file <- system.file("extdata/ldetect", "EUR.b38.ldetect.regions.RDS", package = "ctwas")
region_info <- readRDS(region_file)
region_info <- subset(region_info, chrom == 16)

head(region_info)
```


`region_info` contains the region definitions, and IDs of the regions (by default, we use the convention <chrom_start_stop> for the region IDs).

In this tutorial, we use chr16 as an example, so we specify `chrom = 16`. In real cTWAS analysis, you should run the entire genome. 

### LD matrices

cTWAS performs its analysis region-by-region. The preferred way to run cTWAS is to provide pre-computed LD matrices for each region. 

To run cTWAS analysis with LD, we need reference LD matrices for the predefined regions. By default, we need a LD matrix for each region (preferably in `.RDS` format) and a variant information table accompanying each LD matrix. The `.RDS` files store the LD correlation matrix  (R) for each region (a $p \times p$ matrix, $p$ is the number of variants in the region). The accompanying `.Rvar` files include variant information for the region, and the order of its rows should match the order of rows and columns in the `.RDS` file.

We have precomputed reference LD matrices (`.RDS` files) and the variant information tables (`.Rvar` files) accompanying the LD matrices for both UKB and 1000G European Phase3 references. The complete LD matrices of European individuals from UK Biobank can be downloaded [here](https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn). 
On the University of Chicago RCC cluster, the b38 reference is available at `/project2/mstephens/wcrouse/UKB_LDR_0.1/` and the b37 reference is available at `/project2/mstephens/wcrouse/UKB_LDR_0.1_b37/`.

The columns of the `.Rvar` file include information on chromosome, variant name, position in base pairs, and the alternative and reference alleles. The `variance` column is the variance of each variant prior to standardization; this is required for PredictDB weights but not FUSION weights. PredictDB weights should be scaled by the variance before computing gene expression. This is because PredictDB weights assume that variant genotypes are not standardized before imputation, but our implementation assumes standardized variant genotypes. If variance information is missing, or if weights are in PredictDB format but are already on the standardized scale (e.g. if they were converted from FUSION to PredictDB format), this scaling can be turned off by setting `scale_predictdb_weights=FALSE` in the `preprocess_weights()` function. We've also included information on allele frequency in the variant info, but this is optional.

The naming convention for the LD matrices is `[filestem]_chr[#].R_snp.[start]_[end].RDS`. 
Each variant should be uniquely assigned to a region, and the regions should be left closed and right open, i.e. [start, stop). The positions of the LD matrices must match exactly the positions specified by the region file. Do not include multiallelic variants in the LD reference. 

#### Create LD matrices from individual level genotype data

cTWAS provides a function,  `convert_geno_to_LD_matrix()` to convert genotype files and definition of LD regions to the LD matrices (`.RDS` files) and corresponding variant information (`.Rvar` files). 

We provide example R scripts to generate the LD matrices for UK biobank or 1000 genomes genotype files [here](https://github.com/xinhe-lab/ctwas/tree/multigroup_test/inst/extdata/scripts)

### Mapping reference SNPs and LD matrices to regions

To map reference SNPs and LD matrices to regions, we need to add two additional columns "LD_matrix" and "SNP_info" to `region_info`:  "LD_matrix" stores the filenames to the precomputed LD (R) matrices (`.RDS` files in our precomputed reference LD files),  "SNP_info" stores the filenames of corresponding variant information (`.Rvar` files). 

The following paths and filenames were from the University of Chicago RCC cluster, please replace them with your own paths. 
```{r add_LD_paths, eval=FALSE}
LD_dir <- "/project2/mstephens/wcrouse/UKB_LDR_0.1/"
genome_version <- "b38"

LD_filestem <- sprintf("ukb_%s_0.1_chr%s.R_snp.%s_%s", genome_version, region_info$chrom, region_info$start, region_info$stop)

region_metatable <- region_info
region_metatable$LD_matrix <- file.path(LD_dir, paste0(LD_filestem, ".RDS"))
region_metatable$SNP_info <- file.path(LD_dir, paste0(LD_filestem, ".Rvar"))
```


The `create_snp_LD_map()` function takes input of `region_metatable`, and returns processed `region_info`, `snp_map` and `LD_map`.

```{r create_snp_LD_map, eval=FALSE}
res <- create_snp_LD_map(region_metatable)
region_info <- res$region_info
snp_map <- res$snp_map
LD_map <- res$LD_map
```



When `use_LD = FALSE`, we need `region_info` (region definitions, no need to provide "LD_matrix" and "SNP_info" columns) and `ref_snp_info`, a data frame containing information of all variants from the reference. 

We have the reference variant information from all the LD matrices from UK Biobank in [hg38](https://uchicago.box.com/s/t089or92dkovv0epkrjvxq8r9db9ys99) and [hg19](https://uchicago.box.com/s/ufko2gjagcb693dob4khccqubuztb9pz).

We then use the `create_snp_map()` function to preprocess region definitions, and map variants from the LD reference to regions, which returns processed `region_info` and `snp_map`.

```{r create_snp_map, eval=FALSE}
ref_snp_info_file <- system.file("extdata/sample_data", "ukb_b38_0.1_chr16_var_info.Rvar.gz", package = "ctwas")

ref_snp_info <- fread(ref_snp_info_file, sep = "\t")
class(ref_snp_info) <- "data.frame"

res <- create_snp_map(region_info, ref_snp_info)
region_info <- res$region_info
snp_map <- res$snp_map
```



## Data harmonization

There are a few potential problems when preparing the input data. First, the variants in the three sets of input data, GWAS summary statistics, weights, and the reference data, may not match. Only variants in all three input data will be used in cTWAS. The `preprocess_weights()` function described below will automatically select variants in all three input data. So it is important to maximize the overlap of the variants in the three sets. This can be done for example, by imputing GWAS summary statistics of the variants missing in GWAS but in the LD reference. Another useful pre-processing step is to perform Minor allele frequency (MAF) filtering on the GWAS data so that only those with MAF above a certain cutoff would be used in the analysis, ideally the same cutoff used in the references. Additionally, when building the prediction models of gene expression, it is better to impute the genotype data using the LD reference, if possible.  All these steps should be done before running cTWAS.  

The second potential problem is that the effect alleles in the prediction model, GWAS and LD reference may not agree with each other, thus we need to ``harmonize'' the data to ensure that the effect alleles match. If the data are not already harmonized, we provide some options for harmonization. These inputs should be harmonized prior to cTWAS analysis, i.e. the reference and alternative alleles for each variant should match across all three data sources. 
    
Another potential problem is the LD of the GWAS data (in-sample LD) do not match the reference LD. This can lead to false positives in fine-mapping tools. Diagnostic tools including [SuSiE-RSS][susierss_diagnostic], and [DENTIST][DENTIST], have been developed to check possible LD mismatch. Because it is very time consuming to run the LD mismatch diagnosis for all the regions across the genome, we will perform LD mismatch diagnosis and adjustment only for selected regions with high PIP signals in the post-processing section. Please see the tutorial "Post-processing cTWAS results" for details.

Below we describe several functions to harmonize the input datasets. 

### Harmonizing GWAS z-scores and the reference data

The `preprocess_z_snp()` function harmonizes GWAS z-scores and the reference data based on the included allele information.

```{r preprocess_z_snp, eval=FALSE}
z_snp <- readRDS(system.file("extdata/sample_data", "LDL_example.z_snp.RDS", package = "ctwas"))

z_snp <- preprocess_z_snp(z_snp, snp_map, 
                          drop_multiallelic = TRUE, 
                          drop_strand_ambig = TRUE)
```


By default, it will drop multiallelic variants and strand ambiguous variants. You can change these by setting the options `drop_multiallelic` and `drop_strand_ambig`.

### Harmonizing prediction models and the reference data

We use the `preprocess_weight()` function to harmonize the PredictDB/FUSION prediction models and LD reference. 

In this example, we will use liver gene expression models trained on GTEx v8 in the PredictDB format. We can download both the prediction models (.db) and the covariances between variants in the prediction models (.txt.gz) from the link below. The covariances can optionally be used for computing LD.
```{bash, eval=FALSE}
wget https://zenodo.org/record/3518299/files/mashr_eqtl.tar
tar -xvf mashr_eqtl.tar
```



```{r preprocess_weights, eval=FALSE}
weight_file   <- system.file("extdata/sample_data", "expression_Liver.db", package = "ctwas")
weights       <- preprocess_weights(weight_file,
                                    region_info,
                                    z_snp$id,
                                    snp_map,                                  
                                    weight_format = "PredictDB",
                                    ncore = 6,
                                    drop_strand_ambig = TRUE,
                                    scale_predictdb_weights = TRUE,
                                    load_predictdb_LD = TRUE,
                                    filter_protein_coding_genes = TRUE)
```



We specify PredictDB or FUSION format in `weight_format` and the specific method in FUSION by `fusion_method`. 
By default, it will drop strand ambiguous variants (`drop_strand_ambig = TRUE`) and limit to protein coding genes (`filter_protein_coding_genes = TRUE`).
For PredictDB weights, we set `scale_predictdb_weights = TRUE`, because PredictDB assumes the unstandardized genotypes (see the discussion in the "Reference data" section). For FUSION weights or weights converted from eQTLs, we `scale_predictdb_weights = FALSE`.
For PredictDB models and weights converted from top eQTLs, we set `load_predictdb_LD = TRUE` to load pre-computed correlations (`.txt.gz` file) between weight variants. 
For FUSION models, we set `load_predictdb_LD = FALSE` to calculate correlations between weight variants using the LD reference. 
We could set `ncore` to parallelize the computation.

## Computing Z-scores of genes

After we have done data preprocessing, we compute Z-scores of genes. This step basically performs Transcriptome-wide association studies (TWAS) on each gene with a prediction model. The underlying calculation is based on [S-PrediXcan][S-PrediXcan].

The `compute_gene_z` function computes Z-scores for the genes using preprocessed SNP Z-scores (`z_snp`) and the preprocessed weights (`weights`). 

```{r compute_gene_z, eval=FALSE}
z_gene <- compute_gene_z(z_snp, weights)
```



[reference]: https://xinhe-lab.github.io/ctwas/reference/index.html
[UKBB_LD_ref]: https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn
[LDetect]: https://github.com/endrebak/ldetect
[PredictDB]: http://predictdb.org/
[FUSION_format]: http://gusevlab.org/projects/fusion/#compute-your-own-predictive-models
[S-PrediXcan]: https://www.nature.com/articles/s41467-018-03621-1
[susierss_diagnostic]: https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html
[DENTIST]: https://github.com/Yves-CHEN/DENTIST/
