---
title: "Running cTWAS analysis"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Running cTWAS analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```



Running cTWAS analysis involves several major steps: preparing input data, computing associations of genes with the phenotype, estimating parameters, screening for candidate genomic regions, and fine-mapping these regions. 

We require first running the data preprocessing and harmonization steps. Please follow the "Preparing cTWAS input data" tutorial.

Load the packages
```{r load_package, message=FALSE}
library(ctwas)
```


Let's first load the preprocessed input data,including GWAS summary statistics (`z_snp`), the processed weights of genes (`weights`), the region definitions (`region_info`), and the list of SNPs and information of LD matrix of each region (`snp_map` and `LD_map`): :

```{r load_ctwas_res}
z_snp <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.z_snp.RDS", package = "ctwas"))

weights <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.weights.RDS", package = "ctwas"))

region_info <- readRDS(system.file("extdata/sample_data", "LDL_example.region_info.RDS", package = "ctwas"))

snp_map <- readRDS(system.file("extdata/sample_data", "LDL_example.snp_map.RDS", package = "ctwas"))

LD_map <- readRDS(system.file("extdata/sample_data", "LDL_example.LD_map.RDS", package = "ctwas"))
```


We have two ways of running cTWAS main analysis: running the main function, or running separate modules. The first is easier to run, and the second has the advantage of giving you more control of the analysis (see below). 


## Running cTWAS main function

In this section, we will show how to perform full cTWAS analysis using the main function.

We use the function `ctwas_sumstats()` to run cTWAS analysis with LD. It takes as input preprocessed GWAS z-scores (`z_snp`), and prediction models of genes (`weights`). In addition, it requires the reference data including information about regions (`region_info`), reference variant list (`snp_map`) and reference LD (`LD_map`). 

The function performs the main steps: (i) computing Z-scores of genes; (i) estimating model parameters; (iii) screening regions potentially containing causal genes, and (iv) fine-mapping those regions to discover causal genes. 

Now, we can run cTWAS using the `ctwas_sumstats()` function: 
```{r ctwas_sumstats, eval=FALSE}
ctwas_res <- ctwas_sumstats(z_snp, 
                            weights, 
                            region_info, 
                            snp_map, 
                            LD_map,
                            thin = 0.1,
                            maxSNP = 20000,
                            filter_nonSNP_PIP = FALSE,
                            min_nonSNP_PIP = 0.5,
                            ncore = 6, 
                            save_cor = TRUE, 
                            cor_dir = "./cor_matrix")
```


Important arguments that control how this function is executed:

+ `z_gene`: computed Z-scores of the genes. If not provided, it will compute those using `z_snp` and `weights`.

+ `thin`: the fraction of SNPs used to estimate parameters and screen regions likely containing signals. By sampling a fraction of SNPs, this speeds up computation. By default, we use `thin = 0.1`, i.e. 10% of the SNPs. To use all SNPs in parameter estimation and screening regions, one could change the `thin` parameter to 1. 

+ `maxSNP` Maximum number of SNPs in a region (e.g. 20000). Default is Inf, no limit. This can be useful if there are too many SNPs in a region and you don't have enough memory to run the program.

+ `filter_nonSNP_PIP` and `min_nonSNP_PIP`: If `filter_nonSNP_PIP = TRUE`, it will compute the non-SNP PIP, i.e. the total PIP of all molecular traits,  for each region (using the estimated parameters) and select regions with non-SNP PIP > `min_nonSNP_PIP` (0.5 by default`). 

+ `ncore`:  the number of cores to parallelize computing over regions.

+ `save_cor` and `cor_dir`: If `save_cor = TRUE`, we save the correlation matrices in the `cor_dir`directory, which will be used later for visualizing the results.

To understand and summarize the results from cTWAS, please read the tutorial “Summarizing and visualizing cTWAS results”. 

Note: One can run cTWAS without using reference LD data. See “A minimal tutorial of running cTWAS without LD”. That tutorial describes how to run cTWAS in a single step. 


## Running cTWAS modules

This section describes how to perform cTWAS analysis by running each of the modules separately. This would allow more flexible controls of the cTWAS modules, e.g. running each module with different settings (memory, parallelization, etc.). We could also save the results from each module for future access, that way, we can re-run just one module, if necessary, without running the entire process. One useful way of running cTWAS is to first run the first few steps to estimate model parameters and assess their values. If this step does not converge, or the parameters look strange, it is best to pause and understand why.

### Computing Z-scores of genes

After we have done data preprocessing, we compute Z-scores of genes. This step basically performs Transcriptome-wide association studies (TWAS) on each gene with a prediction model. The underlying calculation is based on [S-PrediXcan][S-PrediXcan].

The `compute_gene_z` function computes Z-scores for the genes using preprocessed SNP Z-scores (`z_snp`) and the preprocessed weights (`weights`). 

```{r compute_gene_z, eval=FALSE}
z_gene <- compute_gene_z(z_snp, weights)
```




### Assemble input data for the regions

Because cTWAS runs region-by-region, we will first need to prepare the data for each region, combining information of all variants and genes in that region. After data preprocessing and computation of Z-scores of genes, we assemble the input data for all the regions using the function `assemble_region_data()`.  It assigns genes, variants and their Z-scores to each region. The assignment of a gene to regions is based on the weights, i.e. variants in the prediction model, of the gene. 
```{r assemble_region_data, eval=FALSE}
res <- assemble_region_data(region_info, 
                            z_snp, 
                            z_gene, 
                            weights, 
                            snp_map,
                            thin = 0.1, 
                            maxSNP = 20000, 
                            ncore = 6)
region_data <- res$region_data
boundary_genes <- res$boundary_genes
```


The function has a few arguments to control its behavior. The `thin` argument randomly selects a subset of variants (10% when thin = 0.1) to use during parameter estimation and screening for candidate regions. This helps reduce computation. If `thin = 1`, it will use all the SNPs. `maxSNP` sets a maximum on the number of variants that can be in a single region to prevent memory issues during fine-mapping. If the number of SNPs in a region is more than `maxSNP`, It will trim the SNPs (randomly, by default) in the region. `ncore` specifies the number of cores to use when parallelizing over regions.

This function returns `region_data`, a list object containing input data (IDs of genes and variants, their Z-scores, genomic boundaries, etc.) for each of the regions. It also returns 
a data frame `boundary_genes`, containing the information of genes whose weights cross region boundaries. The information would be used in later analysis and post-processing. 


### Estimating parameters

We use the `est_param()` function to estimate two sets of parameters: the prior inclusion probabilities and the prior effect size variance for genes and variants, separately. 

This step will run the EM algorithm to estimate parameters and return the estimated parameters (`group_prior`, `group_prior_var`, etc.). For technical reasons (see Methods of the cTWAS paper), it will run two rounds of EM algorithm. In the first round, it uses fewer iterations (`niter_prefit=3` by default) to get rough parameter estimates, and using these values, select regions for the analysis in the second run of EM. We use more iterations in the second round (`niter=30` by default) to get accurate parameter estimates. 

```{r est_param, eval=FALSE}
param <- est_param(region_data, 
                   niter_prefit = 3,
                   niter = 30, 
                   ncore = 6)
group_prior <- param$group_prior
group_prior_var <- param$group_prior_var
```


*Note:*, we used sample data (from chr16) in this example, however, in real data analysis, parameter estimation should be done using data from the *entire genome*.

### Screening regions

After parameter estimation, we use the `screen_regions()` function to perform a screening process to select regions with likely causal signals in genes. To find these regions, we perform an initial fine-mapping with thinned SNPs (a subset of all variants, see the Section “Assemble input data”). Only these regions would be subject to full fine-mapping analysis, using all variants, in the final step. Thus screening regions can save computational time. 

cTWAS have two options to screen regions:
 
+ screening by the number of causal signals: If `filter_L = TRUE` (the default option), cTWAS will first estimate the number of causal signals (L) for each region. This step is done using uniform prior, treating all variants and genes equally. From the results, we obtain the number of credible sets (L). We then select regions with at least one credible set, L > 0. 

+ screening by non-SNP PIPs: If `filter_nonSNP_PIP = TRUE`, cTWAS will compute the non-SNP PIP, i.e. total PIP of all genes, for each region (using estimated L and estimated priors). It then selects regions by applying a  `min_nonSNP_PIP` cutoff (0.5, by default). 

When running with LD (`use_LD = TRUE`), we recommend screening by the number of causal signals, i.e. setting `filter_L = TRUE` and `filter_nonSNP_PIP = FALSE`. 

```{r screen_regions_cs, eval=FALSE}
screen_region_res <- screen_regions(region_data,
                                    use_LD = TRUE,
                                    LD_map = LD_map,
                                    snp_map = snp_map, 
                                    weights = weights,
                                    filter_L = TRUE,
                                    filter_nonSNP_PIP = FALSE,
                                    ncore = 6)
screened_region_data <- screen_region_res$screened_region_data
L <- screen_region_res$L
```


It returns `screened_region_data` which contains a subset of region data for selected regions. It also returns the estimated L (number of credible sets) for the selected regions, which could be used in the fine-mapping step to set L for each region.

When running without LD (`use_LD = FALSE`), cTWAS would only perform screening regions by non-SNP PIPs, and it will automatically set `filter_nonSNP_PIP = TRUE` and `filter_L = FALSE`.

```{r screen_regions_nonSNP_PIP, eval=FALSE}
screen_region_res <- screen_regions(region_data,
                                    use_LD = FALSE,
                                    group_prior = group_prior,
                                    group_prior_var = group_prior_var,
                                    filter_L = FALSE,
                                    filter_nonSNP_PIP = TRUE,
                                    min_nonSNP_PIP = 0.5,
                                    ncore = 6)
screened_region_data <- screen_region_res$screened_region_data
```


After selecting regions, we expand the screened regions with all SNPs for final fine-mapping. (This is not needed when `thin = 1`.)

```{r expand_region_data, eval=FALSE}
screened_region_data <- expand_region_data(screened_region_data,
                                           snp_map,
                                           z_snp,
                                           maxSNP = 20000,
                                           ncore = ncore)
```


If the number of SNPs in a region is more than `maxSNP`, It will trim the SNPs in the region (according to the Z-scores, by default).

### Fine-mapping screened regions

We now perform the fine-mapping step for each of the screened regions with the estimated parameters.

The `finemap_regions()` function performs fine-mapping for the regions in the `screened_region_data`.

It first computes correlation matrices among all variables included in fine-mapping (SNPs, and genes), then runs fine-mapping with estimated parameters (`group_prior` and `group_prior_var`) and estimated number of causal signals (`L`), and returns a data frame with fine-mapping results for the regions. 
```{r finemap_regions, eval=FALSE}
finemap_res <- finemap_regions(screened_region_data,
                               use_LD = TRUE,
                               LD_map = LD_map,
                               snp_map = snp_map,
                               weights = weights,
                               group_prior = group_prior,
                               group_prior_var = group_prior_var,
                               L = screen_region_res$L,
                               save_cor = TRUE,
                               cor_dir = "./cor_matrix",
                               ncore = ncore)
```


We can save the computed correlation matrices to the directory `cor_dir` for future access, by setting `save_cor = TRUE`.

*Note*: it is also possible to run fine-mapping without LD:
```{r finemap_regions_noLD, eval=FALSE}
finemap_res <- finemap_regions(screened_region_data,
                               use_LD = FALSE,
                               group_prior = group_prior,
                               group_prior_var = group_prior_var,
                               ncore = 6)
```



### Fine-mapping a single region

We could also run fine-mapping for a single region (or several regions of interest) with precomputed parameters. This is particularly useful to perform a deep analysis on genes/regions of interest. 

Here we show an example for fine mapping a single region "16_71020125_72901251".

We first expand the region of interest with the full set of SNPs using the assembled `region_data`. We also need the number of causal signals (`L`) for this region. This is obtained from the output of the earlier `screen_regions()` function.

```{r example_region, eval=FALSE}
region_id <- "16_71020125_72901251"
selected_region_data <- expand_region_data(region_data[region_id],
                                           snp_map,
                                           z_snp,
                                           maxSNP = 20000)

selected_region_L <- screen_region_res$L[region_id]
```


We could then perform fine-mapping for this region of interest:

```{r finemap_single_region, eval=FALSE}
finemap_region_res <- finemap_regions(selected_region_data,
                                      use_LD = TRUE,
                                      LD_map = LD_map,
                                      snp_map = snp_map,
                                      weights = weights,
                                      group_prior = group_prior,
                                      group_prior_var = group_prior_var,
                                      L = selected_region_L,
                                      save_cor = TRUE,
                                      cor_dir = "./cor_matrix")
```


We could also perform fine-mapping without LD:

```{r finemap_single_region_noLD, eval=FALSE}
finemap_region_res <- finemap_regions(selected_region_data,
                                      use_LD = FALSE,
                                      group_prior = group_prior,
                                      group_prior_var = group_prior_var)
```






[reference]: https://xinhe-lab.github.io/ctwas/reference/index.html
[UKBB_LD_ref]: https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn
[LDetect]: https://github.com/endrebak/ldetect
[PredictDB]: http://predictdb.org/
[FUSION_format]: http://gusevlab.org/projects/fusion/#compute-your-own-predictive-models
[S-PrediXcan]: https://www.nature.com/articles/s41467-018-03621-1
[susierss_diagnostic]: https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html
[DENTIST]: https://github.com/Yves-CHEN/DENTIST/

