<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Adjusting for Genetic Confounders in Transcriptome-Wide Association
  Studies Improves Discovery of Risk Genes of Complex Traits • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Adjusting for Genetic Confounders in Transcriptome-Wide Association
  Studies Improves Discovery of Risk Genes of Complex Traits">
<meta property="og:description" content='Implements the "causal-TWAS" (cTWAS) method
    for identifying causal genes in genome-wide association studies
    (GWAS) described in S. Zhao et al (2024)
    &lt;DOI:10.1038/s41588-023-01648-9&gt;. cTWAS generalizes methods for
    "transcriptome-wide association studies" (TWAS) in that when
    analyzing a gene it adjusts for not only other nearby genes but
    also nearby genetic variants.'>
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/software_updates.html">Software updates</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="ctwas-an-r-package-for-integrating-molecular-qtls-and-gwas-for-gene-discovery">ctwas: an R package for integrating molecular QTLs and GWAS for gene discovery<a class="anchor" aria-label="anchor" href="#ctwas-an-r-package-for-integrating-molecular-qtls-and-gwas-for-gene-discovery"></a>
</h1></div>
<p><a href="https://github.com/xinhe-lab/ctwas/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/xinhe-lab/ctwas/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></p>
<p>Expression quantitative trait loci (eQTLs) have often been used to nominate candidate genes from genome-wide association studies (GWAS). However, commonly used methods are susceptible to false positives largely due to linkage disequilibrium (LD) of eQTLs with causal variants acting on the phenotype directly.</p>
<p>Our method, “causal-TWAS” (cTWAS), addresses this challenge by borrowing ideas from statistical fine-mapping. It is a generalization of methods for transcriptome-wide association studies (TWAS), but when analyzing any gene, it adjusts for other nearby genes <em>and all nearby genetic variants.</em></p>
<p>While the <a href="https://doi.org/10.1038/s41588-023-01648-9" class="external-link">published paper</a> analyzes a single eQTL dataset, the latest version of cTWAS extends the method to integrate multiple groups of prediction models, allowing for joint analysis of multiple types of molecular traits, across potentially different tissues, cell types or conditions.</p>
<p>If you have already run cTWAS before, see <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/software_updates.html" class="external-link">this page</a> for the software updates.</p>
<p>You can <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">browse source code</a> and <a href="https://github.com/xinhe-lab/ctwas/issues" class="external-link">report a bug</a> here.</p>
<p>You can also join our <a href="https://groups.google.com/g/ctwas_users" class="external-link">Google Group</a> to ask questions, report issues, or receive notifications of software updates.</p>
<div class="section level2">
<h2 id="install-ctwas">Install ctwas<a class="anchor" aria-label="anchor" href="#install-ctwas"></a>
</h2>
<p>Use “remotes” to install the latest version of ctwas from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>,ref <span class="op">=</span> <span class="st">"multigroup"</span><span class="op">)</span></span></code></pre></div>
<p>Currently, ctwas has only been tested on Linux systems.</p>
<p>We recommend installing and running ctwas on a high-performance computing system.</p>
</div>
<div class="section level2">
<h2 id="running-ctwas">Running ctwas<a class="anchor" aria-label="anchor" href="#running-ctwas"></a>
</h2>
<p>Running a cTWAS analysis involves four main steps:</p>
<ol style="list-style-type: decimal">
<li><p>Preparing the input data.</p></li>
<li><p>Computing associations of genes with the phenotype (Z-scores).</p></li>
<li><p>Estimating the model parameters.</p></li>
<li><p>Fine-mapping causal genes</p></li>
</ol>
<p>The outputs of cTWAS are posterior inclusion probabilities (PIPs) for all variants and genes.</p>
<p>To learn more about the ctwas R package, we recommend starting with this introductory tutorial:</p>
<p><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/minimal_tutorial.html" class="external-link">A minimal tutorial of how to run cTWAS without LD</a></p>
<p>To run the full cTWAS, follow these tutorials:</p>
<ul>
<li><p><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/preparing_input_data.html" class="external-link">Preparing input data</a></p></li>
<li><p><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/running_ctwas_analysis.html" class="external-link">Running cTWAS analysis</a></p></li>
<li><p><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/summarizing_results.html" class="external-link">Summarizing and visualizing cTWAS results</a></p></li>
<li><p><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/sample_report.html" class="external-link">Sample report of cTWAS results</a></p></li>
<li><p><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/postprocessing.html" class="external-link">Post-processing cTWAS results</a></p></li>
</ul>
<p>In addition, we have some useful functions to help run cTWAS, e.g. for creating your own reference LD data:</p>
<ul><li><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/utility_functions.html" class="external-link">cTWAS utility functions</a></li></ul>
<p>We also have a FAQ page for some common questions:</p>
<ul><li><a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/FAQ.html" class="external-link">FAQ</a></li></ul>
<p><img style="display:block;margin:auto" width="700" height="300" src="reference/figures/workflow.png"></p>
</div>
<div class="section level2">
<h2 id="citing-this-work">Citing this work<a class="anchor" aria-label="anchor" href="#citing-this-work"></a>
</h2>
<p>If you find the <code>ctwas</code> package or any of the source code in this repository useful for your work, please cite:</p>
<blockquote><p>Zhao S, Crouse W, Qian S, Luo K, Stephens M, He X. Adjusting for genetic confounders in transcriptome-wide association studies improves discovery of risk genes of complex traits. Nature Genetics 56, 336–347 (2024). <a href="https://doi.org/10.1038/s41588-023-01648-9" class="external-link uri">https://doi.org/10.1038/s41588-023-01648-9</a></p></blockquote>
</div>
<div class="section level2">
<h2 id="useful-resources">Useful resources<a class="anchor" aria-label="anchor" href="#useful-resources"></a>
</h2>
<p>We have pre-computed reference LD matrices and variant information of European samples from UK Biobank. They can be downloaded <a href="https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn" class="external-link">here</a>.</p>
<p>cTWAS requires the expression prediction models, or weights, of genes. The pre-computed weights of GTEx expression and splicing traits can be downloaded from <a href="https://predictdb.org/post/2021/07/21/gtex-v8-models-on-eqtl-and-sqtl/" class="external-link">PredictDB</a>.</p>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>We thank the authors of <code>susieR</code> package for using their codes.</p>
<p>Original <code>susieR</code> code obtained by:</p>
<pre><code>git clone git@github.com:stephenslab/susieR.git
git checkout c7934c0</code></pre>
<p>Minor edits to make it accept different prior variances for each variable.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/xinhe-lab/ctwas/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ctwas</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Siming Zhao <br><small class="roles"> Author </small>  </li>
<li>Wesley Crouse <br><small class="roles"> Author </small>  </li>
<li>Sheng Qian <br><small class="roles"> Author </small>  </li>
<li>Kaixuan Luo <br><small class="roles"> Author, maintainer </small>  </li>
<li>Matthew Stephens <br><small class="roles"> Author </small>  </li>
<li>Xin He <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
