<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Using cTWAS with summary statistics genome-wide • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Using cTWAS with summary statistics genome-wide"><meta property="og:description" content="ctwas"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multigroup_ctwas_pipeline.html">Pipepline of using multigroup cTWAS with summary statistics</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>




<script src="ctwas_sumstats_genomewide_files/accessible-code-block-0.0.1/empty-anchor.js"></script>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using cTWAS with summary statistics genome-wide</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
            <h4 data-toc-skip class="date">2024-04-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/ctwas_sumstats_genomewide.Rmd" class="external-link"><code>vignettes/ctwas_sumstats_genomewide.Rmd</code></a></small>
      <div class="hidden name"><code>ctwas_sumstats_genomewide.Rmd</code></div>

    </div>

    
    
<p>This document demonstrates how to use the summary statistics version of cTWAS using tutorial data from a single group.</p>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Install <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>remotes<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;xinhe-lab/ctwas&quot;</span>, <span class="dt">ref =</span> <span class="st">&quot;multigroup_test&quot;</span>)</span></code></pre></div>
<p>Load <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># library(ctwas)</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>devtools<span class="op">::</span><span class="kw">load_all</span>(<span class="st">&quot;~/projects/cTWAS_package/multigroup_test/ctwas&quot;</span>)</span></code></pre></div>
</div>
<div id="preparing-the-input-data" class="section level2">
<h2>Preparing the input data</h2>
<p>The inputs for the summary statistics version of cTWAS include GWAS summary statistics for variants, prediction models for genes, and LD reference.</p>
<p>We used the directories and files on the University of Chicago RCC cluster as examples. If you are at UChicago, you can load those data from RCC as below.</p>
<div id="gwas-z-scores" class="section level3">
<h3>GWAS z-scores</h3>
<p>For this analysis, we will use summary statistics from a GWAS of LDL cholesterol in the UK Biobank. We will download the VCF from the IEU Open GWAS Project.</p>
<p>Next, we will read the summary statistics. Then, we will compute the z-scores and format the input data. We will also collect the sample size, which will be useful later. We will save this output for convenience.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="cf">if</span> (<span class="kw">file.exists</span>(z_snp_outfile)){</span>
<span id="cb3-2"><a href="#cb3-2"></a>  z_snp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(z_snp_outfile)</span>
<span id="cb3-3"><a href="#cb3-3"></a>} <span class="cf">else</span> {</span>
<span id="cb3-4"><a href="#cb3-4"></a>  <span class="co"># read the data using the VariantAnnotation package</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>  z_snp &lt;-<span class="st"> </span>VariantAnnotation<span class="op">::</span><span class="kw">readVcf</span>(<span class="st">&quot;/project2/xinhe/shared_data/multigroup_ctwas/test_data/ukb-d-30780_irnt.vcf.gz&quot;</span>)</span>
<span id="cb3-6"><a href="#cb3-6"></a>  z_snp &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(gwasvcf<span class="op">::</span><span class="kw">vcf_to_tibble</span>(z_snp))</span>
<span id="cb3-7"><a href="#cb3-7"></a></span>
<span id="cb3-8"><a href="#cb3-8"></a>  <span class="co"># compute the z-scores</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>  z_snp<span class="op">$</span>Z &lt;-<span class="st"> </span>z_snp<span class="op">$</span>ES<span class="op">/</span>z_snp<span class="op">$</span>SE</span>
<span id="cb3-10"><a href="#cb3-10"></a></span>
<span id="cb3-11"><a href="#cb3-11"></a>  <span class="co"># collect sample size (most frequent sample size for all variants)</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>  gwas_n &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">names</span>(<span class="kw">sort</span>(<span class="kw">table</span>(z_snp<span class="op">$</span>SS),<span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span>]))</span>
<span id="cb3-13"><a href="#cb3-13"></a></span>
<span id="cb3-14"><a href="#cb3-14"></a>  <span class="co"># subset the columns and format the column names</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>  z_snp &lt;-<span class="st"> </span>z_snp[,<span class="kw">c</span>(<span class="st">&quot;rsid&quot;</span>, <span class="st">&quot;ALT&quot;</span>, <span class="st">&quot;REF&quot;</span>, <span class="st">&quot;Z&quot;</span>)]</span>
<span id="cb3-16"><a href="#cb3-16"></a>  <span class="kw">colnames</span>(z_snp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;id&quot;</span>, <span class="st">&quot;A1&quot;</span>, <span class="st">&quot;A2&quot;</span>, <span class="st">&quot;z&quot;</span>)</span>
<span id="cb3-17"><a href="#cb3-17"></a></span>
<span id="cb3-18"><a href="#cb3-18"></a>  <span class="co"># drop multiallelic variants (id not unique)</span></span>
<span id="cb3-19"><a href="#cb3-19"></a>  z_snp &lt;-<span class="st"> </span>z_snp[<span class="op">!</span>(z_snp<span class="op">$</span>id <span class="op">%in%</span><span class="st"> </span>z_snp<span class="op">$</span>id[<span class="kw">duplicated</span>(z_snp<span class="op">$</span>id)]),]</span>
<span id="cb3-20"><a href="#cb3-20"></a></span>
<span id="cb3-21"><a href="#cb3-21"></a>  <span class="co"># save the formatted z-scores and GWAS sample size</span></span>
<span id="cb3-22"><a href="#cb3-22"></a>  <span class="kw">saveRDS</span>(z_snp, <span class="dt">file=</span>z_snp_outfile)</span>
<span id="cb3-23"><a href="#cb3-23"></a>  <span class="kw">saveRDS</span>(gwas_n, <span class="dt">file=</span><span class="kw">file.path</span>(outputdir, <span class="st">&quot;gwas_n.RDS&quot;</span>))</span>
<span id="cb3-24"><a href="#cb3-24"></a>  <span class="kw">save</span>(z_snp, gwas_n, <span class="dt">file =</span> <span class="kw">file.path</span>(outputdir, <span class="kw">paste0</span>(z_snp_stem, <span class="st">&quot;.Rdata&quot;</span>)))</span>
<span id="cb3-25"><a href="#cb3-25"></a>}</span></code></pre></div>
<p>After the previous step, we can load the data and look at the format. <code>z_snp</code> is a data frame, and each row is a variant. <code>A1</code> is the alternate allele, and <code>A2</code> is the reference allele. The sample size for this GWAS is <code>N=343,621</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>z_snp &lt;-<span class="st"> </span><span class="kw">load</span>(<span class="kw">file.path</span>(outputdir, <span class="kw">paste0</span>(z_snp_stem, <span class="st">&quot;.Rdata&quot;</span>)))</span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="kw">head</span>(z_snp)</span>
<span id="cb4-4"><a href="#cb4-4"></a></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co"># sample size</span></span>
<span id="cb4-6"><a href="#cb4-6"></a>gwas_n</span></code></pre></div>
</div>
<div id="prediction-models" class="section level3">
<h3>Prediction models</h3>
<p>In this version, we use the prediction models in PredictDB format.</p>
<p>Please check <a href="http://predictdb.org/">PredictDB</a> for the format of PredictDB weights.</p>
<p>For this analysis, we will use liver gene expression models trained on GTEx v8 in the PredictDB format.</p>
<p>To specify weights in PredictDB format, provide the path to the <code>.db</code> file.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>weight &lt;-<span class="st"> &quot;/project2/xinhe/shared_data/multigroup_ctwas/test_data/mashr_Liver_nolnc.db&quot;</span></span></code></pre></div>
</div>
<div id="ld-reference-and-regions" class="section level3">
<h3>LD reference and regions</h3>
<div id="defining-regions" class="section level4">
<h4>Defining regions</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>region_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata/ldetect&quot;</span>, <span class="st">&quot;EUR.b38.bed&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ctwas&quot;</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a>region_info &lt;-<span class="st"> </span><span class="kw">read.table</span>(region_file, <span class="dt">header =</span> <span class="ot">TRUE</span>)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="kw">colnames</span>(region_info)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chrom&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;stop&quot;</span>)</span>
<span id="cb6-4"><a href="#cb6-4"></a>region_info<span class="op">$</span>chrom &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">gsub</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;&quot;</span>, region_info<span class="op">$</span>chrom))</span>
<span id="cb6-5"><a href="#cb6-5"></a>region_info<span class="op">$</span>region_tag &lt;-<span class="st"> </span><span class="kw">paste0</span>(region_info<span class="op">$</span>chr, <span class="st">&quot;:&quot;</span>, region_info<span class="op">$</span>start, <span class="st">&quot;-&quot;</span>, region_info<span class="op">$</span>stop)</span>
<span id="cb6-6"><a href="#cb6-6"></a></span>
<span id="cb6-7"><a href="#cb6-7"></a>ld_R_dir &lt;-<span class="st"> &quot;/project2/mstephens/wcrouse/UKB_LDR_0.1&quot;</span></span>
<span id="cb6-8"><a href="#cb6-8"></a>filestem &lt;-<span class="st"> &quot;ukb_b38_0.1&quot;</span></span>
<span id="cb6-9"><a href="#cb6-9"></a>ld_filestem &lt;-<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">&quot;%s_chr%s.R_snp.%s_%s&quot;</span>, filestem, region_info<span class="op">$</span>chrom, region_info<span class="op">$</span>start, region_info<span class="op">$</span>stop)</span>
<span id="cb6-10"><a href="#cb6-10"></a>region_info<span class="op">$</span>LD_matrix &lt;-<span class="st"> </span><span class="kw">file.path</span>(ld_R_dir, <span class="kw">paste0</span>(ld_filestem, <span class="st">&quot;.RDS&quot;</span>))</span>
<span id="cb6-11"><a href="#cb6-11"></a>region_info<span class="op">$</span>SNP_info &lt;-<span class="st"> </span><span class="kw">file.path</span>(ld_R_dir, <span class="kw">paste0</span>(ld_filestem, <span class="st">&quot;.Rvar&quot;</span>))</span>
<span id="cb6-12"><a href="#cb6-12"></a></span>
<span id="cb6-13"><a href="#cb6-13"></a>region_info[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,]</span></code></pre></div>
</div>
<div id="ld-matrices" class="section level4">
<h4>LD matrices</h4>
<p>To use LD matrices for the LD reference, provide a directory containing all of the <code>.RDS</code> matrix files and matching <code>.Rvar</code> variant information files.</p>
<p>The complete LD matrices of European individuals from UK Biobank can be downloaded <a href="https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn">here</a>. On the University of Chicago RCC cluster, the b38 reference is available at <code>/project2/mstephens/wcrouse/UKB_LDR_0.1/</code> and the b37 reference is available at <code>/project2/mstephens/wcrouse/UKB_LDR_0.1_b37/</code>.</p>
</div>
</div>
<div id="data-harmonization" class="section level3">
<h3>Data harmonization</h3>
<p>These inputs should be harmonized prior to cTWAS analysis (i.e. the reference and alternative alleles for each variant should match across all three data sources).</p>
<div id="harmonizing-gwas-z-scores-and-ld-reference-and-detect-ld-mismatches" class="section level4">
<h4>Harmonizing GWAS z-scores and LD reference and detect LD mismatches</h4>
<p>The <code>preprocess_z()</code> function first harmonizes GWAS z-scores and LD reference based on the included allele information, then performs LD mismatch diagnosis using the <code>kriging_rss()</code> function from SuSiE RSS, and flips signs for variants detected with allele flipping issues. Note, the LD mismatch diagnosis procedure can be time consuming (we skip this step here).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>res &lt;-<span class="st"> </span><span class="kw">preprocess_z</span>(z_snp, </span>
<span id="cb7-2"><a href="#cb7-2"></a>                    region_info, </span>
<span id="cb7-3"><a href="#cb7-3"></a>                    gwas_n,</span>
<span id="cb7-4"><a href="#cb7-4"></a>                    <span class="dt">drop_multiallelic =</span> <span class="ot">TRUE</span>,</span>
<span id="cb7-5"><a href="#cb7-5"></a>                    <span class="dt">drop_strand_ambig =</span> <span class="ot">TRUE</span>,</span>
<span id="cb7-6"><a href="#cb7-6"></a>                    <span class="dt">detect_ld_mismatch =</span> <span class="ot">FALSE</span>)</span>
<span id="cb7-7"><a href="#cb7-7"></a>z_snp &lt;-<span class="st"> </span>res<span class="op">$</span>z_snp</span>
<span id="cb7-8"><a href="#cb7-8"></a>ld_mismatch_res &lt;-<span class="st"> </span>res<span class="op">$</span>ld_mismatch_res</span></code></pre></div>
</div>
<div id="harmonizing-prediction-models-and-ld-reference" class="section level4">
<h4>Harmonizing prediction models and LD reference</h4>
<p>The <code>preprocess_weight()</code> function harmonizes the PredictDB prediction models and LD reference. This only needs to be done once per combination of prediction models and LD reference.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>weight_fileweight &lt;-<span class="st"> &quot;/project2/xinhe/kevinluo/cTWAS/ctwas_tutorial/weights/eqtl/mashr/mashr_Liver.db&quot;</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>outputdir &lt;-<span class="st"> &quot;/project2/xinhe/kevinluo/cTWAS/ctwas_tutorial/&quot;</span></span>
<span id="cb8-3"><a href="#cb8-3"></a></span>
<span id="cb8-4"><a href="#cb8-4"></a>res &lt;-<span class="st"> </span><span class="kw">preprocess_weight</span>(weight, </span>
<span id="cb8-5"><a href="#cb8-5"></a>                         region_info, </span>
<span id="cb8-6"><a href="#cb8-6"></a>                         <span class="dt">drop_strand_ambig =</span> <span class="ot">TRUE</span>,</span>
<span id="cb8-7"><a href="#cb8-7"></a>                         <span class="dt">write_db =</span> <span class="ot">TRUE</span>,</span>
<span id="cb8-8"><a href="#cb8-8"></a>                         <span class="dt">outputdir =</span> <span class="kw">paste0</span>(outputdir, <span class="st">&quot;/preprocessed_weights/&quot;</span>),</span>
<span id="cb8-9"><a href="#cb8-9"></a>                         <span class="dt">outname =</span> <span class="st">&quot;mashr_Liver_nolnc&quot;</span>)</span>
<span id="cb8-10"><a href="#cb8-10"></a>weight_table &lt;-<span class="st"> </span>res<span class="op">$</span>weight_table</span>
<span id="cb8-11"><a href="#cb8-11"></a>extra_table &lt;-<span class="st"> </span>res<span class="op">$</span>extra_table</span>
<span id="cb8-12"><a href="#cb8-12"></a></span>
<span id="cb8-13"><a href="#cb8-13"></a>preprocessed_weight =<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">paste0</span>(outputdir, <span class="st">&quot;/preprocessed_weights/&quot;</span>), <span class="st">&quot;mashr_Liver_nolnc.db&quot;</span>)</span>
<span id="cb8-14"><a href="#cb8-14"></a>res &lt;-<span class="st"> </span><span class="kw">read_weights</span>(preprocessed_weight, </span>
<span id="cb8-15"><a href="#cb8-15"></a>                    region_info, </span>
<span id="cb8-16"><a href="#cb8-16"></a>                    z_snp, </span>
<span id="cb8-17"><a href="#cb8-17"></a>                    <span class="dt">scale_by_ld_variance=</span><span class="ot">TRUE</span>,</span>
<span id="cb8-18"><a href="#cb8-18"></a>                    <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb8-19"><a href="#cb8-19"></a>weight_list &lt;-<span class="st"> </span>res<span class="op">$</span>weight_list</span>
<span id="cb8-20"><a href="#cb8-20"></a>weight_info &lt;-<span class="st"> </span>res<span class="op">$</span>weight_info</span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="kw">saveRDS</span>(res, <span class="kw">file.path</span>(<span class="kw">paste0</span>(outputdir, <span class="st">&quot;/preprocessed_weights/&quot;</span>), <span class="st">&quot;mashr_Liver_nolnc_processed_weights.rds&quot;</span>))</span></code></pre></div>
</div>
</div>
</div>
<div id="running-ctwas-genome-wide" class="section level2">
<h2>Running cTWAS genome-wide</h2>
<div id="imputing-gene-z-scores" class="section level3">
<h3>Imputing gene z-scores</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>res &lt;-<span class="st"> </span><span class="kw">compute_gene_z</span>(z_snp, region_info, weight_list, weight_info, <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a>gene_info &lt;-<span class="st"> </span>res<span class="op">$</span>gene_info</span>
<span id="cb9-3"><a href="#cb9-3"></a>z_gene &lt;-<span class="st"> </span>res<span class="op">$</span>z_gene</span></code></pre></div>
</div>
<div id="estimating-parameters" class="section level3">
<h3>Estimating parameters</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>thin &lt;-<span class="st"> </span><span class="fl">0.1</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>max_snp_region &lt;-<span class="st"> </span><span class="dv">20000</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>ncore &lt;-<span class="st"> </span><span class="dv">6</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>group_prior_var_structure &lt;-<span class="st"> &quot;independent&quot;</span></span></code></pre></div>
<p>using 6 cores and with 56GB of RAM available, this step took approximately 7 hours.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>param &lt;-<span class="st"> </span><span class="kw">est_param</span>(<span class="dt">z_snp =</span> z_snp,</span>
<span id="cb11-2"><a href="#cb11-2"></a>                   <span class="dt">region_info =</span> region_info,</span>
<span id="cb11-3"><a href="#cb11-3"></a>                   <span class="dt">z_gene =</span> z_gene,</span>
<span id="cb11-4"><a href="#cb11-4"></a>                   <span class="dt">gene_info =</span> gene_info,</span>
<span id="cb11-5"><a href="#cb11-5"></a>                   <span class="dt">thin =</span> thin,</span>
<span id="cb11-6"><a href="#cb11-6"></a>                   <span class="dt">max_snp_region =</span> max_snp_region,</span>
<span id="cb11-7"><a href="#cb11-7"></a>                   <span class="dt">group_prior_var_structure =</span> group_prior_var_structure,</span>
<span id="cb11-8"><a href="#cb11-8"></a>                   <span class="dt">ncore =</span> ncore)</span></code></pre></div>
</div>
<div id="assessing-parameter-estimates" class="section level3">
<h3>Assessing parameter estimates</h3>
<p>We provide code to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>ctwas_parameters &lt;-<span class="st"> </span><span class="kw">summarize_param</span>(param, gwas_n, <span class="dt">thin =</span> thin)</span>
<span id="cb12-2"><a href="#cb12-2"></a></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co"># number of variants in the analysis</span></span>
<span id="cb12-4"><a href="#cb12-4"></a>ctwas_parameters<span class="op">$</span>n_snps</span>
<span id="cb12-5"><a href="#cb12-5"></a></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co"># number of genes in the analysis</span></span>
<span id="cb12-7"><a href="#cb12-7"></a>ctwas_parameters<span class="op">$</span>n_genes</span>
<span id="cb12-8"><a href="#cb12-8"></a></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co"># estimated prior inclusion probability</span></span>
<span id="cb12-10"><a href="#cb12-10"></a>ctwas_parameters<span class="op">$</span>group_prior</span>
<span id="cb12-11"><a href="#cb12-11"></a></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co"># estimated prior effect size</span></span>
<span id="cb12-13"><a href="#cb12-13"></a>ctwas_parameters<span class="op">$</span>group_prior_var</span>
<span id="cb12-14"><a href="#cb12-14"></a></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co"># estimated enrichment of genes over variants</span></span>
<span id="cb12-16"><a href="#cb12-16"></a>ctwas_parameters<span class="op">$</span>enrichment</span>
<span id="cb12-17"><a href="#cb12-17"></a></span>
<span id="cb12-18"><a href="#cb12-18"></a><span class="co"># PVE explained by genes and variants</span></span>
<span id="cb12-19"><a href="#cb12-19"></a>ctwas_parameters<span class="op">$</span>group_pve</span>
<span id="cb12-20"><a href="#cb12-20"></a></span>
<span id="cb12-21"><a href="#cb12-21"></a><span class="co"># total heritability (sum of PVE)</span></span>
<span id="cb12-22"><a href="#cb12-22"></a>ctwas_parameters<span class="op">$</span>total_pve</span>
<span id="cb12-23"><a href="#cb12-23"></a></span>
<span id="cb12-24"><a href="#cb12-24"></a><span class="co"># attributable heritability</span></span>
<span id="cb12-25"><a href="#cb12-25"></a>ctwas_parameters<span class="op">$</span>attributable_pve</span>
<span id="cb12-26"><a href="#cb12-26"></a></span>
<span id="cb12-27"><a href="#cb12-27"></a><span class="co"># plot convergence</span></span>
<span id="cb12-28"><a href="#cb12-28"></a>ctwas_parameters<span class="op">$</span>convergence_plot</span></code></pre></div>
</div>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<p>Here are some details about the computing environment, including the versions of R, and the R packages, used to generate these results.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw">sessionInfo</span>()</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>
