<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using cTWAS with summary statistics • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using cTWAS with summary statistics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.38</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="ctwas_summary_statistics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using cTWAS with summary statistics</h1>
                        <h4 data-toc-skip class="author">Wesley Crouse, Sheng Qian, Kaixuan Luo, Xin He</h4>
            
            <h4 data-toc-skip class="date">2024-09-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/vignettes/ctwas_summary_statistics.Rmd" class="external-link"><code>vignettes/ctwas_summary_statistics.Rmd</code></a></small>
      <div class="hidden name"><code>ctwas_summary_statistics.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This document demonstrates how to use the summary statistics version of cTWAS. Running cTWAS involves four main steps: preparing input data, imputing gene z-scores, estimating parameters, and fine-mapping the genes and variants. The output of cTWAS are posterior inclusion probabilities (PIPs) for all variants and genes with expression models. This document will cover each of these topics in detail. We also describe some of the options available at each step and the analysis considerations behind these options.</p>
<p>We will start by examining the input data and running the cTWAS analysis at a single locus using parameters previously estimated in the cTWAS paper. Running cTWAS with fixed parameters at a single locus is relatively fast, and is the simplest way to use cTWAS. We will then provide a workflow for the same analysis, but using the full data and estimating parameters.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Install <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>,ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
<p>Load <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preparing-the-input-data">Preparing the input data<a class="anchor" aria-label="anchor" href="#preparing-the-input-data"></a>
</h2>
<p>The inputs for the summary statistics version of cTWAS include GWAS summary statistics for variants, prediction models for genes, and LD reference.</p>
<p>We used the directories and files on the University of Chicago RCC cluster as examples. If you are at UChicago, you can load those data from RCC as below.</p>
<div class="section level3">
<h3 id="data-harmonization">Data harmonization<a class="anchor" aria-label="anchor" href="#data-harmonization"></a>
</h3>
<p>There are a few potential problems when preparing the input data. First, the variants in the three sets of input data may not match. Only variants in all three input data will be used in cTWAS. So it is important to maximize the overlap of the variants in the three sets. This can be done for example, by imputing GWAS summary statistics of the variants missing in GWAS but in the LD reference. Another useful pre-processing step is to perform Minor allele frequency (MAF) filtering on the GWAS data so that only those with MAF above a certain cutoff would be used in the analysis, ideally the same cutoff used in the LD references. Additionally, when building the prediction models of gene expression, it is better to impute the genotype data using the LD reference, if possible. All these steps should be done before running cTWAS.</p>
<p>The second potential problem is that the effect alleles in the prediction model, GWAS and LD reference may not agree with each other, thus we need to ``harmonize’’ the data to ensure that the effect alleles match. If the data are not already harmonized, we provide some options for harmonization. Please see the topics on <a href="https://xinhe-lab.github.io/ctwas/articles/harmonization.html" class="external-link">harmonization</a> for more details.</p>
<p>The third potential problem is the LD of the GWAS data (in-sample LD) do not match the reference LD. This can lead to false positives in fine-mapping tools. Diagnostic tools including <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link">SuSiE-RSS</a>, and <a href="https://github.com/Yves-CHEN/DENTIST/" class="external-link">DENTIST</a>, have been developed to check possible LD mismatch.</p>
</div>
<div class="section level3">
<h3 id="gwas-z-scores">GWAS z-scores<a class="anchor" aria-label="anchor" href="#gwas-z-scores"></a>
</h3>
<p>For this analysis, we will use summary statistics from a GWAS of LDL cholesterol in the UK Biobank. We will download the VCF from the IEU Open GWAS Project.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set the working directory, download the summary statistics, and unzip the file.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget https://gwas.mrcieu.ac.uk/files/ukb-d-30780_irnt/ukb-d-30780_irnt.vcf.gz -P gwas_summary_stats"</span><span class="op">)</span></span>
<span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/R.utils/man/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/ukb-d-30780_irnt.vcf.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Next, we will read the summary statistics. Then, we will compute the z-scores and format the input data. We will also collect the sample size, which will be useful later. We will save this output for convenience.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read the data using the VariantAnnotation package</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/ukb-d-30780_irnt.vcf"</span><span class="op">)</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">gwasvcf</span><span class="fu">::</span><span class="fu">vcf_to_tibble</span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute the z-scores</span></span>
<span><span class="va">z_snp</span><span class="op">$</span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">$</span><span class="va">ES</span><span class="op">/</span><span class="va">z_snp</span><span class="op">$</span><span class="va">SE</span></span>
<span></span>
<span><span class="co"># collect sample size (most frequent sample size for all variants)</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">SS</span><span class="op">)</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset the columns and format the column names</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>, <span class="st">"ALT"</span>, <span class="st">"REF"</span>, <span class="st">"Z"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"z"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># drop multiallelic variants (id not unique)</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># save the formatted z-scores and GWAS sample size</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">z_snp</span>, file<span class="op">=</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/ukb-d-30780_irnt.RDS"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">gwas_n</span>, file<span class="op">=</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/gwas_n.RDS"</span><span class="op">)</span></span></code></pre></div>
<p>After the previous step, we can load the data and look at the format. <code>z_snp</code> is a data frame, and each row is a variant. <code>A1</code> is the alternate allele, and <code>A2</code> is the reference allele. The sample size for this GWAS is <code>N=343,621</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/ukb-d-30780_irnt.RDS"</span><span class="op">)</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/gwas_n.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            id A1 A2           z</span></span>
<span><span class="co">## 1 rs530212009  C CA -1.10726803</span></span>
<span><span class="co">## 2  rs12238997  G  A -1.05210759</span></span>
<span><span class="co">## 3 rs371890604  C  G -1.27724687</span></span>
<span><span class="co">## 4 rs144155419  A  G -0.84645309</span></span>
<span><span class="co">## 5 rs189787166  T  A -0.05504189</span></span>
<span><span class="co">## 6 rs148120343  C  T -1.23068731</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample size</span></span>
<span><span class="va">gwas_n</span> </span></code></pre></div>
<pre><code><span><span class="co">## [1] 343621</span></span></code></pre>
<p>If your GWAS summary statistics is of other formats, please extract relevant columns (rsid, alternative allele, reference allele, zscore) and convert it to the format shown above.</p>
</div>
<div class="section level3">
<h3 id="prediction-models">Prediction models<a class="anchor" aria-label="anchor" href="#prediction-models"></a>
</h3>
<p>Prediction models can be specified in either PredictDB or FUSION format. PredictDB is the recommended format, as it has information about correlations among SNPs included in the model. This is useful for recovering strand ambiguous variants. In this user guide, we focus on the PredictDB format, but will provide information on using FUSION format.</p>
<p>In terms of the choice of prediction models, cTWAS performs best when prediction models are sparse, i.e. they have relatively few variants per gene. As the density of variants increases, it becomes computationally more expensive. Dense variants may also lead to a problem with region merging. Basically, if the variants in the prediction model of a gene spans two LD-independent regions, it would be unclear to cTWAS what region the gene should be assigned to. So cTWAS will attempt to merge the two regions. But if many genes have dense variants in their prediction models, region merging could be excessive, leading to very large regions and hurting the performance of cTWAS. Given this consideration, we recommend to choose sparse prediction models such as Lasso. If using dense prediction models, we recommend to remove variants with weights below a threshold from the prediction models.</p>
<p>Often, a research may perform eQTL mapping and have a list of significant eQTLs without explicitly building prediction models. In such cases, it is possible to run cTWAS. This can be done simply by using top eQTL per gene as the prediction model. One can create PredictDB format data from the eQTL list - the details will be added later.</p>
<div class="section level4">
<h4 id="predictdb-format">PredictDB format<a class="anchor" aria-label="anchor" href="#predictdb-format"></a>
</h4>
<p>Please check <a href="http://predictdb.org/" class="external-link">PredictDB</a> for the format of PredictDB weights. To specify weights in PredictDB format, provide the path to the <code>.db</code> file.</p>
<p>For this analysis, we will use liver gene expression models trained on GTEx v8 in the PredictDB format. We will download both the prediction models (<code>.db</code>) and the covariances between variants in the prediction models (<code>.txt.gz</code>). The covariances can optionally be used during harmonization to recover strand ambiguous variants.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget https://zenodo.org/record/3518299/files/mashr_eqtl.tar"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract to ./weights folder </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"mkdir /project2/xinhe/shared_data/ctwas_tutorial/weights"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf mashr_eqtl.tar -C /project2/xinhe/shared_data/ctwas_tutorial/weights"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"rm mashr_eqtl.tar"</span><span class="op">)</span></span></code></pre></div>
<p>In the paper, we used PredictDB models for liver gene expression. We also performed an additional preprocessing step to remove lncRNAs from the prediction models. This can be done using the following code:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org" class="external-link">RSQLite</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify the weight to remove lncRNA from</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights/eqtl/mashr/mashr_Liver.db"</span></span>
<span></span>
<span><span class="co"># read the PredictDB weights</span></span>
<span><span class="va">sqlite</span> <span class="op">&lt;-</span> <span class="fu">dbDriver</span><span class="op">(</span><span class="st">"SQLite"</span><span class="op">)</span></span>
<span><span class="va">db</span> <span class="op">=</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="va">weight</span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">db</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select * from weights"</span><span class="op">)</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select * from extra"</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset to protein coding genes only</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span>  <span class="va">extra_table</span><span class="op">[</span><span class="va">extra_table</span><span class="op">$</span><span class="va">gene_type</span><span class="op">==</span><span class="st">"protein_coding"</span>,,drop<span class="op">=</span><span class="cn">F</span><span class="op">]</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="va">weights_table</span><span class="op">[</span><span class="va">weights_table</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># read and subset the covariances</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/fileutils.html" class="external-link">file_path_sans_ext</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span>, <span class="st">".txt.gz"</span><span class="op">)</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="va">weight_info</span><span class="op">[</span><span class="va">weight_info</span><span class="op">$</span><span class="va">GENE</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># write the .db file and the covariances</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc"</span>, showWarnings<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc.db"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">db</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc.db"</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"extra"</span>, <span class="va">extra_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"weights"</span>, <span class="va">weights_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">weight_info_gz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc.txt.gz"</span>, <span class="st">"w"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">weight_info</span>, <span class="va">weight_info_gz</span>, sep<span class="op">=</span><span class="st">" "</span>, quote<span class="op">=</span><span class="cn">F</span>, row.names<span class="op">=</span><span class="cn">F</span>, col.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">close</a></span><span class="op">(</span><span class="va">weight_info_gz</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># specify the weight for the analysis</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc.db"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="fusion-format">FUSION format<a class="anchor" aria-label="anchor" href="#fusion-format"></a>
</h4>
<p>Please check <a href="http://gusevlab.org/projects/fusion/#compute-your-own-predictive-models" class="external-link">FUSION/TWAS</a> for the format of FUSION weights. Below is an example.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_fusion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_fusion_weights"</span>, <span class="st">"Tissue"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span></code></pre></div>
<p>To specify weights in FUSION format, provide the directory that contains all the <code>.rdata</code> files as above. We assume a file with the same name as the directory with the suffix <code>.pos</code> is present in the same level as the directory. The program will search for this file automatically. For example, we have both the directory <code>Tissue/</code> and <code>Tissue.pos</code> present under the <code>extdata/example_fusion_weights</code> folder. Note that FUSION weights don’t come with covariance information among variants. So you need to turn off the option “recover_strand_ambig_wgt” in the “impute_expr_z” function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># directory and .pos file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">weight_fusion</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Tissue"     "Tissue.pos"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># .rda files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">weight_fusion</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Tissue.gene1.wgt.RDat"  "Tissue.gene10.wgt.RDat" "Tissue.gene11.wgt.RDat"</span></span>
<span><span class="co">##  [4] "Tissue.gene12.wgt.RDat" "Tissue.gene13.wgt.RDat" "Tissue.gene14.wgt.RDat"</span></span>
<span><span class="co">##  [7] "Tissue.gene15.wgt.RDat" "Tissue.gene16.wgt.RDat" "Tissue.gene17.wgt.RDat"</span></span>
<span><span class="co">## [10] "Tissue.gene18.wgt.RDat" "Tissue.gene19.wgt.RDat" "Tissue.gene2.wgt.RDat" </span></span>
<span><span class="co">## [13] "Tissue.gene20.wgt.RDat" "Tissue.gene3.wgt.RDat"  "Tissue.gene4.wgt.RDat" </span></span>
<span><span class="co">## [16] "Tissue.gene5.wgt.RDat"  "Tissue.gene6.wgt.RDat"  "Tissue.gene7.wgt.RDat" </span></span>
<span><span class="co">## [19] "Tissue.gene8.wgt.RDat"  "Tissue.gene9.wgt.RDat"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">weight_fusion</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="ld-reference-and-regions">LD reference and regions<a class="anchor" aria-label="anchor" href="#ld-reference-and-regions"></a>
</h3>
<p>LD reference information can be provided to cTWAS as either individual-level genotype data (in PLINK format), or as genetic correlation matrices (termed “R matrices”) for regions that are approximately LD-independent. These regions must also be specified, regardless of how the LD reference information is provided.</p>
<p>cTWAS performs its analysis region-by-region. If individual-level genotype data is used for the LD reference, variants are assigned to regions, and then correlations matrices are computed within each region. This can be computationally expensive if there are many individuals or if variants in the LD reference are dense. For this reason, the preferred way to run cTWAS is to provide pre-computed LD matrices for each region.</p>
<p>It is critical that the genome build (e.g. hg38) of the LD reference matches the genome build used to train the prediction models. The genome build of the GWAS summary statistics does not matter because variant positions are determined by the LD reference.</p>
<p>The choice of LD reference population is important for fine-mapping. Best practice for fine-mapping is to use in-sample LD reference (LD computed using the subjects in the GWAS sample). If in-sample LD reference is not an option, the LD reference should be as representative of the population in the GWAS sample as possible. Given that cTWAS is an extended fine-mapping algorithm, and that gene z-scores are computed using the observed GWAS z-scores, which reflect patterns of LD in the GWAS population, our recommendation is to match the LD reference to the GWAS population, not the population used to build the prediction models.</p>
<div class="section level4">
<h4 id="defining-regions">Defining regions<a class="anchor" aria-label="anchor" href="#defining-regions"></a>
</h4>
<p>cTWAS includes pre-defined regions based on European (EUR), Asian (ASN), or African (AFR) populations, using either genome build b38 or b37. These regions were previously generated using <a href="https://github.com/endrebak/ldetect" class="external-link">LDetect</a>. These regions are specified in <code>ctwas_rss</code> function using the <code>ld_regions</code> and <code>ld_regions_version</code> arguments, respectively (see function <a href="https://xinhe-lab.github.io/ctwas/reference/index.html" class="external-link">reference</a>).</p>
<p>Here we focus on an example region containing the HPR locus using b38 European region file, which is included in the package.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ldetect"</span>, <span class="st">"EUR.b38.bed"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">regions_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">regions</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">locus_chr</span> <span class="op">&lt;-</span> <span class="st">"chr16"</span></span>
<span><span class="va">locus_start</span> <span class="op">&lt;-</span> <span class="fl">71020125</span></span>
<span></span>
<span><span class="va">regions_df</span> <span class="op">&lt;-</span> <span class="va">regions_df</span><span class="op">[</span><span class="va">regions_df</span><span class="op">$</span><span class="va">chr</span><span class="op">==</span><span class="va">locus_chr</span> <span class="op">&amp;</span> <span class="va">regions_df</span><span class="op">$</span><span class="va">start</span><span class="op">==</span><span class="va">locus_start</span>,<span class="op">]</span></span>
<span><span class="va">regions_df</span></span></code></pre></div>
<pre><code><span><span class="co">##        chr    start     stop</span></span>
<span><span class="co">## 1465 chr16 71020125 72901251</span></span></code></pre>
<p>We could save custom regions in <code>.bed</code> format, which would be used later in our analysis.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/regions"</span>, showWarnings<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">regions_file</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/regions/regions_subset.bed"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">regions_df</span>, file<span class="op">=</span><span class="va">regions_file</span>, row.names<span class="op">=</span><span class="cn">F</span>, col.names<span class="op">=</span><span class="cn">T</span>, sep<span class="op">=</span><span class="st">"\t"</span>, quote <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>Note that the default behavior of cTWAS is to merge regions that have a gene spanning the region boundary (<code>merge=T</code>). For this reason, the regions specified here may not correspond exactly to the final regions analyzed by cTWAS. See the section on <a href="https://xinhe-lab.github.io/ctwas/articles/regionmerging.html" class="external-link">region merging</a> for more details.</p>
</div>
<div class="section level4">
<h4 id="ld-matrices">LD matrices<a class="anchor" aria-label="anchor" href="#ld-matrices"></a>
</h4>
<p>To use LD matrices for the LD reference, provide a directory containing all of the <code>.RDS</code> matrix files and matching <code>.Rvar</code> variant information files.</p>
<p>The complete LD matrices of European individuals from UK Biobank can be downloaded <a href="https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn" class="external-link">here</a>. On the University of Chicago RCC cluster, the b38 reference is available at <code>/project2/mstephens/wcrouse/UKB_LDR_0.1/</code> and the b37 reference is available at <code>/project2/mstephens/wcrouse/UKB_LDR_0.1_b37/</code>.</p>
<p>We illustrated below how cTWAS uses the LD matrices. We have included an LD matrix for the HPR locus that we analyzed in the paper as part of the package. The complete LD matrix for this region was too large to include in the package, so we include only half of the variants at this locus, including the ones needed for the prediction models at this locus.</p>
<p>This example LD matrix can be loaded directly from the package:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_R_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ld_matrices"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">ld_R_dir</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "example_locus_chr16.R_snp.71020125_72901251.RDS" </span></span>
<span><span class="co">## [2] "example_locus_chr16.R_snp.71020125_72901251.Rvar"</span></span></code></pre>
<p>The <code>.RDS</code> file is in <code>.RDS</code> format. It stores the LD correlation matrix for a region (a <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">p \times p</annotation></semantics></math> matrix, <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math> is the number of variants in the region). We require that for each <code>.RDS</code> file, in the same directory, there is a corresponding file with the same stem but ending with the suffix <code>.Rvar</code>. This <code>.Rvar</code> files includes variant information for the region, and the order of its rows must match the order of rows and columns in the <code>.RDS</code> file. The format of these files is:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># correlation matrix</span></span>
<span><span class="va">R_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ld_matrices"</span>, <span class="st">"example_locus_chr16.R_snp.71020125_72901251.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">R_snp</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]      [,2]       [,3]      [,4]      [,5]</span></span>
<span><span class="co">## [1,]  1.0000000 0.2416067 -0.1926180 0.2300365 0.2332161</span></span>
<span><span class="co">## [2,]  0.2416067 1.0000000  0.8464862 0.9471165 0.9738176</span></span>
<span><span class="co">## [3,] -0.1926180 0.8464862  1.0000000 0.8268301 0.8540398</span></span>
<span><span class="co">## [4,]  0.2300365 0.9471165  0.8268301 1.0000000 0.9713106</span></span>
<span><span class="co">## [5,]  0.2332161 0.9738176  0.8540398 0.9713106 1.0000000</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variant info</span></span>
<span><span class="va">R_snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ld_matrices"</span>, <span class="st">"example_locus_chr16.R_snp.71020125_72901251.Rvar"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">R_snp_info</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   chrom         id      pos alt ref   variance allele_freq</span></span>
<span><span class="co">## 1    16  rs9936840 71020448   G   A 0.09481989   0.9419410</span></span>
<span><span class="co">## 2    16   rs917007 71022770   G   A 0.47443074   0.5229496</span></span>
<span><span class="co">## 3    16 rs11648149 71025765   G   A 0.45222513   0.6014605</span></span>
<span><span class="co">## 4    16  rs5817733 71028512   A   C 0.45914416   0.5484138</span></span>
<span><span class="co">## 5    16 rs11647909 71029403   T   C 0.49039131   0.5220587</span></span></code></pre>
<p>The columns of the <code>.Rvar</code> file include information on chromosome, variant name, position in base pairs, and the alternative and reference alleles. The <code>variance</code> column is the variance of each variant prior to standardization; this is required for PredictDB weights but not FUSION weights. PredictDB weights should be scaled by the variance before imputing gene expression. This is because PredictDB weights assume that variant genotypes are not standardized before imputation, but our implementation assumes standardized variant genotypes. If variance information is missing, or if weights are in PredictDB format but are already on the standardized scale (e.g. if they were converted from FUSION to PredictDB format), this scaling can be turned off using the option <code>scale_by_ld_variance=F</code> using the multigroup version of cTWAS (this option hasn’t been based forward to the main branch yet). We’ve also include information on allele frequency in the variant info, but this is optional and not used by cTWAS.</p>
<p>The naming convention for the LD matrices is <code>[filestem]_chr[#].R_snp.[start]_[end].RDS</code>. cTWAS expects that all <code>.RDS</code> and <code>.Rvar</code> files in the directory contain LD information, so no other files with these suffixes should be in the directory. Each variant should be uniquely assigned to a region, and the regions should be left closed and right open, i.e. [start, stop). The positions of the LD matrices must match exactly the positions specified by the region file. Do not include invariant or multiallelic variants in the LD reference.</p>
</div>
<div class="section level4">
<h4 id="convert-genotype-data-to-ld-matrices">Convert genotype data to LD matrices<a class="anchor" aria-label="anchor" href="#convert-genotype-data-to-ld-matrices"></a>
</h4>
<p>We could use the <code>convert_geno_to_LD_matrix</code> function to convert genotype files (and LD regions) to the LD matrices and corresponding variant information.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify LD reference</span></span>
<span><span class="va">ldref_dir</span> <span class="op">&lt;-</span> <span class="st">"/gpfs/data/xhe-lab/ukb_LDR/genotype_data_0.1"</span></span>
<span><span class="va">genotype_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ldref_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ukb_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".pgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># the output Rvar files use the positions and allele information in varinfo_files</span></span>
<span><span class="va">varinfo_files</span> <span class="op">&lt;-</span> <span class="st">"/gpfs/data/xhe-lab/ukb_LDR/neale_lab/neale_variants_hg38.bim"</span></span>
<span><span class="co"># prepare a data frame region_info for LD regions with columns "chr", "start", and "stop"</span></span>
<span><span class="co"># the positions should match those in varinfo_files</span></span>
<span><span class="va">region_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ldetect"</span>, <span class="st">"EUR.b38.bed"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">region_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># specify output</span></span>
<span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/gpfs/data/xhe-lab/ctwas/LDR/UKB_b38/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"ukb_b38_0.1"</span></span>
<span></span>
<span><span class="co"># generate LD matrices (.RDS) and variant info (.Rvar)</span></span>
<span><span class="co"># update region info with paths of LD matrices and variant info files</span></span>
<span><span class="co"># we could run this for one chromosome or multiple chromosomes</span></span>
<span><span class="va">updated_region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convert_geno_to_LD_matrix.html">convert_geno_to_LD_matrix</a></span><span class="op">(</span><span class="va">region_info</span>, </span>
<span>                                                 <span class="va">genotype_files</span>, </span>
<span>                                                 <span class="va">varinfo_files</span>,</span>
<span>                                                 chrom <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>                                                 outputdir <span class="op">=</span> <span class="va">outputdir</span>, </span>
<span>                                                 outname <span class="op">=</span> <span class="va">outname</span><span class="op">)</span></span></code></pre></div>
<p>We provide example R scripts to generate the LD matrices for UKB or 1KG European genotype files.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to UKB b38 script</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/scripts"</span>, <span class="st">"convert_geno_to_LD_matrix_UKB_b38.R"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># path to UKB b37 script</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/scripts"</span>, <span class="st">"convert_geno_to_LD_matrix_UKB_b37.R"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># path to 1KG European b37 script</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/scripts"</span>, <span class="st">"convert_geno_to_LD_matrix_1KG_EUR_b37.R"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-ctwas-at-a-single-locus">Running cTWAS at a single locus<a class="anchor" aria-label="anchor" href="#running-ctwas-at-a-single-locus"></a>
</h2>
<p>To speed computation in our single locus example, we will subset the prediction models to only the genes at the HPR locus that we analyzed in the paper. Subsetting the prediction models is not strictly necessary (simply specifying a single region for the fine-mapping step will yield the same output), but subsetting the weights lets us avoid imputing all the genes during the imputation step.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the genes to subset to:</span></span>
<span><span class="va">gene_subset</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CMTR2"</span>, <span class="st">"ZNF23"</span>, <span class="st">"CHST4"</span>, <span class="st">"ZNF19"</span>, <span class="st">"TAT"</span>, <span class="st">"MARVELD3"</span>, <span class="st">"PHLPP2"</span>, <span class="st">"ATXN1L"</span>, <span class="st">"ZNF821"</span>, <span class="st">"PKD1L3"</span>, <span class="st">"HPR"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># subset to selected genes only</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span>  <span class="va">extra_table</span><span class="op">[</span><span class="va">extra_table</span><span class="op">$</span><span class="va">genename</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gene_subset</span>,,drop<span class="op">=</span><span class="cn">F</span><span class="op">]</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="va">weights_table</span><span class="op">[</span><span class="va">weights_table</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># subset the covariances</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="va">weight_info</span><span class="op">[</span><span class="va">weight_info</span><span class="op">$</span><span class="va">GENE</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># write the .db file and the covariances</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc_subset.db"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">db</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc_subset.db"</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"extra"</span>, <span class="va">extra_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"weights"</span>, <span class="va">weights_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">weight_info_gz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc_subset.txt.gz"</span>, <span class="st">"w"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">weight_info</span>, <span class="va">weight_info_gz</span>, sep<span class="op">=</span><span class="st">" "</span>, quote<span class="op">=</span><span class="cn">F</span>, row.names<span class="op">=</span><span class="cn">F</span>, col.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">close</a></span><span class="op">(</span><span class="va">weight_info_gz</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># specify the weight for the analysis</span></span>
<span><span class="va">weight_subset</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/weights_nolnc/mashr_Liver_nolnc_subset.db"</span></span></code></pre></div>
<p>We also included this final subset of liver prediction models in the cTWAS package.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/weights_nolnc"</span>, <span class="st">"mashr_Liver_nolnc_subset.db"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span></code></pre></div>
<p>We will also subset the GWAS z-scores to only the variants in this region:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_snp_subset</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">[</span><span class="va">z_snp</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">R_snp_info</span><span class="op">$</span><span class="va">id</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">z_snp_subset</span>, file<span class="op">=</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/gwas_summary_stats/ukb-d-30780_irnt_subset.RDS"</span><span class="op">)</span></span></code></pre></div>
<p>We’ve also included this file as part of the package:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_snp_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/summary_stats"</span>, <span class="st">"ukb-d-30780_irnt_subset.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="imputing-gene-z-scores">Imputing gene z-scores<a class="anchor" aria-label="anchor" href="#imputing-gene-z-scores"></a>
</h3>
<p>Now that the input data is prepared, we are ready to impute gene z-scores using cTWAS. This step basically performs Transcriptome-wide association studies (TWAS) on each gene with a prediction model. The underlying calculation is based on <a href="https://www.nature.com/articles/s41467-018-03621-1" class="external-link">S-PrediXcan</a>. To impute gene z-scores at this locus, we use the <code>impute_expr_z</code> function. Here, we use the option <code>harmonize_z = T</code> and <code>harmonize_wgt = T</code> to harmonize both the z-scores and the weights to the LD reference. Because the z-scores and the LD reference are from the same source (UK Biobank), we expect that the strand is consistent between the GWAS and LD reference, so we use the option <code>strand_ambig_action_z = "none"</code> to treat strand ambiguous variants as unambiguous. The prediction models are from a different population (GTEx), so we use the option <code>recover_strand_ambig_wgt</code> to recover strand ambiguous variants in the prediction models. The detailed procedure is described in the paper.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/results/single_locus/"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outputdir</span>, showWarnings<span class="op">=</span><span class="cn">F</span>, recursive<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example_locus"</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># impute gene z-scores</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_expr_z.html">impute_expr_z</a></span><span class="op">(</span>z_snp <span class="op">=</span> <span class="va">z_snp_subset</span>,</span>
<span>                     weight <span class="op">=</span> <span class="va">weight_subset</span>,</span>
<span>                     ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                     outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                     outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                     harmonize_z <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     harmonize_wgt <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     strand_ambig_action_z <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                     recover_strand_ambig_wgt <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>The log tells us that there is only chromosome information for chromosome 16, as expected. For each chromosome, it also tells us that we’ve harmonized (flipped) the GWAS z-scores and weights, and that we’ve harmonized strand ambiguous variants for the weights. Chromsome 16 has 11 prediction models and all of them are successfully imputed. The warnings tell us that 21 of the chromosomes did not have any imputed genes.</p>
<p>We’ve now successfully harmonized the data and imputed the gene z-scores. The function has returned a list containing the imputed gene z-scores, the harmonized GWAS z-scores. We will store them for use in the next step. Make sure to store the harmonized GWAS z-scores or this will introduce inconsistencies during the fine-mapping step.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_gene_subset</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">z_gene</span> <span class="co"># imputed gene z-scores</span></span>
<span><span class="va">z_snp_subset</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">z_snp</span> <span class="co"># harmonized GWAS z-scores</span></span>
<span><span class="va">ld_exprvarfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".exprvar"</span><span class="op">)</span> <span class="co"># position information for imputed genes</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_gene_subset</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_snp_subset</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In the directory, we have generated 3 files per chromosome during gene imputation. The <code>exprqc.Rd</code> files contain QC information about imputation for each gene. The <code>.exprvar</code> file contains position information for each imputed gene, including start positions and end positions; these positions are determined by the first and last variant positions for each gene prediction model. The <code>_ld_R_*.txt</code> file contains information about the LD regions and matrices used.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">outputdir</span>, pattern<span class="op">=</span><span class="st">"chr16"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "example_locus_chr16.expr.gz"   "example_locus_chr16.exprqc.Rd"</span></span>
<span><span class="co">## [3] "example_locus_chr16.exprvar"   "example_locus_ld_R_chr16.txt"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="fine-mapping-with-fixed-parameters">Fine-mapping with fixed parameters<a class="anchor" aria-label="anchor" href="#fine-mapping-with-fixed-parameters"></a>
</h3>
<p>After imputing gene z-scores, we are ready to run the cTWAS analysis. The full analysis involves first estimating parameters from the data, and then fine-mapping the genes and variants using these estimated parameters. This can be computationally intensive, so for this example, we will only run the fine-mapping step at the HPR locus, using the parameters we estimated in the paper.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load imputed gene z-scores, the harmonized GWAS z-scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that cTWAS expects the output of <code>impute_expr_z</code>. Currently, it is not possible to specify gene z-scores obtained using other software. This is because we cannot ensure that the LD reference used for imputation with other software is the same as the LD reference used for fine-mapping. This scenario would be fine for parameter estimation, which does not depend on LD, but it would be problematic for fine-mapping.</p>
<p>Let’s run the cTWAS fine-mapping step at the HPR locus using the estimated parameters:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the estimated prior inclusion probabilities for genes and variants from the paper</span></span>
<span><span class="va">group_prior</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0107220302</span>, <span class="fl">0.0001715896</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the estimated effect sizes for genes and variants from the paper</span></span>
<span><span class="va">group_prior_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">41.327666</span>, <span class="fl">9.977841</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run ctwas_rss</span></span>
<span><span class="fu"><a href="../reference/ctwas_rss.html">ctwas_rss</a></span><span class="op">(</span>z_gene <span class="op">=</span> <span class="va">z_gene_subset</span>,</span>
<span>          z_snp <span class="op">=</span> <span class="va">z_snp_subset</span>,</span>
<span>          ld_exprvarfs <span class="op">=</span> <span class="va">ld_exprvarfs</span>,</span>
<span>          ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>          ld_regions_custom <span class="op">=</span> <span class="va">regions_file</span>,</span>
<span>          outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>          outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>          estimate_group_prior <span class="op">=</span> <span class="cn">F</span>,</span>
<span>          estimate_group_prior_var <span class="op">=</span> <span class="cn">F</span>,</span>
<span>          group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>          group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span><span class="op">)</span></span></code></pre></div>
<p>Most of the arguments of this function should be self-explained. The arguments <code>estimate_group_prior</code> and <code>estimate_group_prior_var</code> control whether cTWAS will perform estimation of the two sets of parameters: the prior inclusion probabilities and the prior effect size variance. The <code>outputdir</code> argument specifies the directory to save the output files, and the <code>outname</code> argument is used as the prefix for the output files.</p>
<p>This function computed variant-by-gene and gene-by-gene correlation information (“R matrix info”) for all regions, by chromosome. These matrices are saved in a directory called <code>[outname]_LDR</code>. Finally, cTWAS ran a single iteration of SuSiE for all regions and reported the parameters used.</p>
<p>In the output directory, we’ve created a number of files. We’ve created the <code>[outname]_LDR</code> directory with the gene correlation information; this folder is not created if using genotype information instead of LD matrices. The <code>[outname]_ld_R_*.txt</code> files contains information about the LD regions and matrices used. We also created two files related to region indexing, <code>[outname].regions.txt</code> and <code>[outname].regionlist.RDS</code>.</p>
<p>The cTWAS results are in the <code>[outname].susieIrss.txt</code> file. This file has an entry for each gene and variant reporting the PIP (<code>susie_pip</code>), its confidence set (<code>cs_index</code>), and its effect size (<code>mu2</code>). It also reports information on the confidence set (<code>cs_index</code>) that each gene or variant is assigned to.</p>
</div>
<div class="section level3">
<h3 id="viewing-the-results">Viewing the results<a class="anchor" aria-label="anchor" href="#viewing-the-results"></a>
</h3>
<p>We will add the gene names to the results (the PredictDB weights use Ensembl IDs as the primary identifier), as well as the z-scores for each SNP and gene, and sort the result by the PIP:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load cTWAS results</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">".susieIrss.txt"</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load gene information from PredictDB</span></span>
<span><span class="va">sqlite</span> <span class="op">&lt;-</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbDriver</span><span class="op">(</span><span class="st">"SQLite"</span><span class="op">)</span></span>
<span><span class="va">db</span> <span class="op">=</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="va">weight_subset</span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">db</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select gene, genename, gene_type from extra"</span><span class="op">)</span></span>
<span><span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add gene names to cTWAS results</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gene_info</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span>, <span class="va">gene_info</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add z-scores to cTWAS results</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_snp_subset</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">z_snp_subset</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_gene_subset</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span>, <span class="va">z_gene_subset</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># display the results for the top 10 PIPs</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ctwas_res</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      chrom                id      pos type region_tag1 region_tag2 cs_index</span></span>
<span><span class="co">## 11      16 ENSG00000261701.6 72063820 gene          16           1        1</span></span>
<span><span class="co">## 1303    16        rs77303550 72045758  SNP          16           1        2</span></span>
<span><span class="co">## 1299    16          rs763665 72044144  SNP          16           1        5</span></span>
<span><span class="co">## 1347    16          rs217181 72080103  SNP          16           1        2</span></span>
<span><span class="co">## 1379    16        rs11075921 72098230  SNP          16           1        5</span></span>
<span><span class="co">## 1513    16        rs35549608 72193499  SNP          16           1        5</span></span>
<span><span class="co">## 919     16        rs10459804 71799928  SNP          16           1        4</span></span>
<span><span class="co">## 1363    16        rs12708925 72088807  SNP          16           1        3</span></span>
<span><span class="co">## 1366    16         rs3852781 72089657  SNP          16           1        3</span></span>
<span><span class="co">## 1516    16         rs9923575 72196213  SNP          16           1        5</span></span>
<span><span class="co">##       susie_pip       mu2 genename          z</span></span>
<span><span class="co">## 11   1.00000000 328.21450      HPR -17.962770</span></span>
<span><span class="co">## 1303 0.82978774 168.96398     &lt;NA&gt;  13.732910</span></span>
<span><span class="co">## 1299 0.71516249  78.39512     &lt;NA&gt;  11.285714</span></span>
<span><span class="co">## 1347 0.17021637 165.09713     &lt;NA&gt;  13.553655</span></span>
<span><span class="co">## 1379 0.11989262  71.79442     &lt;NA&gt;  11.020351</span></span>
<span><span class="co">## 1513 0.08657578  70.24465     &lt;NA&gt;  10.532917</span></span>
<span><span class="co">## 919  0.08543960  47.84390     &lt;NA&gt;  -9.159242</span></span>
<span><span class="co">## 1363 0.07087696 125.64998     &lt;NA&gt;  -2.482822</span></span>
<span><span class="co">## 1366 0.06519302 125.58267     &lt;NA&gt;  -2.493077</span></span>
<span><span class="co">## 1516 0.06451556  67.92297     &lt;NA&gt;   8.067235</span></span></code></pre>
<p>To plot the results, we also want to update the gene positions. cTWAS assigns gene positions based on the first variant in the gene’s prediction model, but we want to visualize gene locations using their transcription start site. First, we get gene information for all genes on chromosome 16 and subset to protein coding genes.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download all entries for ensembl on chromosome 16</span></span>
<span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useEnsembl.html" class="external-link">useEnsembl</a></span><span class="op">(</span>biomart<span class="op">=</span><span class="st">"ENSEMBL_MART_ENSEMBL"</span>, dataset<span class="op">=</span><span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span></span>
<span><span class="va">G_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html" class="external-link">getBM</a></span><span class="op">(</span>filters<span class="op">=</span> <span class="st">"chromosome_name"</span>, attributes<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hgnc_symbol"</span>,<span class="st">"chromosome_name"</span>,<span class="st">"start_position"</span>,<span class="st">"end_position"</span>,<span class="st">"gene_biotype"</span>, <span class="st">"ensembl_gene_id"</span>, <span class="st">"strand"</span><span class="op">)</span>, values<span class="op">=</span><span class="fl">16</span>, mart<span class="op">=</span><span class="va">ensembl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset to protein coding genes and fix empty gene names</span></span>
<span><span class="va">G_list</span> <span class="op">&lt;-</span> <span class="va">G_list</span><span class="op">[</span><span class="va">G_list</span><span class="op">$</span><span class="va">gene_biotype</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"protein_coding"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">G_list</span><span class="op">$</span><span class="va">hgnc_symbol</span><span class="op">[</span><span class="va">G_list</span><span class="op">$</span><span class="va">hgnc_symbol</span><span class="op">==</span><span class="st">""</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"-"</span></span>
<span></span>
<span><span class="co"># set TSS based on start/end position and strand</span></span>
<span><span class="va">G_list</span><span class="op">$</span><span class="va">tss</span> <span class="op">&lt;-</span> <span class="va">G_list</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"end_position"</span>, <span class="st">"start_position"</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">G_list</span><span class="op">)</span>,<span class="va">G_list</span><span class="op">$</span><span class="va">strand</span><span class="op">/</span><span class="fl">2</span><span class="op">+</span><span class="fl">1.5</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">G_list</span>, file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"G_list.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we update the position for each gene to the TSS.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"G_list.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove the version number from the ensembl IDs</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">ensembl</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"[.]"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># update the gene positions to TSS</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">pos</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">G_list</span><span class="op">$</span><span class="va">tss</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">ensembl</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span>, <span class="va">G_list</span><span class="op">$</span><span class="va">ensembl_gene_id</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>And we are now ready to plot the results at the HPR locus. There are some limited options to control the location of legends and labels, but making a publication-ready plot will probably require manual adjustment of these features. Note that the plot does not exactly match the figure in the paper because we are only using a subset of the variants in our example LD reference.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># genome-wide Bonferroni threshold used for TWAS</span></span>
<span><span class="va">twas_sig_thresh</span> <span class="op">&lt;-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">9881</span></span>
<span></span>
<span><span class="fu"><a href="../reference/ctwas_locus_plot.html">ctwas_locus_plot</a></span><span class="op">(</span>ctwas_res <span class="op">=</span> <span class="va">ctwas_res</span>,</span>
<span>                 region_tag <span class="op">=</span> <span class="st">"16_1"</span>,</span>
<span>                 xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6</span>,<span class="fl">72.4</span><span class="op">)</span>,</span>
<span>                 ymax_twas <span class="op">=</span> <span class="fl">85</span>,</span>
<span>                 twas_sig_thresh <span class="op">=</span> <span class="va">twas_sig_thresh</span>,</span>
<span>                 alt_names <span class="op">=</span> <span class="st">"genename"</span>, <span class="co"># the column that specify gene names</span></span>
<span>                 legend_panel <span class="op">=</span> <span class="st">"TWAS"</span>, <span class="co"># the panel to plot legend</span></span>
<span>                 legend_side<span class="op">=</span><span class="st">"left"</span>, <span class="co"># the position of the panel</span></span>
<span>                 outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                 outname <span class="op">=</span> <span class="va">outname</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ctwas_summary_statistics_files/figure-html/locus_plot_HPR-1.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-ctwas-genome-wide">Running cTWAS genome-wide<a class="anchor" aria-label="anchor" href="#running-ctwas-genome-wide"></a>
</h2>
<p>We now provide a workflow for the same analysis, but using the full data and estimating parameters. This can be computationally intensive. We performed these analyses with 6 cores and 56GB of RAM, although the exact resource requirements will vary with the numbers of genes and variants provided.</p>
<div class="section level3">
<h3 id="imputing-gene-z-scores-1">Imputing gene z-scores<a class="anchor" aria-label="anchor" href="#imputing-gene-z-scores-1"></a>
</h3>
<p>To impute gene z-scores genome-wide, we specify the full set of GWAS summary statistics, the full PredictDB liver weights, and the full set of LD reference matrices. This step can be slow, especially with both of the advanced harmonization options turned on. There is a parallelized version of this function in the <code>multigroup</code> branch of cTWAS, currently under development. As specified here, imputation took approximately 6 hours, using a single core (the only option using the <code>main</code> branch).</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/results/whole_genome/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example_genome"</span></span>
<span></span>
<span><span class="va">ld_R_dir</span> <span class="op">&lt;-</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1/"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outputdir</span>, showWarnings<span class="op">=</span><span class="cn">F</span>, recursive<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get gene z-scores</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">ld_exprvarfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".exprvar"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_expr_z.html">impute_expr_z</a></span><span class="op">(</span>z_snp <span class="op">=</span> <span class="va">z_snp</span>,</span>
<span>                     weight <span class="op">=</span> <span class="va">weight</span>,</span>
<span>                     ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                     outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                     outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                     harmonize_z <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     harmonize_wgt <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     strand_ambig_action_z <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                     recover_strand_ambig_wgt <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span>  <span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">z_snp</span></span>
<span>  <span class="va">ld_exprvarfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".exprvar"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_gene</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_snp</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimating-parameters-and-fine-mapping">Estimating parameters and fine-mapping<a class="anchor" aria-label="anchor" href="#estimating-parameters-and-fine-mapping"></a>
</h3>
<p>Now that we’ve imputed gene z-scores for the full genome, we are ready to run the full cTWAS analysis. The full analysis involves first estimating parameters from the data, and then fine-mapping the genes and variants using these estimated parameters.</p>
<p>As mentioned previously, the full cTWAS analysis can computationally expensive, so we will specify several options to make the analysis faster. The <code>thin</code> argument randomly selects 10% of variants to use during the parameter estimation and initial fine-mapping steps, reducing computation. The <code>max_snp_region</code> sets a maximum on the number of variants that can be in a single (merged) region to prevent memory issues during fine-mapping. The <code>ncore</code> argument specifies the number of cores to use when parallelizing over regions.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">thin</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">max_snp_region</span> <span class="op">&lt;-</span> <span class="fl">20000</span></span>
<span><span class="va">ncore</span> <span class="op">&lt;-</span> <span class="fl">6</span></span></code></pre></div>
<p>We pass these arguments the <code>ctwas_rss</code> and run the full analysis. As specified, using 6 cores and with 56GB of RAM available, this step took approximately 7 hours.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimating parameters</span></span>
<span><span class="fu"><a href="../reference/ctwas_rss.html">ctwas_rss</a></span><span class="op">(</span>z_gene <span class="op">=</span> <span class="va">z_gene</span>,</span>
<span>          z_snp <span class="op">=</span> <span class="va">z_snp</span>,</span>
<span>          ld_exprvarfs <span class="op">=</span> <span class="va">ld_exprvarfs</span>,</span>
<span>          ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>          ld_regions <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>          ld_regions_version <span class="op">=</span> <span class="st">"b38"</span>,</span>
<span>          outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>          outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>          thin <span class="op">=</span> <span class="va">thin</span>,</span>
<span>          max_snp_region <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>          ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span></code></pre></div>
<p>The files in the output directory are similar to the single locus example, but there are several additional files since we’ve performed parameter estimation. Parameter estimation involves two steps. The first step obtains a rough estimate for the parameters. These estimates are saved in the <code>*.s1.susieIrssres.Rd</code> file. The SuSiE output from the last iteration in this step is saved in <code>*.s1.susieIrss.txt</code>; this file contains the total number of variants analyzed by cTWAS, as used in PVE estimation, scaled by <code>thin</code>. The second step obtains a more precise estimate for the parameters using a subset of regions. These estimates are saved in the <code>*.s2.susieIrssres.Rd</code> file; the final entry of this file is the estimated parameters, scaled by the <code>thin</code> parameter. The <code>*.s2.susieIrss.txt</code> file contains the SuSiE output from the final iteration of this step.</p>
<p>After parameter estimation, cTWAS performs a first pass at fine-mapping, using the proportion of variants specified in <code>thin</code>. These results are saved in the <code>*.s3.susieIrss.txt</code>. If <code>thin</code> is specified, then for regions with a gene having PIP greater than <code>rerun_gene_PIP</code>, cTWAS will make a final pass, analyzing these regions using the full set of variants. The default value of <code>rerun_gene_PIP</code> is 0.8, but it can be changed by the user. The final output of cTWAS is <code>*.susieIrss.txt</code>; this file contains results for all genes, all variants in regions with strong gene signals, and 10% of variants in other regions.</p>
</div>
<div class="section level3">
<h3 id="assessing-parameter-estimates">Assessing parameter estimates<a class="anchor" aria-label="anchor" href="#assessing-parameter-estimates"></a>
</h3>
<p>We provide code to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_summarize_parameters.html">ctwas_summarize_parameters</a></span><span class="op">(</span>outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                                               outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                                               gwas_n <span class="op">=</span> <span class="va">gwas_n</span>,</span>
<span>                                               thin <span class="op">=</span> <span class="va">thin</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># number of variants in the analysis</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">n_snps</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 8696600</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of genes in the analysis</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">n_genes</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9881</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimated prior inclusion probability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior</span></span></code></pre></div>
<pre><code><span><span class="co">##         gene          SNP </span></span>
<span><span class="co">## 0.0107220302 0.0001715896</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimated prior effect size</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span></code></pre></div>
<pre><code><span><span class="co">##      gene       SNP </span></span>
<span><span class="co">## 41.327666  9.977841</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimated enrichment of genes over variants</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">enrichment</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 62.48649</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PVE explained by genes and variants</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##       gene        SNP </span></span>
<span><span class="co">## 0.01274204 0.04333086</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># total heritability (sum of PVE)</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">total_pve</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.0560729</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># attributable heritability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">attributable_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##      gene       SNP </span></span>
<span><span class="co">## 0.2272407 0.7727593</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot convergence</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">convergence_plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ctwas_summary_statistics_files/figure-html/summarize_parameters-1.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
</div>
<div class="section level3">
<h3 id="viewing-the-results-1">Viewing the results<a class="anchor" aria-label="anchor" href="#viewing-the-results-1"></a>
</h3>
<p>As before, we will add the gene names to the results (the PredictDB weights use Ensembl IDs as the primary identifier), as well as the z-scores for each SNP and gene. We then show all genes with PIP greater than 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load cTWAS results</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">".susieIrss.txt"</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load gene information from PredictDB</span></span>
<span><span class="va">sqlite</span> <span class="op">&lt;-</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbDriver</span><span class="op">(</span><span class="st">"SQLite"</span><span class="op">)</span></span>
<span><span class="va">db</span> <span class="op">=</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="va">weight</span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">db</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select gene, genename, gene_type from extra"</span><span class="op">)</span></span>
<span><span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add gene names to cTWAS results</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gene_info</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span>, <span class="va">gene_info</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add z-scores to cTWAS results</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_gene</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">]</span>, <span class="va">z_gene</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># display the genes with PIP &gt; 0.8</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">ctwas_res</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span> <span class="op">&amp;</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         chrom                 id       pos type region_tag1 region_tag2</span></span>
<span><span class="co">## 879255      1 ENSG00000134222.16 109275684 gene           1          67</span></span>
<span><span class="co">## 1064449    16  ENSG00000261701.6  72063820 gene          16          38</span></span>
<span><span class="co">## 913684      2 ENSG00000125629.14 118088372 gene           2          69</span></span>
<span><span class="co">## 910002      2  ENSG00000143921.6  43839108 gene           2          27</span></span>
<span><span class="co">## 1027995    11 ENSG00000149485.18  61829161 gene          11          34</span></span>
<span><span class="co">## 957882      6 ENSG00000204599.14  30324306 gene           6          24</span></span>
<span><span class="co">## 999614      9 ENSG00000165029.15 104906792 gene           9          53</span></span>
<span><span class="co">## 990822      8 ENSG00000173273.15   9315699 gene           8          12</span></span>
<span><span class="co">## 1049636    13 ENSG00000183087.14 113849020 gene          13          62</span></span>
<span><span class="co">## 1111836    20 ENSG00000100979.14  45906012 gene          20          28</span></span>
<span><span class="co">## 1096890    19 ENSG00000105287.12  46713856 gene          19          33</span></span>
<span><span class="co">## 923078      2  ENSG00000163083.5 120552693 gene           2          70</span></span>
<span><span class="co">## 894050      1 ENSG00000143771.11 224356827 gene           1         114</span></span>
<span><span class="co">## 967636      7 ENSG00000105866.13  21553355 gene           7          19</span></span>
<span><span class="co">## 950469      5 ENSG00000151292.17 123513182 gene           5          75</span></span>
<span><span class="co">## 1093233    19  ENSG00000255974.7  40847202 gene          19          28</span></span>
<span><span class="co">## 1096921    19 ENSG00000176920.11  48700572 gene          19          33</span></span>
<span><span class="co">## 1088486    18 ENSG00000119537.15  63367187 gene          18          35</span></span>
<span><span class="co">## 977014      7 ENSG00000015520.14  44541277 gene           7          32</span></span>
<span><span class="co">## 977015      7 ENSG00000136271.10  44575121 gene           7          32</span></span>
<span><span class="co">## 992988      9 ENSG00000155158.20  15280189 gene           9          13</span></span>
<span><span class="co">## 987201      7 ENSG00000087087.18 100875204 gene           7          62</span></span>
<span><span class="co">## 1005378     9  ENSG00000160447.6 128703202 gene           9          66</span></span>
<span><span class="co">## 1010141    10 ENSG00000119965.12 122945179 gene          10          77</span></span>
<span><span class="co">## 938455      5 ENSG00000152684.10  52787392 gene           5          31</span></span>
<span><span class="co">## 1084624    17  ENSG00000173757.9  42276835 gene          17          25</span></span>
<span><span class="co">## 928668      2 ENSG00000123612.15 157625480 gene           2          94</span></span>
<span><span class="co">## 867597      1 ENSG00000162607.12  62436136 gene           1          39</span></span>
<span><span class="co">## 852390      1  ENSG00000179023.8  18480845 gene           1          13</span></span>
<span><span class="co">## 1017410    11 ENSG00000177685.16    827713 gene          11           1</span></span>
<span><span class="co">## 1021453    11 ENSG00000179119.14  18634724 gene          11          13</span></span>
<span><span class="co">## 987196      7  ENSG00000172336.4 100705751 gene           7          62</span></span>
<span><span class="co">## 861235      1 ENSG00000142765.17  27342119 gene           1          19</span></span>
<span><span class="co">## 901554      2  ENSG00000151360.9   3594771 gene           2           2</span></span>
<span><span class="co">## 1042437    12  ENSG00000118971.7   4275678 gene          12           4</span></span>
<span><span class="co">##         cs_index susie_pip        mu2 genename          z</span></span>
<span><span class="co">## 879255         1 1.0000000 1667.48423    PSRC1 -41.687336</span></span>
<span><span class="co">## 1064449        1 0.9999997  162.98924      HPR -17.962770</span></span>
<span><span class="co">## 913684         1 0.9999957   68.37263   INSIG2  -8.982702</span></span>
<span><span class="co">## 910002         1 0.9999453  312.10767    ABCG8 -20.293982</span></span>
<span><span class="co">## 1027995        1 0.9998404  163.46291    FADS1  12.926351</span></span>
<span><span class="co">## 957882         1 0.9985885   71.89752   TRIM39   8.840164</span></span>
<span><span class="co">## 999614         4 0.9955314   70.13575    ABCA1   7.982017</span></span>
<span><span class="co">## 990822         1 0.9910883   76.13390     TNKS  11.038564</span></span>
<span><span class="co">## 1049636        1 0.9883382   71.11508     GAS6  -8.923688</span></span>
<span><span class="co">## 1111836        2 0.9883283   61.28528     PLTP  -5.732491</span></span>
<span><span class="co">## 1096890        2 0.9871587   29.99928    PRKD2   5.072217</span></span>
<span><span class="co">## 923078         1 0.9825109   73.79864    INHBB  -8.518936</span></span>
<span><span class="co">## 894050         1 0.9789702   40.69262    CNIH4   6.145535</span></span>
<span><span class="co">## 967636         1 0.9759456  101.98305      SP4  10.693191</span></span>
<span><span class="co">## 950469         1 0.9745191   83.85862  CSNK1G3   9.116291</span></span>
<span><span class="co">## 1093233        1 0.9650551   31.87939   CYP2A6   5.407028</span></span>
<span><span class="co">## 1096921        1 0.9641649  104.43211     FUT2 -11.927107</span></span>
<span><span class="co">## 1088486        1 0.9602264   24.59603     KDSR  -4.526287</span></span>
<span><span class="co">## 977014         1 0.9515004   86.82732   NPC1L1 -10.761931</span></span>
<span><span class="co">## 977015         2 0.9495892   59.83144    DDX56   9.641861</span></span>
<span><span class="co">## 992988         0 0.9450026   23.14602   TTC39B  -4.334495</span></span>
<span><span class="co">## 987201         2 0.9405116   32.60063     SRRT   5.424996</span></span>
<span><span class="co">## 1005378        1 0.9380390   47.46054     PKN3  -6.620563</span></span>
<span><span class="co">## 1010141        1 0.9371487   37.07672 C10orf88  -6.787850</span></span>
<span><span class="co">## 938455         1 0.9363014   70.56030     PELO   8.288398</span></span>
<span><span class="co">## 1084624        1 0.9336311   30.56463   STAT5B   5.426252</span></span>
<span><span class="co">## 928668         0 0.9320372   25.78698   ACVR1C  -4.687370</span></span>
<span><span class="co">## 867597         1 0.8941255  252.81074     USP1  16.258211</span></span>
<span><span class="co">## 852390         0 0.8393363   22.17420  KLHDC7A   4.124187</span></span>
<span><span class="co">## 1017410        0 0.8274447   21.52123  CRACR2B  -3.989585</span></span>
<span><span class="co">## 1021453        1 0.8251281   33.41511  SPTY2D1  -5.557123</span></span>
<span><span class="co">## 987196         3 0.8234854   40.37437     POP7  -5.845258</span></span>
<span><span class="co">## 861235         0 0.8163154   22.15199    SYTL1  -3.962854</span></span>
<span><span class="co">## 901554         1 0.8133753   28.04053     ALLC   4.919066</span></span>
<span><span class="co">## 1042437        0 0.8041948   22.63603    CCND2  -4.065830</span></span></code></pre>
<p>We will also visualize the POLK locus that we highlighted in the paper. As mentioned previously, the cTWAS output includes results for all genes, but it only includes a fraction of variants if the <code>thin</code> parameter is specified. There are complete variant results in regions with gene PIP greater than <code>rerun_gene_PIP</code>, but there are incomplete variant results in regions with weaker gene signals. The POLK locus is a “null” result; after running cTWAS, there is no strong gene signal. Thus, in order to plot the full set of variants at this locus, we must specify <code>rerun_ctwas=T</code> to run cTWAS again at this locus using all variants. This calls <code>ctwas_rss</code> and logs its progress; it also create a temporary folder within the output directory to store the results. After doing this the first time, you can set <code>rerun_load_only=T</code> to make changes to the plot without calling cTWAS again.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># genome-wide bonferroni threshold used for TWAS</span></span>
<span><span class="va">twas_sig_thresh</span> <span class="op">&lt;-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show cTWAS result for POLK and store region_tag</span></span>
<span><span class="va">ctwas_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">genename</span><span class="op">==</span><span class="st">"POLK"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">region_tag</span> <span class="op">&lt;-</span> <span class="st">"5_45"</span></span>
<span></span>
<span><span class="co"># plot the POLK locus</span></span>
<span><span class="fu"><a href="../reference/ctwas_locus_plot.html">ctwas_locus_plot</a></span><span class="op">(</span>ctwas_res <span class="op">=</span> <span class="va">ctwas_res</span>,</span>
<span>                 region_tag <span class="op">=</span> <span class="va">region_tag</span>,</span>
<span>                 xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>,<span class="fl">75.8</span><span class="op">)</span>,</span>
<span>                 twas_sig_thresh <span class="op">=</span> <span class="va">twas_sig_thresh</span>,</span>
<span>                 alt_names <span class="op">=</span> <span class="st">"genename"</span>,</span>
<span>                 legend_panel <span class="op">=</span> <span class="st">"cTWAS"</span>,</span>
<span>                 legend_side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>                 outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                 outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                 rerun_ctwas <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                 z_snp <span class="op">=</span> <span class="va">z_snp</span>,</span>
<span>                 z_gene <span class="op">=</span> <span class="va">z_gene</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="cleaning-up">Cleaning up<a class="anchor" aria-label="anchor" href="#cleaning-up"></a>
</h2>
<p>The summary statistics version of cTWAS generates a lot of files. cTWAS added variant-by-gene and gene-by-gene correlation information (“R matrix info”) for all regions in a directory called <code>[outname]_LDR</code>, and it also stored thinned version of the LD matrices in the same directory. We store all these files to speed computation, and we leave them unzipped at the end of the analysis so that they can be used with <code>ctwas_locus_plot</code>. However, once you are done with an analysis, we recommend zipping the contents of the <code>[outname]_LDR</code> folder.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"tar -zcvf "</span>, <span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">"_LDR.tar.gz "</span>, <span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">"_LDR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rm -r "</span>, <span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">"_LDR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>Here are some details about the computing environment, including the versions of R, and the R packages, used to generate these results.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: CentOS Linux 7 (Core)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] cowplot_1.1.1 ggplot2_3.5.1 RSQLite_2.3.7 ctwas_0.1.38 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.1  xfun_0.45         bslib_0.7.0       lattice_0.22-6   </span></span>
<span><span class="co">##  [5] pgenlibr_0.3.7    generics_0.1.3    colorspace_2.1-0  vctrs_0.6.5      </span></span>
<span><span class="co">##  [9] htmltools_0.5.8.1 yaml_2.3.9        utf8_1.2.4        blob_1.2.4       </span></span>
<span><span class="co">## [13] rlang_1.1.4       pkgdown_2.1.0     jquerylib_0.1.4   pillar_1.9.0     </span></span>
<span><span class="co">## [17] withr_3.0.0       glue_1.7.0        DBI_1.2.3         bit64_4.0.5      </span></span>
<span><span class="co">## [21] foreach_1.5.2     lifecycle_1.0.4   munsell_0.5.1     gtable_0.3.0     </span></span>
<span><span class="co">## [25] ragg_1.3.2        htmlwidgets_1.6.4 codetools_0.2-20  evaluate_0.24.0  </span></span>
<span><span class="co">## [29] memoise_2.0.1     labeling_0.4.3    knitr_1.48        tzdb_0.4.0       </span></span>
<span><span class="co">## [33] fastmap_1.2.0     fansi_1.0.6       highr_0.11        logging_0.10-108 </span></span>
<span><span class="co">## [37] Rcpp_1.0.12       readr_2.1.5       scales_1.3.0      cachem_1.1.0     </span></span>
<span><span class="co">## [41] desc_1.4.3        jsonlite_1.8.8    farver_2.1.2      systemfonts_1.1.0</span></span>
<span><span class="co">## [45] fs_1.6.4          bit_4.0.5         textshaping_0.4.0 hms_1.1.3        </span></span>
<span><span class="co">## [49] digest_0.6.36     dplyr_1.1.4       grid_4.2.0        cli_3.6.3        </span></span>
<span><span class="co">## [53] tools_4.2.0       magrittr_2.0.3    sass_0.4.9        tibble_3.2.1     </span></span>
<span><span class="co">## [57] pkgconfig_2.0.3   Matrix_1.4-1      data.table_1.15.4 rmarkdown_2.27   </span></span>
<span><span class="co">## [61] iterators_1.0.14  R6_2.5.1          compiler_4.2.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
