<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A minimal tutorial of running cTWAS without LD • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A minimal tutorial of running cTWAS without LD">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/minimal_ctwas_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/single_group" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="minimal_ctwas_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A minimal tutorial of running cTWAS without LD</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-07-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/single_group/vignettes/minimal_ctwas_tutorial.Rmd" class="external-link"><code>vignettes/minimal_ctwas_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>minimal_ctwas_tutorial.Rmd</code></div>

    </div>

    
    
<p>This document demonstrates how to run a simple analysis using cTWAS with GWAS summary statistics.</p>
<p>One complication of using cTWAS is the use of LD matrices. The LD data are often large, creating challenges in data storage and sharing, and in running the analysis (e.g. not all data can be fit into memory). It is possible to run the entire cTWAS analysis without using LD data if one assumes at most one causal signal per LD region. The downside is that cTWAS may miss causal genes in some regions. Nevertheless, this may be a worthy trade-off, especially in the first pass. This tutorial shows how to perform a simpler analysis without LD.</p>
<p>When running cTWAS in this way, it is similar to colocalization analysis. In the Supplementary Notes of the cTWAS paper, we explained that in the situation where a region contains a single gene with a single eQTL, cTWAS is similar to coloc, and equivalent to Enloc, a version of colocalization analysis. Still, compared to Enloc and coloc, running cTWAS without LD has several benefits: the important parameters of colocalization are estimated from data; cTWAS jointly analyzes multiple genes; and each gene is allowed to have multiple QTL variants.</p>
<p>First, install <code>ctwas</code> if you haven’t already:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>,ref <span class="op">=</span> <span class="st">"single_group"</span><span class="op">)</span></span></code></pre></div>
<p>Currently, <code>ctwas</code> has only been tested on Linux systems.</p>
<p>We recommend running <code>ctwas</code> on a High-Performance Computing system.</p>
<p>Now load the <code>ctwas</code> package and a few other packages needed to perform the analysis:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">ctwas</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="preparing-the-input-data">Preparing the input data<a class="anchor" aria-label="anchor" href="#preparing-the-input-data"></a>
</h2>
<p>The “no-LD” version of cTWAS requires several input datasets: (i) GWAS summary statistics; (ii) prediction models (called “weights” in our documentation) and the reference data, including information about all the variants; and (iii) the definition of the genomic regions. We describe below the steps for data preparation and preprocessing before running cTWAS. More details about data preparation, including preparing the LD data, which is required for full cTWAS, are given in the tutorial, “Preparing cTWAS input data”.</p>
<p>For GWAS summary statistics, we use sample GWAS data of LDL on chromosome 16. These data are provided in the R package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.z_snp.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We also need the GWAS sample size for summarizing the cTWAS result.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span></code></pre></div>
<div class="section level3">
<h3 id="reference-data">Reference data<a class="anchor" aria-label="anchor" href="#reference-data"></a>
</h3>
<p>cTWAS assumes that the genome is partitioned into approximately independent LD regions. The definitions of these regions need to be provided. You will also need to have a list of SNPs, <code>snp_info</code>, as a “reference”, with the information of these SNPs. This is similar to LD reference data, needed in fine-mapping GWAS loci with summary statistics. Also similar to fine-mapping, the variant reference should match in the population of GWAS samples.</p>
<p>It is critical that the genome build (e.g. hg38) of the reference match the genome build used to train the prediction models. By contrast, the genome build of the GWAS summary statistics does not matter because variant positions are determined by the reference.</p>
<p>We require a data frame <code>region_info</code> containing the region definitions, with the following columns: “chrom”, “start”, “stop”, for the genomic coordinates of the regions, and “region_id” for the IDs of the regions (by default, we use the convention <chrom_start_stop> for the region IDs).</chrom_start_stop></p>
<p>Here we use the b38 European LDetect blocks, which is again included in the R package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ldetect"</span>, <span class="st">"EUR.b38.ldetect.regions.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">region_file</span><span class="op">)</span></span></code></pre></div>
<p>Additionally, we require <code>ref_snp_info</code> as a “reference” of the variants. <code>ref_snp_info</code> is a list, and each element of the list contains a data frame with the positions and allele information of the variants in a region. We provide a function, <code><a href="../reference/preprocess_region_LD_snp_info.html">preprocess_region_LD_snp_info()</a></code> to preprocess region definitions and map variants from the reference to regions, which takes input of <code>region_info</code> and a data frame <code>ref_snp_info</code> with variant information from the entire genome, and returns processed <code>region_info</code> and <code>snp_info</code>.</p>
<p>We have the lists of reference variant information from all the LD regions in the genome in <a href="https://uchicago.box.com/s/t089or92dkovv0epkrjvxq8r9db9ys99" class="external-link">hg38</a> and <a href="https://uchicago.box.com/s/ufko2gjagcb693dob4khccqubuztb9pz" class="external-link">hg19</a>. They are also available on the University of Chicago RCC cluster at <code>/project2/xinhe/shared_data/multigroup_ctwas/LD_reference/</code>.</p>
<p>In this tutorial, we use chr16 as an example, so we specify <code>chrom = 16</code>. In real cTWAS analysis, you should run the entire genome.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_chrom</span> <span class="op">&lt;-</span> <span class="fl">16</span></span>
<span><span class="va">ref_snp_info_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"ukb_b38_0.1_chr16_var_info.Rvar.gz"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">ref_snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">ref_snp_info_file</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">ref_snp_info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"data.frame"</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_region_LD_snp_info.html">preprocess_region_LD_snp_info</a></span><span class="op">(</span><span class="va">region_info</span>, </span>
<span>                                     <span class="va">ref_snp_info</span>, </span>
<span>                                     chrom <span class="op">=</span> <span class="va">example_chrom</span>, </span>
<span>                                     use_LD <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">region_info</span></span>
<span><span class="va">snp_info</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">snp_info</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chrom   start    stop          region_id</span></span>
<span><span class="co">## 1    16   10054 1157206   16_10054_1157206</span></span>
<span><span class="co">## 2    16 1157206 2714828 16_1157206_2714828</span></span>
<span><span class="co">## 3    16 2714828 3951195 16_2714828_3951195</span></span>
<span><span class="co">## 4    16 3951195 5068344 16_3951195_5068344</span></span>
<span><span class="co">## 5    16 5068344 5841579 16_5068344_5841579</span></span>
<span><span class="co">## 6    16 5841579 6843550 16_5841579_6843550</span></span></code></pre>
<p>Note: in this simple tutorial, we use the “no-LD” version, so we set <code>use_LD = FALSE</code> in the call to preprocess_region_LD_snp_info() above.</p>
</div>
<div class="section level3">
<h3 id="harmonizing-gwas-summary-statistics">Harmonizing GWAS summary statistics<a class="anchor" aria-label="anchor" href="#harmonizing-gwas-summary-statistics"></a>
</h3>
<p>The GWAS summary statistics needs to be “harmonized” before use. This means, for example, the allele definitions need to be consistent with the reference data. Harmonization is achieved through the <code><a href="../reference/preprocess_z_snp.html">preprocess_z_snp()</a></code> function. We filter out multiallelic variants and strand-ambiguous variants by default.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_z_snp.html">preprocess_z_snp</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">snp_info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-11 11:36:23 INFO::Preprocessing z_snp...</span></span>
<span><span class="co">## 2024-07-11 11:36:23 INFO::z_snp has 407828 variants in total</span></span>
<span><span class="co">## 2024-07-11 11:36:23 INFO::267361 variants left after filtering by snp_info</span></span>
<span><span class="co">## 2024-07-11 11:36:23 INFO::Drop 30 multiallelic variants</span></span>
<span><span class="co">## 2024-07-11 11:36:23 INFO::Harmonize 267331 variants in both GWAS and LD reference</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Flip signs for 267289 (99.98%) variants</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Remove 45738 strand ambiguous variants</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::221593 variants left after preprocessing and harmonization.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="prediction-models-of-genes">Prediction models of genes<a class="anchor" aria-label="anchor" href="#prediction-models-of-genes"></a>
</h3>
<p>We need a list <code>weights</code> which contains preprocessed weights from PredictDB or FUSION prediction models. We provide a function <code>preprocess_weight()</code> to harmonize the prediction models and LD reference.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"expression_Liver.db"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_weights.html">preprocess_weights</a></span><span class="op">(</span><span class="va">weight_file</span>,</span>
<span>                              <span class="va">region_info</span>,</span>
<span>                              <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span>,</span>
<span>                              <span class="va">snp_info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-11 11:36:24 INFO::Load weight: /tmp/jobs/40832708/RtmpirD79E/temp_libpath502a5fe3e2a8/ctwas/extdata/sample_data/expression_Liver.db</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Loading PredictDB weights ...</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Keep protein coding genes only</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::weight_name: expression_Liver</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Remove 1 genes without predictdb LD</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Number of genes with weights: 11479</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::Number of variants in weights: 16929</span></span>
<span><span class="co">## 2024-07-11 11:36:24 INFO::529 genes and 410 variants left after filtering by GWAS and LD reference</span></span>
<span><span class="co">## 2024-07-11 11:36:25 INFO::Harmonizing and processing weights ...</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="running-the-ctwas-analysis">Running the cTWAS analysis<a class="anchor" aria-label="anchor" href="#running-the-ctwas-analysis"></a>
</h2>
<p>We use the function <code><a href="../reference/ctwas_sumstats_noLD.html">ctwas_sumstats_noLD()</a></code> to run cTWAS analysis without LD. It takes as input preprocessed GWAS z-scores (<code>z_snp</code>), <code>weights</code>, <code>region_info</code>, <code>snp_info</code>, and will perform the main steps: (i) computing Z-scores of genes; (i) estimating model parameters; (iii) screening regions potentially containing causal genes, and (iv) fine-mapping those regions.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_sumstats_noLD.html">ctwas_sumstats_noLD</a></span><span class="op">(</span><span class="va">z_snp</span>,</span>
<span>                                 <span class="va">weights</span>,</span>
<span>                                 <span class="va">region_info</span>,</span>
<span>                                 <span class="va">snp_info</span><span class="op">)</span></span></code></pre></div>
<p>Note that if the Z-scores of the genes, <code>z_gene</code>, have already been computed when preparing cTWAS input data, we could specify the <code>z_gene</code> argument in the function, then it will skip computing z-scores of genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">screened_region_data</span></span></code></pre></div>
<p>ctwas_sumstats_noLD() returns several things, including:</p>
<ul>
<li><p><code>param</code>: the estimated parameters,</p></li>
<li><p><code>finemap_res</code>: fine-mapping results,</p></li>
<li><p><code>boundary_genes</code>: the genes that cross region boundaries (we will need these in the post-processing step),</p></li>
<li><p><code>z_gene</code>: the Z-scores of genes,</p></li>
<li><p><code>region_data</code>: assembled region data for all the regions,</p></li>
<li><p><code>screened_region_data</code>: region data with full SNPs for the screened regions.</p></li>
</ul>
<p><em>Note:</em>, we used sample data (from chr16) in this example, however, in real data analysis, we should use data from the <em>entire genome</em> to estimate parameters.</p>
</div>
<div class="section level2">
<h2 id="summarizing-and-visualizing-the-ctwas-results">Summarizing and visualizing the cTWAS results<a class="anchor" aria-label="anchor" href="#summarizing-and-visualizing-the-ctwas-results"></a>
</h2>
<div class="section level3">
<h3 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h3>
<p>First, we use <code><a href="../reference/summarize_param.html">summarize_param()</a></code> to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes. These PVE values allow us to compute how heritability of the phenotype is partitioned among variants and groups of genes.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<p>Make plots of how estimated parameters converge during the execution of the program:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_convergence_plots.html">make_convergence_plots</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="minimal_ctwas_tutorial_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="840"><p class="caption">
 
</p>
</div>
<p>These plots show the estimated prior inclusion probability, and prior effect size, enrichment and PVE over the iterations of parameter estimation. In particular, genes have higher prior inclusion probability, enrichment and PVE than variants.</p>
</div>
<div class="section level3">
<h3 id="interpreting-ctwas-results">Interpreting cTWAS results<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-results"></a>
</h3>
<p>To interpret and visualize the results, we first need to add gene annotations (gene names and gene types) to the fine-mapping result.</p>
<p>We need a user defined data frame <code>gene_annot</code> that should contain the following columns: ‘chrom’, ‘start’, ‘end’, ‘gene_id’, ‘gene_name’ and ‘gene_type’ for each gene.</p>
<p>If you only used gene expression data, you can use the following function to obtain the gene annotations from the Ensembl database. We use <code>EnsDb.Hsapiens.v86</code> for the example data, please choose the Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">id</span>, split <span class="op">=</span> <span class="st">"[|]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db</a></span><span class="op">(</span><span class="va">ens_db</span>, <span class="va">gene_ids</span><span class="op">)</span></span></code></pre></div>
<p>We then use the <code><a href="../reference/anno_finemap_res.html">anno_finemap_res()</a></code> function to add gene annotations to the fine-mapping result, and use gene mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">snp_info</span>, <span class="va">gene_annot</span>,</span>
<span>                                use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-11 11:36:28 INFO::Annotating ctwas finemapping result ...</span></span>
<span><span class="co">## 2024-07-11 11:36:28 INFO::add gene_name and gene_type</span></span>
<span><span class="co">## 2024-07-11 11:36:28 INFO::use gene mid positions</span></span>
<span><span class="co">## 2024-07-11 11:36:28 INFO::add SNP positions from snp_info</span></span></code></pre>
<p>We list genes with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"gene"</span> <span class="op">&amp;</span> <span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    chrom      pos           gene_id gene_name      gene_type</span></span>
<span><span class="co">## 10    16 72070235 ENSG00000261701.6       HPR protein_coding</span></span>
<span><span class="co">##                                    id group         z            region_id</span></span>
<span><span class="co">## 10 ENSG00000261701.6|expression_Liver  gene -18.40315 16_71020125_72901251</span></span>
<span><span class="co">##    susie_pip      mu2</span></span>
<span><span class="co">## 10         1 336.4736</span></span></code></pre>
<p>In this example, HPR gene is the only gene that satisfies this threshold.</p>
</div>
<div class="section level3">
<h3 id="visualizing-the-ctwas-results">Visualizing the cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-the-ctwas-results"></a>
</h3>
<p>We could make locus plots for regions of interest, and zoom in the region by specifying the <code>locus_range</code>. For example:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-11 11:36:29 INFO::focus gene: HPR</span></span>
<span><span class="co">## 2024-07-11 11:36:29 INFO::focus id: ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 2024-07-11 11:36:29 INFO::plot locus range: chr16 71600000,72400000</span></span>
<span><span class="co">## 2024-07-11 11:36:29 INFO::gene: HPR</span></span>
<span><span class="co">## 2024-07-11 11:36:29 INFO::QTL positions: 72063820,72063928</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="minimal_ctwas_tutorial_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
<p>The locus plot shows several tracks in the example region in chromosome 6, including GWAS -log10(p-value), PIPs, QTLs of the focus gene (by default, the focus gene is the gene with the highest PIP), and the gene track. In this example, we zoomed in a locus ( 71.6 - 72.4 Mb) around the HPR gene, which has the highest PIP (PIP = 1), and the red dotted line shows the PIP cutoff of 0.8.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
