<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Running cTWAS analysis • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Running cTWAS analysis"><meta property="og:description" content="ctwas"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>




<script src="running_ctwas_analysis_files/accessible-code-block-0.0.1/empty-anchor.js"></script>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Running cTWAS analysis</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2025-05-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/vignettes/running_ctwas_analysis.Rmd" class="external-link"><code>vignettes/running_ctwas_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>running_ctwas_analysis.Rmd</code></div>

    </div>

    
    
<p>Running cTWAS analysis involves several major steps: preparing input data, computing associations of molecular traits with the phenotype, estimating parameters, screening for candidate genomic regions, and fine-mapping these regions.</p>
<p>In this version of cTWAS, we allow the joint analysis of multiple groups of molecular traits. This could be: eQTL of multiple tissues; or eQTLs, splicing QTLs and other types of QTLs in a single tissue. In a more complex setting, multiple types of QTL data from multiple tissues/cell types. Each group is defined by its “type” (kind of molecular traits), and “context” (tissue, cell type, condition, etc.). If you have multiple datasets from the same type and context, and you prefer to treat them as separate groups, then you can give them different context names, e.g. “context” = “GTEx_brain” and “CMC_brain”.</p>
<p>We require first running the data preprocessing and harmonization steps. Please follow the <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/preparing_input_data.html">“Preparing cTWAS input data”</a> tutorial.</p>
<p>Load the packages</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(ctwas)</span></code></pre></div>
<p>Let’s first load the preprocessed input data,including GWAS summary statistics (<code>z_snp</code>), the processed weights of molecular traits (<code>weights</code>), the region definitions (<code>region_info</code>), and the list of SNPs and information of LD matrix of each region (<code>snp_map</code> and <code>LD_map</code>):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>z_snp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/sample_data&quot;</span>, <span class="st">&quot;LDL_example.preprocessed.z_snp.RDS&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ctwas&quot;</span>))</span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a>weights &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/sample_data&quot;</span>, <span class="st">&quot;LDL_example.preprocessed.weights.RDS&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ctwas&quot;</span>))</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a>region_info &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/sample_data&quot;</span>, <span class="st">&quot;LDL_example.region_info.RDS&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ctwas&quot;</span>))</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a>snp_map &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/sample_data&quot;</span>, <span class="st">&quot;LDL_example.snp_map.RDS&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ctwas&quot;</span>))</span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a>LD_map &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/sample_data&quot;</span>, <span class="st">&quot;LDL_example.LD_map.RDS&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ctwas&quot;</span>))</span></code></pre></div>
<p>We have two ways of running cTWAS main analysis: running the main function, or running separate modules. The first is easier to run, and the second has the advantage of giving you more control of the analysis (see below).</p>
<div id="running-ctwas-main-function" class="section level2">
<h2>Running cTWAS main function</h2>
<p>In this section, we will show how to perform full cTWAS analysis using the main function.</p>
<p>We use the function <code>ctwas_sumstats()</code> to run cTWAS analysis with LD. It takes as input preprocessed GWAS z-scores (<code>z_snp</code>), and prediction models of molecular traits (<code>weights</code>). In addition, it requires the reference data including information about regions (<code>region_info</code>), reference variant list (<code>snp_map</code>) and reference LD (<code>LD_map</code>).</p>
<p>The function performs the main steps: (i) computing Z-scores of molecular traits; (i) estimating model parameters; (iii) screening regions potentially containing causal molecular traits, and (iv) fine-mapping those regions to discover causal molecular traits.</p>
<p>Now, we can run cTWAS using the <code>ctwas_sumstats()</code> function:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>ctwas_res &lt;-<span class="st"> </span><span class="kw">ctwas_sumstats</span>(z_snp, </span>
<span id="cb3-2"><a href="#cb3-2"></a>                            weights, </span>
<span id="cb3-3"><a href="#cb3-3"></a>                            region_info, </span>
<span id="cb3-4"><a href="#cb3-4"></a>                            LD_map, </span>
<span id="cb3-5"><a href="#cb3-5"></a>                            snp_map, </span>
<span id="cb3-6"><a href="#cb3-6"></a>                            <span class="dt">thin =</span> <span class="dv">1</span>,</span>
<span id="cb3-7"><a href="#cb3-7"></a>                            <span class="dt">maxSNP =</span> <span class="dv">20000</span>,</span>
<span id="cb3-8"><a href="#cb3-8"></a>                            <span class="dt">min_group_size =</span> <span class="dv">100</span>, </span>
<span id="cb3-9"><a href="#cb3-9"></a>                            <span class="dt">group_prior_var_structure =</span> <span class="st">&quot;shared_all&quot;</span>, </span>
<span id="cb3-10"><a href="#cb3-10"></a>                            <span class="dt">min_nonSNP_PIP =</span> <span class="fl">0.5</span>,</span>
<span id="cb3-11"><a href="#cb3-11"></a>                            <span class="dt">min_abs_corr =</span> <span class="fl">0.1</span>, </span>
<span id="cb3-12"><a href="#cb3-12"></a>                            <span class="dt">ncore =</span> <span class="dv">6</span>, </span>
<span id="cb3-13"><a href="#cb3-13"></a>                            <span class="dt">ncore_LD =</span> <span class="dv">4</span>,</span>
<span id="cb3-14"><a href="#cb3-14"></a>                            <span class="dt">save_cor =</span> <span class="ot">TRUE</span>,</span>
<span id="cb3-15"><a href="#cb3-15"></a>                            <span class="dt">cor_dir =</span> <span class="st">&quot;./cor_matrix&quot;</span>,</span>
<span id="cb3-16"><a href="#cb3-16"></a>                            <span class="dt">force_compute_cor =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Important arguments that control how this function is executed:</p>
<ul>
<li><p><code>z_gene</code>: computed Z-scores of the molecular traits. If not provided, it will compute those using <code>z_snp</code> and <code>weights</code>.</p></li>
<li><p><code>thin</code>: the fraction of SNPs used to estimate parameters and screen regions likely containing signals. By sampling a fraction of SNPs, this speeds up computation. By default, we use <code>thin = 1</code>, i.e. all SNPs. You can speed up computation by sampling a fraction of SNPs, e.g. <code>thin=0.1</code> will use 10% of SNPs.</p></li>
<li><p><code>maxSNP</code>: maximum number of SNPs in a region (e.g. 20000). Default is Inf, no limit. This can be useful if there are too many SNPs in a region and you don’t have enough memory to run the program.</p></li>
<li><p><code>min_group_size</code>: minimum number of variables in a region (default is 100). Groups with fewer variables will be removed from cTWAS analysis, because parameters cannot be reliably estimated for groups with too few variables.</p></li>
<li><p><code>group_prior_var_stucture</code>: options to control how the effect size variance parameters are shared among groups. For example, “shared_all” (default option) allows all groups to share the same variance parameter. “independent” allows all groups to have their own separate variance parameters. More options can be found in the “Estimating parameters” section below.</p></li>
<li><p><code>filter_L</code>: if <code>filter_L = TRUE</code> (the default option), cTWAS will first estimate the number of causal signals (<code>L</code>) for each region. This step is done using uniform prior, treating all variants and genes equally. From the results, we obtain the number of credible sets (<code>L</code>). We then select regions with at least one credible set, L &gt; 0.</p></li>
<li><p><code>filter_nonSNP_PIP</code> and <code>min_nonSNP_PIP</code>: if <code>filter_nonSNP_PIP = TRUE</code>, it will compute the non-SNP PIP, i.e. the total PIP of all molecular traits, for each region (using the estimated parameters) and select regions with non-SNP PIP &gt; <code>min_nonSNP_PIP</code> (0.5 by default).</p></li>
<li><p><code>min_abs_corr</code>: Minimum absolute correlation allowed in a credible set. cTWAS jointly fine-maps both SNPs and molecular traits, the correlations between SNPs and molecular traits could be lower than standard fine-mapping with SNPs. We use a lower cutoff (0.1 by default).</p></li>
<li><p><code>ncore</code>: computation of genomic regions can be performed in parallel. This parameter controls how many cores are used in the parallelization.</p></li>
<li><p><code>ncore_LD</code>: this sets the number of cores in parallelization when LD matrices are needed. In screening regions and fine-mapping steps, we need larger memory to load and process LD matrices. We recommend using fewer cores in <code>ncore_LD</code> than <code>ncore</code>.</p></li>
<li><p><code>cor_dir</code>, <code>save_cor</code>, and <code>force_compute_cor</code>: we need correlation matrices during fine-mapping and visualizing the results with LD. If <code>cor_dir</code> is provided, cTWAS will check the <code>cor_dir</code> directory. If precomputed correlation matrices are available in <code>cor_dir</code>, cTWAS will simply load those. Otherwise, it will compute the correlation matrices and save to the <code>cor_dir</code> directory if <code>save_cor = TRUE</code>. You can set <code>force_compute_cor = TRUE</code> so that it always computes the correlation matrices. The correlation matrices are large datasets, so they are not saved by default. <em>Note</em>: you should use a different <code>cor_dir</code> directory for each cTWAS analysis; otherwise, it would cause mismatch between Z-scores and correlation matrices in the old <code>cor_dir</code>.</p></li>
</ul>
<p>To understand and summarize the results from cTWAS, please read the tutorial <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/summarizing_results.html">“Summarizing and visualizing cTWAS results”</a>.</p>
<p><em>Note:</em> One can run cTWAS without using reference LD data. See <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/minimal_tutorial.html">“A minimal tutorial of running cTWAS without LD”</a>. That tutorial describes how to run cTWAS in a single step.</p>
</div>
<div id="running-ctwas-modules" class="section level2">
<h2>Running cTWAS modules</h2>
<p>This section describes how to perform cTWAS analysis by running each of the modules separately. This would allow more flexible controls of the cTWAS modules, e.g. running each module with different settings (memory, parallelization, etc.). We could also save the results from each module for future access, that way, we can re-run just one module, if necessary, without running the entire process. One useful way of running cTWAS is to first run the first few steps to estimate model parameters and assess their values. If this step does not converge, or the parameters look strange, it is best to pause and understand why.</p>
<p>We note that one can also run cTWAS without using LD information - see the tutorial <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/minimal_tutorial.html">“A minimal tutorial of running cTWAS without LD”</a>. It is also possible to do this analysis one step at a time. We describe below how each step can be done in appropriate places.</p>
<div id="computing-z-scores-of-molecular-traits" class="section level3">
<h3>Computing Z-scores of molecular traits</h3>
<p>After we have done data preprocessing, we compute Z-scores of molecular traits. This step basically performs transcriptome-wide association studies (TWAS) on each molecular trait with a prediction model. The underlying calculation is based on <a href="https://www.nature.com/articles/s41467-018-03621-1">S-PrediXcan</a>.</p>
<p>The <code>compute_gene_z</code> function computes Z-scores for the molecular traits using preprocessed SNP Z-scores (<code>z_snp</code>) and the preprocessed weights (<code>weights</code>). <code>ncore</code> specifies the number of cores to use when parallelizing over genes.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>z_gene &lt;-<span class="st"> </span><span class="kw">compute_gene_z</span>(z_snp, weights, <span class="dt">ncore =</span> <span class="dv">6</span>)</span></code></pre></div>
<p>We recommend filtering out groups with too few variables (molecular traits), as parameters cannot be reliably estimated for groups with too few variables.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>z_gene &lt;-<span class="st"> </span><span class="kw">filter_z_gene_by_group_size</span>(z_gene, <span class="dt">min_group_size =</span> <span class="dv">100</span>)</span></code></pre></div>
</div>
<div id="assemble-input-data-for-the-regions" class="section level3">
<h3>Assemble input data for the regions</h3>
<p>Because cTWAS runs region-by-region, we will first need to prepare the data for each region, combining information of all variants and molecular traits in that region. After data preprocessing and computation of Z-scores of molecular traits, we assemble the input data for all the regions using the function <code>assemble_region_data()</code>. It assigns molecular traits, variants and their Z-scores to each region. The assignment of a molecular trait to regions is based on the weights, i.e. variants in the prediction model, of the molecular trait.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>res &lt;-<span class="st"> </span><span class="kw">assemble_region_data</span>(region_info, </span>
<span id="cb6-2"><a href="#cb6-2"></a>                            z_snp, </span>
<span id="cb6-3"><a href="#cb6-3"></a>                            z_gene, </span>
<span id="cb6-4"><a href="#cb6-4"></a>                            weights, </span>
<span id="cb6-5"><a href="#cb6-5"></a>                            snp_map,</span>
<span id="cb6-6"><a href="#cb6-6"></a>                            <span class="dt">thin =</span> <span class="dv">1</span>, </span>
<span id="cb6-7"><a href="#cb6-7"></a>                            <span class="dt">maxSNP =</span> <span class="dv">20000</span>,</span>
<span id="cb6-8"><a href="#cb6-8"></a>                            <span class="dt">min_group_size =</span> <span class="dv">100</span>, </span>
<span id="cb6-9"><a href="#cb6-9"></a>                            <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb6-10"><a href="#cb6-10"></a>region_data &lt;-<span class="st"> </span>res<span class="op">$</span>region_data</span>
<span id="cb6-11"><a href="#cb6-11"></a>boundary_genes &lt;-<span class="st"> </span>res<span class="op">$</span>boundary_genes</span></code></pre></div>
<p>The function has a few arguments to control its behavior. The <code>thin</code> argument randomly selects a subset of variants (10% when thin = 0.1) to use during parameter estimation for candidate regions. This helps reduce computation. When <code>thin = 1</code> (default option), it will use all the SNPs. <code>maxSNP</code> sets a maximum on the number of variants that can be in a single region to prevent memory issues during fine-mapping. If the number of SNPs in a region is more than <code>maxSNP</code>, it will, by default, choose the top <code>maxSNP</code> SNPs with the highest Z-scores. <code>min_group_size</code> sets the minimum number of variables in a group. Groups with fewer variables will be removed from the analysis, because cTWAS cannot reliably estimate parameters for groups with too few variables. <code>ncore</code> specifies the number of cores to use when parallelizing over regions.</p>
<p>This function returns <code>region_data</code>, a list object containing input data (IDs of molecular traits and variants, their Z-scores, genomic boundaries, etc.) for each of the regions. It also returns a data frame <code>boundary_genes</code>, containing the information of molecular traits whose weights cross region boundaries. The information would be used in later analysis and post-processing.</p>
</div>
<div id="estimating-parameters" class="section level3">
<h3>Estimating parameters</h3>
<p>We use the <code>est_param()</code> function to estimate two sets of parameters: the prior inclusion probabilities and the prior effect size variance for molecular traits and variants, separately.</p>
<p>As described in the Introduction, cTWAS can analyze multiple groups of molecular traits. This step will estimate the prior parameters for each group. But we allow sharing of the prior effect variance parameters among groups. This may improve the accuracy of parameter estimation.There are several options to control how parameters are shared among groups. This can be done by specifying the <code>group_prior_var_stucture</code>:</p>
<ul>
<li>“shared_all” (default option) allows all groups, including SNPs, to share the same variance parameter.</li>
<li>“shared_type” allows all groups in one molecular QTL type to share the same variance parameter.</li>
<li>“shared_context” allows all groups in one context (tissue, cell type, condition) to share the same variance parameter.</li>
<li>“shared_nonSNP” allows all non-SNP groups (i.e. all molecular traits) to share the same variance parameter.</li>
<li>“independent” allows all groups to have their own separate variance parameters.</li>
</ul>
<p>This step will run the EM algorithm to estimate parameters and return the estimated parameters (<code>group_prior</code>, <code>group_prior_var</code>, etc.). For technical reasons (see Methods of the cTWAS paper), it will run two rounds of EM algorithm. In the first round, it uses fewer iterations (<code>niter_prefit=3</code> by default) to get rough parameter estimates, and using these values, select regions for the analysis in the second run of EM. We use more iterations in the second round (maximum <code>niter=50</code> by default) to get accurate parameter estimates. <code>min_group_size</code> sets the minimum number of variables in a group, as parameters cannot be reliably estimated for groups with too few variables.</p>
<p>It also computes enrichment (log scale), and its standard error. We provided two options to assess the statistical significance of enrichment: “G-test” (<code>enrichment_test = "G"</code>, default option), or Fisher’s exact test (<code>enrichment_test = "fisher"</code>).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>param &lt;-<span class="st"> </span><span class="kw">est_param</span>(region_data, </span>
<span id="cb7-2"><a href="#cb7-2"></a>                   <span class="dt">group_prior_var_structure =</span> <span class="st">&quot;shared_all&quot;</span>, </span>
<span id="cb7-3"><a href="#cb7-3"></a>                   <span class="dt">niter_prefit =</span> <span class="dv">3</span>,</span>
<span id="cb7-4"><a href="#cb7-4"></a>                   <span class="dt">niter =</span> <span class="dv">50</span>, </span>
<span id="cb7-5"><a href="#cb7-5"></a>                   <span class="dt">min_group_size =</span> <span class="dv">100</span>,</span>
<span id="cb7-6"><a href="#cb7-6"></a>                   <span class="dt">enrichment_test =</span> <span class="st">&quot;G&quot;</span>,</span>
<span id="cb7-7"><a href="#cb7-7"></a>                   <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb7-8"><a href="#cb7-8"></a>group_prior &lt;-<span class="st"> </span>param<span class="op">$</span>group_prior</span>
<span id="cb7-9"><a href="#cb7-9"></a>group_prior_var &lt;-<span class="st"> </span>param<span class="op">$</span>group_prior_var</span></code></pre></div>
<p><em>Note</em>: we used sample data (from chr16) in this example, however, in real data analysis, parameter estimation should be done using data from the <em>entire genome</em>.</p>
</div>
<div id="screening-regions" class="section level3">
<h3>Screening regions</h3>
<div id="expanding-regions-with-all-snps" class="section level4">
<h4>Expanding regions with all SNPs</h4>
<p>If you had <code>thin</code> &lt; 1 when preparing <code>region_data</code>, you need to expand the regions with all SNPs for screening regions and fine-mapping. If the number of SNPs in a region is more than <code>maxSNP</code>, it will choose the top <code>maxSNP</code> SNPs with the highest Z-scores. This is not needed when <code>thin = 1</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="cf">if</span> (thin <span class="op">&lt;</span><span class="st"> </span><span class="dv">1</span>){</span>
<span id="cb8-2"><a href="#cb8-2"></a>    region_data &lt;-<span class="st"> </span><span class="kw">expand_region_data</span>(region_data,</span>
<span id="cb8-3"><a href="#cb8-3"></a>                                      snp_map,</span>
<span id="cb8-4"><a href="#cb8-4"></a>                                      z_snp,</span>
<span id="cb8-5"><a href="#cb8-5"></a>                                      <span class="dt">maxSNP =</span> <span class="dv">20000</span>,</span>
<span id="cb8-6"><a href="#cb8-6"></a>                                      <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb8-7"><a href="#cb8-7"></a>}</span></code></pre></div>
</div>
<div id="screening-regions-1" class="section level4">
<h4>Screening regions</h4>
<p>We use the <code>screen_regions()</code> function to perform a screening process to select regions with likely causal signals in molecular traits. cTWAS first selects regions with minimum p-values (<code>min_pval</code>) &lt; 5e-8 or any SNPs or all molecular traits (from <code>region_data</code>). Then, cTWAS will perform an initial fine-mapping without LD (L= 1) and compute the non-SNP PIP, i.e. total PIP of all molecular traits, for each region (using estimated priors). It then selects regions by applying a <code>min_nonSNP_PIP</code> cutoff (0.5, by default). Only these selected regions would be subject to full fine-mapping analysis in the final step. Thus screening regions can save computational time.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>screen_res &lt;-<span class="st"> </span><span class="kw">screen_regions</span>(region_data,</span>
<span id="cb9-2"><a href="#cb9-2"></a>                             <span class="dt">group_prior =</span> group_prior,</span>
<span id="cb9-3"><a href="#cb9-3"></a>                             <span class="dt">group_prior_var =</span> group_prior_var,</span>
<span id="cb9-4"><a href="#cb9-4"></a>                             <span class="dt">min_nonSNP_PIP =</span> <span class="fl">0.5</span>,</span>
<span id="cb9-5"><a href="#cb9-5"></a>                             <span class="dt">min_pval =</span> <span class="fl">5e-8</span>,</span>
<span id="cb9-6"><a href="#cb9-6"></a>                             <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb9-7"><a href="#cb9-7"></a>screened_region_data &lt;-<span class="st"> </span>screen_res<span class="op">$</span>screened_region_data</span>
<span id="cb9-8"><a href="#cb9-8"></a>screen_summary &lt;-<span class="st"> </span>screen_res<span class="op">$</span>screen_summary</span></code></pre></div>
</div>
</div>
<div id="fine-mapping-regions" class="section level3">
<h3>Fine-mapping regions</h3>
<div id="fine-mapping-regions-with-ld" class="section level4">
<h4>Fine-mapping regions with LD</h4>
<p>We now perform the fine-mapping step for each of the selected regions from screening with the estimated parameters.</p>
<p>The <code>finemap_regions()</code> function performs fine-mapping for the regions in the <code>screened_region_data</code>.</p>
<p>It first computes correlation matrices among all variables included in fine-mapping (SNPs, and molecular traits), then runs fine-mapping with estimated parameters (<code>group_prior</code> and <code>group_prior_var</code>) with <code>L = 5</code> (default option), and returns fine-mapping results for the regions.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>res &lt;-<span class="st"> </span><span class="kw">finemap_regions</span>(screened_region_data,</span>
<span id="cb10-2"><a href="#cb10-2"></a>                       <span class="dt">LD_map =</span> LD_map,</span>
<span id="cb10-3"><a href="#cb10-3"></a>                       <span class="dt">weights =</span> weights,</span>
<span id="cb10-4"><a href="#cb10-4"></a>                       <span class="dt">group_prior =</span> group_prior,</span>
<span id="cb10-5"><a href="#cb10-5"></a>                       <span class="dt">group_prior_var =</span> group_prior_var,</span>
<span id="cb10-6"><a href="#cb10-6"></a>                       <span class="dt">L =</span> <span class="dv">5</span>,</span>
<span id="cb10-7"><a href="#cb10-7"></a>                       <span class="dt">save_cor =</span> <span class="ot">TRUE</span>,</span>
<span id="cb10-8"><a href="#cb10-8"></a>                       <span class="dt">cor_dir =</span> <span class="st">&quot;./cor_matrix&quot;</span>,</span>
<span id="cb10-9"><a href="#cb10-9"></a>                       <span class="dt">ncore =</span> <span class="dv">4</span>)</span>
<span id="cb10-10"><a href="#cb10-10"></a>finemap_res &lt;-<span class="st"> </span>res<span class="op">$</span>finemap_res</span>
<span id="cb10-11"><a href="#cb10-11"></a>susie_alpha_res &lt;-<span class="st"> </span>res<span class="op">$</span>susie_alpha_res</span></code></pre></div>
<p>cTWAS will automatically check <code>cor_dir</code>. If precomputed correlation matrices are available in <code>cor_dir</code>, cTWAS will simply load and reuse those. Otherwise, it will compute the correlation matrices and save to the <code>cor_dir</code> directory if <code>save_cor = TRUE</code>. You can set <code>force_compute_cor = TRUE</code> so that it always computes the correlation matrices. The correlation matrices are large datasets, so they are not saved by default.</p>
<p><em>Note</em>: when fine-mapping with LD, we need larger memory to load LD matrices and compute correlation matrices. So we recommend increasing memory or reducing the number of <code>ncore</code> for these steps.</p>
<p>It is possible to run fine-mapping using uniform prior, treating all variants and molecular traits equally. To do this, we set <code>group_prior = NULL</code> (uniform prior inclusion probabilities) and <code>group_prior_var = NULL</code> (prior variance = 50 as the default in <code>susie_rss</code>).</p>
</div>
<div id="fine-mapping-regions-without-ld" class="section level4">
<h4>Fine-mapping regions without LD</h4>
<p><em>Note</em>: we could use the <code>finemap_regions_noLD()</code> function to run fine-mapping without LD.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>res &lt;-<span class="st"> </span><span class="kw">finemap_regions_noLD</span>(screened_region_data,</span>
<span id="cb11-2"><a href="#cb11-2"></a>                            <span class="dt">group_prior =</span> group_prior,</span>
<span id="cb11-3"><a href="#cb11-3"></a>                            <span class="dt">group_prior_var =</span> group_prior_var,</span>
<span id="cb11-4"><a href="#cb11-4"></a>                            <span class="dt">ncore =</span> <span class="dv">6</span>)</span>
<span id="cb11-5"><a href="#cb11-5"></a>finemap_res &lt;-<span class="st"> </span>res<span class="op">$</span>finemap_res</span>
<span id="cb11-6"><a href="#cb11-6"></a>susie_alpha_res &lt;-<span class="st"> </span>res<span class="op">$</span>susie_alpha_res</span></code></pre></div>
</div>
</div>
<div id="fine-mapping-a-single-region" class="section level3">
<h3>Fine-mapping a single region</h3>
<p>We could also run fine-mapping for a single region (or several regions of interest) with precomputed parameters. This is particularly useful to perform a deep analysis on molecular traits for regions of interest.</p>
<p>Here we show an example for fine-mapping a single region “16_71020125_72901251”.</p>
<p>We select the data for the region of interest from <code>screened_region_data</code>, obtained from the output of the earlier <code>screen_regions()</code> function.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>region_id &lt;-<span class="st"> &quot;16_71020125_72901251&quot;</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>selected_region_data &lt;-<span class="st"> </span>screened_region_data[region_id]</span></code></pre></div>
<p>We could then perform fine-mapping for this region of interest:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>res &lt;-<span class="st"> </span><span class="kw">finemap_regions</span>(selected_region_data,</span>
<span id="cb13-2"><a href="#cb13-2"></a>                       <span class="dt">LD_map =</span> LD_map,</span>
<span id="cb13-3"><a href="#cb13-3"></a>                       <span class="dt">weights =</span> weights,</span>
<span id="cb13-4"><a href="#cb13-4"></a>                       <span class="dt">group_prior =</span> group_prior,</span>
<span id="cb13-5"><a href="#cb13-5"></a>                       <span class="dt">group_prior_var =</span> group_prior_var,</span>
<span id="cb13-6"><a href="#cb13-6"></a>                       <span class="dt">L =</span> <span class="dv">5</span>,</span>
<span id="cb13-7"><a href="#cb13-7"></a>                       <span class="dt">save_cor =</span> <span class="ot">TRUE</span>,</span>
<span id="cb13-8"><a href="#cb13-8"></a>                       <span class="dt">cor_dir =</span> <span class="st">&quot;./cor_matrix&quot;</span>)</span>
<span id="cb13-9"><a href="#cb13-9"></a>finemap_res &lt;-<span class="st"> </span>res<span class="op">$</span>finemap_res</span>
<span id="cb13-10"><a href="#cb13-10"></a>susie_alpha_res &lt;-<span class="st"> </span>res<span class="op">$</span>susie_alpha_res</span></code></pre></div>
<p>We could also perform fine-mapping without LD:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>res &lt;-<span class="st"> </span><span class="kw">finemap_regions_noLD</span>(selected_region_data,</span>
<span id="cb14-2"><a href="#cb14-2"></a>                            <span class="dt">group_prior =</span> group_prior,</span>
<span id="cb14-3"><a href="#cb14-3"></a>                            <span class="dt">group_prior_var =</span> group_prior_var)</span>
<span id="cb14-4"><a href="#cb14-4"></a>finemap_res &lt;-<span class="st"> </span>res<span class="op">$</span>finemap_res</span>
<span id="cb14-5"><a href="#cb14-5"></a>susie_alpha_res &lt;-<span class="st"> </span>res<span class="op">$</span>susie_alpha_res</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>
