<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing cTWAS results â€¢ ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing cTWAS results">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/minimal_ctwas_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/single_group" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="summarizing_ctwas_results_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing cTWAS results</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-07-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/single_group/vignettes/summarizing_ctwas_results.Rmd" class="external-link"><code>vignettes/summarizing_ctwas_results.Rmd</code></a></small>
      <div class="hidden name"><code>summarizing_ctwas_results.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will show how to summarize and visualize cTWAS results.</p>
<p>Load the packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<p>We need the Ensembl database <code>EnsDb.Hsapiens.v86</code> here in this tutorial, please choose the version of Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre></div>
<p>Load the cTWAS results</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span>
<span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.weights.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.region_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.snp_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LD_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.LD_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.ctwas_sumstats_res.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">screened_region_data</span></span>
<span></span>
<span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"cor_matrix"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h3>
<p>We could use the <code><a href="../reference/summarize_param.html">summarize_param()</a></code> function to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by genes and variants.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<p>Estimated prior inclusion probability:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior</span></span></code></pre></div>
<pre><code><span><span class="co">##         gene          SNP </span></span>
<span><span class="co">## 0.0077029549 0.0002339315</span></span></code></pre>
<p>Estimated prior effect size for genes and variants:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span></code></pre></div>
<pre><code><span><span class="co">##       gene        SNP </span></span>
<span><span class="co">## 210.333277   6.576587</span></span></code></pre>
<p>Estimated enrichment of genes over variants:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">enrichment</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 32.92825</span></span></code></pre>
<p>PVE explained by genes and variants:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##         gene          SNP </span></span>
<span><span class="co">## 0.0019331676 0.0008044242</span></span></code></pre>
<p>Total heritability (sum of PVE):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">total_pve</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.002737592</span></span></code></pre>
<p>Attributable heritability to variants in genes:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">attributable_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##      gene       SNP </span></span>
<span><span class="co">## 0.7061563 0.2938437</span></span></code></pre>
<p>Make convergence plots of estimated parameters</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_convergence_plots.html">make_convergence_plots</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="summarizing_ctwas_results_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="840"><p class="caption">
Â 
</p>
</div>
<p>These plots show the estimated prior inclusion probability, and prior effect size, enrichment and PVE over the iterations of parameter estimation. In particular, genes have higher prior inclusion probability, enrichment and PVE than variants.</p>
</div>
<div class="section level2">
<h2 id="summarizing-ctwas-results">Summarizing cTWAS results<a class="anchor" aria-label="anchor" href="#summarizing-ctwas-results"></a>
</h2>
<p>To interpret and visualize the results, we first need to add gene annotations (gene names and gene types) to the fine-mapping result.</p>
<p>We need a user defined data frame <code>gene_annot</code> that should contain the following columns: â€˜chromâ€™, â€˜startâ€™, â€˜endâ€™, â€˜gene_idâ€™, â€˜gene_nameâ€™ and â€˜gene_typeâ€™ for each gene.</p>
<p>If you only used gene expression data, you can use the following function to obtain the gene annotations from the Ensembl database. We use <code>EnsDb.Hsapiens.v86</code> for the example data, please choose the Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">id</span>, split <span class="op">=</span> <span class="st">"[|]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db</a></span><span class="op">(</span><span class="va">ens_db</span>, <span class="va">gene_ids</span><span class="op">)</span></span></code></pre></div>
<p>We then use the <code><a href="../reference/anno_finemap_res.html">anno_finemap_res()</a></code> function to add gene annotations to the finemapping result, and use gene mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">snp_info</span>, <span class="va">gene_annot</span>,</span>
<span>                                use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-11 11:37:22 INFO::Annotating ctwas finemapping result ...</span></span>
<span><span class="co">## 2024-07-11 11:37:23 INFO::add gene_name and gene_type</span></span>
<span><span class="co">## 2024-07-11 11:37:23 INFO::use gene mid positions</span></span>
<span><span class="co">## 2024-07-11 11:37:23 INFO::add SNP positions from snp_info</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="interpreting-ctwas-results">Interpreting cTWAS results<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-results"></a>
</h2>
<p>We list all genes with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"gene"</span> <span class="op">&amp;</span> <span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="va">sig_finemap_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">sig_finemap_res</span><span class="op">$</span><span class="va">susie_pip</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##    chrom      pos           gene_id gene_name      gene_type</span></span>
<span><span class="co">## 57    16 72070235 ENSG00000261701.6       HPR protein_coding</span></span>
<span><span class="co">##                                    id group         z            region_id</span></span>
<span><span class="co">## 57 ENSG00000261701.6|expression_Liver  gene -18.40315 16_71020125_72901251</span></span>
<span><span class="co">##    susie_pip      mu2 cs_index</span></span>
<span><span class="co">## 57         1 632.3152        1</span></span></code></pre>
<p>In this example, HPR is the only gene that satisfies this threshold.</p>
</div>
<div class="section level2">
<h2 id="visualizing-ctwas-results">Visualizing cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-ctwas-results"></a>
</h2>
<p>We could make locus plots for regions of interest, and zoom in the region by specifying the <code>locus_range</code>.</p>
<p>If we want to show the correlations with the focus gene, we need correlation matrices of the region. Because we have computed and saved correlation matrices during fine-mapping step, we could simply load those precomputed correlation matrices.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_region_cor.html">get_region_cor</a></span><span class="op">(</span>region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>, cor_dir <span class="op">=</span> <span class="va">cor_dir</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>, </span>
<span>               R_snp_gene <span class="op">=</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">R_snp_gene</span>,</span>
<span>               R_gene <span class="op">=</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">R_gene</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-11 11:37:25 INFO::focus gene: HPR</span></span>
<span><span class="co">## 2024-07-11 11:37:25 INFO::focus id: ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 2024-07-11 11:37:25 INFO::plot locus range: chr16 71600000,72400000</span></span>
<span><span class="co">## 2024-07-11 11:37:25 INFO::gene: HPR</span></span>
<span><span class="co">## 2024-07-11 11:37:25 INFO::QTL positions: 72063820,72063928</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_ctwas_results_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
Â 
</p>
</div>
<p>The locus plot shows several tracks in the example region in chromosome 6, including GWAS -log10(p-value), PIPs, QTLs of the focus gene (by default, the focus gene is the gene with the highest PIP), and the gene track. In this example, we zoomed in a locus (71.6 - 72.4 Mb) around the HPR gene, which has the highest PIP (PIP = 1), and the red dotted line shows the PIP cutoff of 0.8.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
