<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing and visualizing cTWAS results • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing and visualizing cTWAS results">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/singlegroup" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="summarizing_results_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing and visualizing cTWAS results</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-07-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/singlegroup/vignettes/summarizing_results.Rmd" class="external-link"><code>vignettes/summarizing_results.Rmd</code></a></small>
      <div class="hidden name"><code>summarizing_results.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will show how to summarize and visualize cTWAS results.</p>
<p>Load the packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">ctwas</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre></div>
<p>We need the Ensembl database <code>EnsDb.Hsapiens.v86</code> here in this tutorial, please choose the version of Ensembl database for your specific data.</p>
<p>Let’s first load the cTWAS input data needed to run this tutorial: GWAS sample size (<code>gwas_n</code>), prediction models of genes (<code>weights</code>), the reference data: including information about regions (<code>region_info</code>), reference variant list (<code>snp_map</code>) and reference LD (<code>LD_map</code>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span>
<span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.weights.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.region_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snp_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.snp_map.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LD_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.LD_map.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We then load the cTWAS result. We focus on examples from running cTWAS with LD here, but the steps below also work for running without LD.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.ctwas_sumstats_res.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">screened_region_data</span></span></code></pre></div>
<p><code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> returns several main output:</p>
<ul>
<li><p><code>z_gene</code>: the Z-scores of genes,</p></li>
<li><p><code>param</code>: the estimated parameters,</p></li>
<li><p><code>finemap_res</code>: fine-mapping results, a data frame of genes and variants, including the IDs, Z-scores, PIPs, credible set indices, as well as region IDs.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                    id group          z            region_id</span></span>
<span><span class="co">## 1 ENSG00000103494.12|expression_Liver  gene -0.4550967 16_53348660_55869862</span></span>
<span><span class="co">## 2 ENSG00000177508.11|expression_Liver  gene -0.3912807 16_53348660_55869862</span></span>
<span><span class="co">## 3 ENSG00000087245.12|expression_Liver  gene  0.5636824 16_53348660_55869862</span></span>
<span><span class="co">## 4 ENSG00000087253.12|expression_Liver  gene -1.3879522 16_53348660_55869862</span></span>
<span><span class="co">## 5 ENSG00000103546.18|expression_Liver  gene  0.5507643 16_53348660_55869862</span></span>
<span><span class="co">## 6 ENSG00000198848.12|expression_Liver  gene -0.6042867 16_53348660_55869862</span></span>
<span><span class="co">##      susie_pip      mu2 cs_index</span></span>
<span><span class="co">## 1 4.155209e-07 1.200426        0</span></span>
<span><span class="co">## 2 4.045011e-07 1.146923        0</span></span>
<span><span class="co">## 3 4.390369e-07 1.310006        0</span></span>
<span><span class="co">## 4 9.776133e-07 2.903492        0</span></span>
<span><span class="co">## 5 4.359028e-07 1.295746        0</span></span>
<span><span class="co">## 6 4.495215e-07 1.356983        0</span></span></code></pre>
<p>It also produces other output:</p>
<ul>
<li><p><code>boundary_genes</code>: the genes that cross region boundaries (we will need these in the post-processing step).</p></li>
<li><p><code>region_data</code>: assembled region data, including Z scores of SNPs and genes, for all the regions,. Note that only data of the subset of SNPs used in screening regions are included.</p></li>
<li><p><code>screened_region_data</code>: the data of all SNPs and genes of the selected regions.</p></li>
</ul>
<div class="section level2">
<h2 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h2>
<p>First, we make plots using <code>use</code>make_convergence_plots()` to see how estimated parameters converge during the execution of the program:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_convergence_plots.html">make_convergence_plots</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="summarizing_results_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="840"><p class="caption">
 
</p>
</div>
<p>These plots show the estimated prior inclusion probability, prior effect size variance, and enrichment over the iterations of parameter estimation. The enrichment is defined as the ratio of the prior of genes over the prior of variants. We generally expect genes to have higher prior inclusion probability than variants. Enrichment values typically range from 20 -100 for expresion traits.</p>
<p>Then, we use <code><a href="../reference/summarize_param.html">summarize_param()</a></code> to obtain estimated parameters (from the last iteration) and to compute the proportion of variance explained (PVE) by variants and genes. These PVE values allow us to compute how heritability of the phenotype is partitioned among variants and groups of genes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="inspecting-the-ctwas-results">Inspecting the cTWAS results<a class="anchor" aria-label="anchor" href="#inspecting-the-ctwas-results"></a>
</h2>
<p>We can view the genes above a certain PIP cutoff, e.g. PIP &gt; 0.8, by:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"gene"</span> <span class="op">&amp;</span> <span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                       id group         z            region_id</span></span>
<span><span class="co">## 17101 ENSG00000261701.6|expression_Liver  gene -18.40315 16_71020125_72901251</span></span>
<span><span class="co">##       susie_pip      mu2 cs_index</span></span>
<span><span class="co">## 17101         1 632.3152        1</span></span></code></pre>
<p>The result, however, does not contain gene names and position information. To add the information, it is helpful to add gene annotations to the fine-mapping result.</p>
<p>To annotate genes, we need <code>gene_annot</code>, a data frame containing gene IDs, gene names, gene types, and position information of the genes. We have a function <code><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db()</a></code> , which extracts gene annotations from the Ensembl database.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">id</span>, split <span class="op">=</span> <span class="st">"[|]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db</a></span><span class="op">(</span><span class="va">ens_db</span>, <span class="va">gene_ids</span><span class="op">)</span></span></code></pre></div>
<p>Note: we used the gene annotations from <code>EnsDb.Hsapiens.v86</code> for the example data. You can choose the Ensembl database for your specific data, or prepare your own gene annotations in the following format:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_annot</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chrom    start      end            gene_id gene_name      gene_type</span></span>
<span><span class="co">## 1    16 53597683 53703938 ENSG00000103494.12  RPGRIP1L protein_coding</span></span>
<span><span class="co">## 2    16 54283304 54286763 ENSG00000177508.11      IRX3 protein_coding</span></span>
<span><span class="co">## 3    16 55389700 55506691 ENSG00000087245.12      MMP2 protein_coding</span></span>
<span><span class="co">## 4    16 55508998 55586670 ENSG00000087253.12    LPCAT2 protein_coding</span></span>
<span><span class="co">## 5    16 55655604 55706192 ENSG00000103546.18    SLC6A2 protein_coding</span></span>
<span><span class="co">## 6    16 55802851 55833337 ENSG00000198848.12      CES1 protein_coding</span></span></code></pre>
<p>We then use the <code><a href="../reference/anno_finemap_res.html">anno_finemap_res()</a></code> function to add gene annotations to the fine-mapping result, and use gene mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">snp_map</span>, <span class="va">gene_annot</span>,</span>
<span>                                use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-25 08:04:03 INFO::Annotating ctwas finemapping result ...</span></span>
<span><span class="co">## 2024-07-25 08:04:05 INFO::add gene_name and gene_type</span></span>
<span><span class="co">## 2024-07-25 08:04:05 INFO::Remove unannotated genes</span></span>
<span><span class="co">## 2024-07-25 08:04:05 INFO::use gene mid positions</span></span>
<span><span class="co">## 2024-07-25 08:04:05 INFO::add SNP positions</span></span></code></pre>
<p>You can now run the command at the beginning of this section to view the genes with additional information. In this example, HPR is the only gene that satisfies PIP &gt; 0.8.</p>
</div>
<div class="section level2">
<h2 id="visualizing-the-ctwas-results">Visualizing the cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-the-ctwas-results"></a>
</h2>
<p>It is often useful to visualize the results of cTWAS. This would help one understand the rationale of how cTWAS chooses a particular gene and identify issues when cTWAS doesn’t behave properly. We could use the <code><a href="../reference/make_locusplot.html">make_locusplot()</a></code> function to make locus plots for regions of interest, and zoom in the region by specifying the <code>locus_range</code>. For example:</p>
<p>The plotting function could color the genes and SNPs based on the correlations among them. If we have computed and saved correlation matrices to <code>cor_dir</code> when running cTWAS with LD, we could simply load those precomputed correlation matrices of the region (<code>R_snp_gene</code> for gene-SNP correlations and <code>R_gene</code> for gene-gene correlations), and then make the locus plot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"cor_matrix"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">cor_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_region_cor.html">get_region_cor</a></span><span class="op">(</span>region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>, cor_dir <span class="op">=</span> <span class="va">cor_dir</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>, </span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>               R_snp_gene <span class="op">=</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">R_snp_gene</span>,</span>
<span>               R_gene <span class="op">=</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">R_gene</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-25 08:04:07 INFO::focal gene: HPR</span></span>
<span><span class="co">## 2024-07-25 08:04:07 INFO::focal id: ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 2024-07-25 08:04:07 INFO::plot locus range: chr16 71600000,72400000</span></span>
<span><span class="co">## 2024-07-25 08:04:08 INFO::gene: HPR</span></span>
<span><span class="co">## 2024-07-25 08:04:08 INFO::QTL positions: 72063820,72063928</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_results_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
<p>The locus plot shows several tracks in the example region in chromosome 6 (71.6 - 72.4 Mb). The top one shows -log10(p-value) of the association of variants and genes with the phenotype. The next track shows the PIPs of variants and genes. The data points are colored by the correlations with the focal gene (the gene with the highest PIP, HPR in this case). The red dotted line shows the PIP cutoff of 0.8. The next track shows the locations of QTLs of the focus gene (by default, the focus gene is the gene with the highest PIP). The bottom is the gene track.</p>
<p>In this plot, one can see that there are several genes with good associations with the phenotype, but only HPR is selected by cTWAS (PIP = 1). This is because once HPR and other likely causal variants are selected by cTWAS, there is no remaining association left in other genes. This example thus illustrates how cTWAS selects likely causal genes while controlling for nearby genes and variants.</p>
<p>For the “no-LD” version, we could still make the locus plot. The difference is that the plot will not color the data points based on the correlations to the focal gene.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-07-25 08:04:10 INFO::focal gene: HPR</span></span>
<span><span class="co">## 2024-07-25 08:04:10 INFO::focal id: ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 2024-07-25 08:04:10 INFO::plot locus range: chr16 71600000,72400000</span></span>
<span><span class="co">## 2024-07-25 08:04:10 INFO::gene: HPR</span></span>
<span><span class="co">## 2024-07-25 08:04:10 INFO::QTL positions: 72063820,72063928</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_results_files/figure-html/locus_plot_noLD-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
