<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing and post-processing cTWAS results • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing and post-processing cTWAS results">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_postprocessing_results.html">Summarizing and post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="summarizing_postprocessing_results_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing and post-processing cTWAS results</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-05-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/summarizing_postprocessing_results.Rmd" class="external-link"><code>vignettes/summarizing_postprocessing_results.Rmd</code></a></small>
      <div class="hidden name"><code>summarizing_postprocessing_results.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="summarizing-and-visualizing-ctwas-result">Summarizing and visualizing cTWAS result<a class="anchor" aria-label="anchor" href="#summarizing-and-visualizing-ctwas-result"></a>
</h2>
<p>Load the packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
<p>Load the cTWAS results from the <code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> function</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># output directory</span></span>
<span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/tutorial/LDL_multitissue_tutorial/output"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"LDL.Liver_Adipose"</span></span>
<span></span>
<span><span class="co"># directory storing computed correlation matrices</span></span>
<span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GWAS sample size</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span>
<span></span>
<span><span class="co"># Load region info</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"region_info.RDS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># # Load z_snp</span></span>
<span><span class="co"># z_snp &lt;- readRDS(file.path(outputdir, paste0(outname, ".preprocessed.z_snp.RDS")))</span></span>
<span></span>
<span><span class="co"># Load weights</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".preprocessed.weights.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the cTWAS results </span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_sumstats_res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span></code></pre></div>
<div class="section level3">
<h3 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h3>
<p>We could use the <code><a href="../reference/summarize_param.html">summarize_param()</a></code> function to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ctwas_parameters</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_parameters.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ctwas_parameters</span></span></code></pre></div>
<pre><code><span><span class="co">## $group_size</span></span>
<span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##               8775              10538            6435220 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $group_prior</span></span>
<span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##       0.0144522860       0.0021325754       0.0002151146 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $group_prior_var</span></span>
<span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##          24.524181          24.524181           9.749157 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $enrichment</span></span>
<span><span class="co">##   expression|liver expression|adipose </span></span>
<span><span class="co">##          67.184133           9.913673 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $group_pve</span></span>
<span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##         0.00905104         0.00160390         0.03927540 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $total_pve</span></span>
<span><span class="co">## [1] 0.04993034</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $attributable_pve</span></span>
<span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##         0.18127336         0.03212276         0.78660388 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $convergence_plot</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_postprocessing_results_files/figure-html/summarize_param-1.png" alt="&amp;nbsp;" width="960"><p class="caption">
 
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="summarizing-ctwas-results">Summarizing cTWAS results<a class="anchor" aria-label="anchor" href="#summarizing-ctwas-results"></a>
</h2>
<p>To interpret and visualize the results, we need to add gene names to the finemapping result.</p>
<p>We use gene annotations from Ensembl database (EnsDb.Hsapiens.v86). Please choose the Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load gene info database</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span></code></pre></div>
<p>We then add gene names and use gene mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract gene IDs from finemapping result</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">id</span>, split <span class="op">=</span> <span class="st">"[|]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">gene_id</span>, split <span class="op">=</span> <span class="st">"[.]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="va">gene_id</span></span>
<span><span class="va">finemap_SNP_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># get gene position info</span></span>
<span><span class="va">gene_annot_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">ens_db</span>, filter <span class="op">=</span> <span class="fu">GeneIdFilter</span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">start_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">gene_annot_gr</span><span class="op">)</span></span>
<span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">end_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">gene_annot_gr</span><span class="op">)</span></span>
<span><span class="co"># use the midpoint as gene position</span></span>
<span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_mid_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="op">(</span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">start_pos</span><span class="op">+</span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">end_pos</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># get gene_name and gene position</span></span>
<span><span class="va">gene_annot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene_id <span class="op">=</span> <span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_id</span>, </span>
<span>                            gene_name <span class="op">=</span> <span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_name</span>, </span>
<span>                            gene_pos <span class="op">=</span> <span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_mid_pos</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add gene name to finemapping result and update with gene positions</span></span>
<span><span class="va">gene_annot_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span>, <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span>
<span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_pos</span><span class="op">[</span><span class="va">gene_annot_idx</span><span class="op">]</span></span>
<span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="va">gene_annot_idx</span><span class="op">]</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">finemap_gene_res</span>, <span class="va">finemap_SNP_res</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="interpreting-ctwas-results">Interpreting cTWAS results<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-results"></a>
</h2>
<p>We then show all molecular traits with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select PIP &gt; 0.8</span></span>
<span><span class="va">finemap_sig_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finemap_sig_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">finemap_sig_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualzing-ctwas-results">Visualzing cTWAS results<a class="anchor" aria-label="anchor" href="#visualzing-ctwas-results"></a>
</h2>
<p>We could make locus plots for regions of interest.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"1:107867043-109761309"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               cor_dir <span class="op">=</span> <span class="va">cor_dir</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               gene_biotype<span class="op">=</span><span class="st">"protein_coding"</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-05-09 15:33:09 INFO::focus gene: PSRC1</span></span>
<span><span class="co">## 2024-05-09 15:33:09 INFO::focus id: ENSG00000134222.16|mashr_Liver</span></span>
<span><span class="co">## 2024-05-09 15:33:09 INFO::locus_range: 107768052,109760983</span></span>
<span><span class="co">## 2024-05-09 15:33:12 INFO::QTL positions: 109275684</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_postprocessing_results_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="960"><p class="caption">
 
</p>
</div>
<p>We could zoom in the region by specifing the <code>locus_range</code>, and highlight the positions using <code>highlight_pos</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"1:107867043-109761309"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               cor_dir <span class="op">=</span> <span class="va">cor_dir</span>, </span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">109e6</span>, <span class="fl">109.6e6</span><span class="op">)</span>,</span>
<span>               gene_biotype<span class="op">=</span><span class="st">"protein_coding"</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>               panel.heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">0.3</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>               highlight_pos <span class="op">=</span> <span class="fl">109275684</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-05-09 15:33:14 INFO::focus gene: PSRC1</span></span>
<span><span class="co">## 2024-05-09 15:33:14 INFO::focus id: ENSG00000134222.16|mashr_Liver</span></span>
<span><span class="co">## 2024-05-09 15:33:14 INFO::locus_range: 1.09e+08,109600000</span></span>
<span><span class="co">## 2024-05-09 15:33:17 INFO::QTL positions: 109275684</span></span>
<span><span class="co">## 2024-05-09 15:33:17 INFO::highlight positions: 109275684</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_postprocessing_results_files/figure-html/locus_plot_2-1.png" alt="&amp;nbsp;" width="960"><p class="caption">
 
</p>
</div>
<p>If we do not have precomputed correlation matrices, we could compute those by providing region_info and region_data</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"1:107867043-109761309"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               region_info <span class="op">=</span> <span class="va">region_info</span>, </span>
<span>               region_data <span class="op">=</span> <span class="va">region_data</span>,</span>
<span>               cor_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">109e6</span>, <span class="fl">109.6e6</span><span class="op">)</span>,</span>
<span>               gene_biotype<span class="op">=</span><span class="st">"protein_coding"</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>               panel.heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">0.3</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>               highlight_pos <span class="op">=</span> <span class="fl">109275684</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="post-processing">Post-processing<a class="anchor" aria-label="anchor" href="#post-processing"></a>
</h2>
<div class="section level3">
<h3 id="merging-regions">Merging regions<a class="anchor" aria-label="anchor" href="#merging-regions"></a>
</h3>
<p>We first select cross-boundary genes to merge, We then identify overlapping regions from these cross-boundary genes, and create <code>merged_region_data</code> and <code>merged_region_info</code> for these merged regions.</p>
<p>We rerun finemapping for the merged regions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">high_PIP_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span> <span class="op">!=</span> <span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;=</span> <span class="fl">0.5</span>, <span class="st">"id"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">selected_boundary_genes</span> <span class="op">&lt;-</span> <span class="va">boundary_genes</span><span class="op">[</span><span class="va">boundary_genes</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">high_PIP_genes</span>, , drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">selected_boundary_genes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_finemap_regions.html">merge_finemap_regions</a></span><span class="op">(</span><span class="va">selected_boundary_genes</span>,</span>
<span>                               <span class="va">region_data</span>,</span>
<span>                               <span class="va">region_info</span>,</span>
<span>                               <span class="va">finemap_res</span>,</span>
<span>                               <span class="va">z_snp</span>,</span>
<span>                               <span class="va">z_gene</span>,</span>
<span>                               <span class="va">weights</span>,</span>
<span>                               expand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                               group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                               group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                               L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                               cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/cor_matrix/"</span><span class="op">)</span>,</span>
<span>                               ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">finemap_merged_regions_res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">finemap_merged_regions_res</span></span>
<span>  <span class="va">merged_region_data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">merged_region_data</span></span>
<span>  <span class="va">merged_region_info</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">merged_region_info</span></span>
<span>  <span class="va">merge_region_id_list</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">merge_region_id_list</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Optionally, we could update the finemapping results for the regions got merged.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">updated_finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/update_merged_regions_finemap_res.html">update_merged_regions_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>                                                        <span class="va">finemap_merged_regions_res</span>, </span>
<span>                                                        <span class="va">merge_region_id_list</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rerun-finemapping-with-l-1-for-regions-with-ld-mismatches">Rerun finemapping with L = 1 for regions with LD mismatches<a class="anchor" aria-label="anchor" href="#rerun-finemapping-with-l-1-for-regions-with-ld-mismatches"></a>
</h3>
<p>LD mismatches between GWAS data and the reference LD could lead to false positives in fine-mapping. Diagnostic tools including <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link">SuSiE-RSS</a>, and <a href="https://github.com/Yves-CHEN/DENTIST/" class="external-link">DENTIST</a>, have been developed to check possible LD mismatch.</p>
<p>Here, we perform the LD mismatch diagnosis using <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link">SuSiE-RSS</a> for selected regions. It is an optional step that users could run after finishing the main cTWAS analysis. Users could choose regions of interest, to be confident of the results.</p>
<p>Here we check the top 10 regions with highest total non-SNP PIPs:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nonSNP_PIP_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_region_nonSNP_PIPs.html">compute_region_nonSNP_PIPs</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">)</span></span>
<span><span class="va">selected_nonSNP_PIP_df</span> <span class="op">&lt;-</span> <span class="va">nonSNP_PIP_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">nonSNP_PIP_df</span><span class="op">$</span><span class="va">nonSNP_PIP</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">selected_region_ids</span> <span class="op">&lt;-</span> <span class="va">selected_nonSNP_PIP_df</span><span class="op">$</span><span class="va">region_id</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<p>We perform LD mismatch diagnosis for these regions using SuSiE RSS to detect problematic SNPs. This returns a list of problematic SNPs (diagnostic test <code>p-value &lt; 5e-8</code> by default), flipped SNPs, and the test statistics of <code>kriging_rss</code> function from SuSiE-RSS.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_region_info</span> <span class="op">&lt;-</span> <span class="va">region_info</span><span class="op">[</span><span class="va">region_info</span><span class="op">$</span><span class="va">region_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">selected_region_ids</span>, <span class="op">]</span></span>
<span><span class="co"># We use SuSiE RSS to perform LD mismatch diagnosis for these high PIP regions.</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detect_ld_mismatch_susie.html">detect_ld_mismatch_susie</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">selected_region_info</span>, <span class="va">gwas_n</span>, </span>
<span>                                ncore <span class="op">=</span> <span class="va">ncore</span>, p_diff_thresh <span class="op">=</span> <span class="fl">5e-8</span><span class="op">)</span></span>
<span><span class="va">problematic_snps</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">problematic_snps</span></span>
<span><span class="va">flipped_snps</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">flipped_snps</span></span>
<span><span class="va">condz_stats</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">condz_stats</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".detect_ld_mismatch_res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We choose large effect genes (or molecular traits) (abs(z) &gt; 3) with problematic SNPs in their weights, and then select regions containing the problematic genes (molecular traits).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">problematic_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_problematic_genes.html">get_problematic_genes</a></span><span class="op">(</span><span class="va">problematic_snps</span>, </span>
<span>                                           <span class="va">weights</span>, </span>
<span>                                           <span class="va">z_gene</span>, </span>
<span>                                           z_thresh <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">problematic_region_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">problematic_genes</span>, <span class="st">"region_id"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We then rerun the fine-mapping with <code>L = 1</code> for the problematic regions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">problematic_region_ids</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># rerun the fine-mapping with L = 1 for the problematic regions</span></span>
<span>  <span class="va">rerun_region_data</span> <span class="op">&lt;-</span> <span class="va">screened_region_data</span><span class="op">[</span><span class="va">problematic_region_ids</span><span class="op">]</span></span>
<span>  <span class="va">finemap_rerun_region_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">rerun_region_data</span>,</span>
<span>                                              <span class="va">region_info</span>,</span>
<span>                                              <span class="va">weights</span>,</span>
<span>                                              group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                              group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                              L <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                     cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span>, </span>
<span>                                              ncore <span class="op">=</span> <span class="va">ncore</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Optimally, we could update finemapping results for problematic regions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">updated_finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/update_finemap_res.html">update_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">finemap_rerun_region_res</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
