<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing and visualizing cTWAS results • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing and visualizing cTWAS results">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="summarizing_results_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing and visualizing cTWAS results</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-12-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/vignettes/summarizing_results.Rmd" class="external-link"><code>vignettes/summarizing_results.Rmd</code></a></small>
      <div class="hidden name"><code>summarizing_results.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will show how to summarize and visualize cTWAS results.</p>
<p>Load the packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">ctwas</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre></div>
<p>We need the Ensembl database <code>EnsDb.Hsapiens.v86</code> (for hg38) in this tutorial, please choose the version of Ensembl database for your specific data (e.g. <code>EnsDb.Hsapiens.v75</code> for hg19).</p>
<p>Let’s first load the cTWAS input data needed to run this tutorial: GWAS sample size (<code>gwas_n</code>), prediction models of molecular traits (<code>weights</code>), the reference data: including information about regions (<code>region_info</code>), reference variant list (<code>snp_map</code>) and reference LD (<code>LD_map</code>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span>
<span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.weights.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.region_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snp_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.snp_map.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LD_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.LD_map.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We then load the cTWAS result for the sample data (chr16) using parameters estimated from the entire genome.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.ctwas_sumstats_res_param_allchrs.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">susie_alpha_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_alpha_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">screen_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">screen_res</span></span></code></pre></div>
<p><code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> returns several main output:</p>
<ul>
<li><p><code>z_gene</code>: the Z-scores of molecular traits,</p></li>
<li><p><code>param</code>: the estimated parameters,</p></li>
<li><p><code>finemap_res</code>: fine-mapping results, a data frame of molecular traits and variants, including their cTWAS IDs (“id”), molecular ID (“molecular_ id”), the regions they belong to (“region_id”), Z-scores (“z”), PIPs (“susie_pip”), credible set indices (“cs_index”), and estimated effect size variance (“mu2”).</p></li>
</ul>
<p>The IDs of molecular traits (“id”) used in cTWAS is in the format of &lt;molecular_ id|weight_name&gt;. “molecular_id” are the IDs originally used in the PredictDB or FUSION weights for genes or other molecular traits, such as Ensembl gene IDs. <code>weight_name</code> is the user defined weight name when preprocessing weights, see the tutorial <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/preparing_input_data.html" class="external-link">“Preparing cTWAS input data”</a>).</p>
<p>It also produces other output:</p>
<ul>
<li><p><code>susie_alpha_res</code>: a data frame with finemapping results of molecular traits and the single effect probabilities (alpha) in all credible sets. This will be used when computing combined gene PIPs below.</p></li>
<li><p><code>boundary_genes</code>: the genes/molecular traits that cross region boundaries (we will need these in the post-processing step).</p></li>
<li><p><code>region_data</code>: assembled region data, including Z scores of SNPs and molecular traits, for all the regions. Note that only data of the subset of SNPs used in screening regions are included.</p></li>
<li><p><code>screen_res</code>: screening regions results, including the data of all SNPs and molecular traits of the selected regions, estimated numbers of causal signals (<code>L</code>) for selected regions, and a data frame with estimated <code>L</code> and non-SNP PIPs for all regions.</p></li>
</ul>
<p>We focus on examples from running cTWAS with LD here, but the steps below also mostly work for running without LD.</p>
<div class="section level2">
<h2 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h2>
<p>First, we make plots using the function <code><a href="../reference/make_convergence_plots.html">make_convergence_plots()</a></code> to see how estimated parameters converge during the execution of the program:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_convergence_plots.html">make_convergence_plots</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="summarizing_results_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
<p>These plots show the estimated prior inclusion probability, prior effect size variance, enrichment and proportion of variance explained (PVE) over the iterations of parameter estimation. The enrichment is defined as the ratio of the prior inclusion probability of molecular traits over the prior inclusion probability of variants. We generally expect molecular traits to have higher prior inclusion probability than variants. Enrichment values typically range from 20 - 100 for expression traits.</p>
<p>Then, we use <code><a href="../reference/summarize_param.html">summarize_param()</a></code> to obtain estimated parameters (from the last iteration) and to compute the PVE by variants and molecular traits.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span>
<span><span class="va">ctwas_parameters</span></span></code></pre></div>
<pre><code><span><span class="co">## $group_size</span></span>
<span><span class="co">##   liver|expression adipose|expression                SNP </span></span>
<span><span class="co">##               8775              10538            7556260 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $group_prior</span></span>
<span><span class="co">##   liver|expression adipose|expression                SNP </span></span>
<span><span class="co">##       0.0128593248       0.0011512596       0.0001901423 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $group_prior_var</span></span>
<span><span class="co">##   liver|expression adipose|expression                SNP </span></span>
<span><span class="co">##           27.68179           27.68179           10.08116 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $enrichment</span></span>
<span><span class="co">##   liver|expression adipose|expression </span></span>
<span><span class="co">##          67.630014           6.054727 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $group_pve</span></span>
<span><span class="co">##   liver|expression adipose|expression                SNP </span></span>
<span><span class="co">##       0.0090903336       0.0009773407       0.0421518279 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $total_pve</span></span>
<span><span class="co">## [1] 0.0522195</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $prop_heritability</span></span>
<span><span class="co">##   liver|expression adipose|expression                SNP </span></span>
<span><span class="co">##         0.17407928         0.01871601         0.80720471</span></span></code></pre>
<p>This function returns several output: <code>group_prior</code> are the estimated prior inclusion probabilities for molecular traits and variants; <code>group_prior_var</code> are estimated prior effect size variance for molecular traits and variants; <code>enrichment</code> are estimated enrichment of molecular traits over variants, defined as ratios of prior inclusion probabilities. Other output parameters are related to the proportion of variance explained (PVE) by molecular traits or variants: <code>group_pve</code> are PVE of molecular traits and variants; <code>total_pve</code> is the total PVE from all molecular traits and variants (sum of <code>group_pve</code>) - basically heritability from all included variables (effectively all common variants); <code>prop_heritability</code> are the proportion of heritability mediated by molecular traits and variants.</p>
<p>These PVE values allow us to compute how heritability of the phenotype is partitioned among variants and groups of molecular traits. From our experience, in single-tissue gene expression analysis, genes typically explain around 5-15% of the total heritability of the trait. In multi-group analysis, the heritability of genes could become lower.</p>
</div>
<div class="section level2">
<h2 id="inspecting-and-summarizing-the-ctwas-results">Inspecting and summarizing the cTWAS results<a class="anchor" aria-label="anchor" href="#inspecting-and-summarizing-the-ctwas-results"></a>
</h2>
<p>We can add p-values (computed from Z-scores) to the fine-mapping results.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/z2p.html">z2p</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">z</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                    id       molecular_id       type context</span></span>
<span><span class="co">## 1 ENSG00000103494.12|liver_expression ENSG00000103494.12 expression   liver</span></span>
<span><span class="co">## 2 ENSG00000177508.11|liver_expression ENSG00000177508.11 expression   liver</span></span>
<span><span class="co">## 3 ENSG00000087245.12|liver_expression ENSG00000087245.12 expression   liver</span></span>
<span><span class="co">## 4 ENSG00000087253.12|liver_expression ENSG00000087253.12 expression   liver</span></span>
<span><span class="co">## 5 ENSG00000103546.18|liver_expression ENSG00000103546.18 expression   liver</span></span>
<span><span class="co">## 6 ENSG00000198848.12|liver_expression ENSG00000198848.12 expression   liver</span></span>
<span><span class="co">##              group            region_id          z    susie_pip      mu2   cs</span></span>
<span><span class="co">## 1 liver|expression 16_53348660_55869862 -0.4550967 1.524699e-06 1.158057 &lt;NA&gt;</span></span>
<span><span class="co">## 2 liver|expression 16_53348660_55869862 -0.3912807 1.485472e-06 1.107746 &lt;NA&gt;</span></span>
<span><span class="co">## 3 liver|expression 16_53348660_55869862  0.5636824 1.608305e-06 1.261103 &lt;NA&gt;</span></span>
<span><span class="co">## 4 liver|expression 16_53348660_55869862 -1.3879522 3.495492e-06 2.759558 &lt;NA&gt;</span></span>
<span><span class="co">## 5 liver|expression 16_53348660_55869862  0.5507643 1.597171e-06 1.247693 &lt;NA&gt;</span></span>
<span><span class="co">## 6 liver|expression 16_53348660_55869862 -0.6042867 1.645537e-06 1.305278 &lt;NA&gt;</span></span>
<span><span class="co">##        pval</span></span>
<span><span class="co">## 1 0.6490397</span></span>
<span><span class="co">## 2 0.6955898</span></span>
<span><span class="co">## 3 0.5729703</span></span>
<span><span class="co">## 4 0.1651516</span></span>
<span><span class="co">## 5 0.5817952</span></span>
<span><span class="co">## 6 0.5456531</span></span></code></pre>
<p>We can view the molecular traits above a certain PIP cutoff, e.g. PIP &gt; 0.8. We also generally recommend limiting the results to molecular traits within credible sets.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">!=</span> <span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">cs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                          id       molecular_id       type</span></span>
<span><span class="co">## 7899  ENSG00000087237.10|adipose_expression ENSG00000087237.10 expression</span></span>
<span><span class="co">## 17549    ENSG00000261701.6|liver_expression  ENSG00000261701.6 expression</span></span>
<span><span class="co">##       context              group            region_id         z susie_pip</span></span>
<span><span class="co">## 7899  adipose adipose|expression 16_55869862_57630418  13.74984 0.9996853</span></span>
<span><span class="co">## 17549   liver   liver|expression 16_71020125_72901251 -18.40315 1.0000000</span></span>
<span><span class="co">##            mu2 cs         pval</span></span>
<span><span class="co">## 7899  177.0699 L1 5.104261e-43</span></span>
<span><span class="co">## 17549 430.9403 L1 1.239454e-75</span></span></code></pre>
<p>Below we describe how to aggregate the information from molecular traits targeting the same genes, to compute the evidence of a gene being causal (denoted as gene PIPs below). Additionally, we will show how to add annotations (e.g. gene names and positions) to the cTWAS results that may be needed in downstream analysis.</p>
<div class="section level3">
<h3 id="computing-gene-pips">Computing gene PIPs<a class="anchor" aria-label="anchor" href="#computing-gene-pips"></a>
</h3>
<p>When we have multiple contexts or different types of molecular traits, it is useful to evaluate the total evidence of a gene being causal, <em>combining evidence of all the molecular traits affecting this gene across all types and contexts.</em></p>
<p>We use the function <code><a href="../reference/combine_gene_pips.html">combine_gene_pips()</a></code> to compute combined PIPs. It groups molecular traits targeting the same gene (we use <code>group_by</code> to select the column to group molecular traits). It then computes combined PIPs across contexts (<code>by = "context"</code>, default option), types (<code>by = "type"</code>), or both (<code>by = "group"</code>).</p>
<p>We have several ways of combining PIPs:</p>
<ul>
<li>“combine_cs” (default option): it first sums PIPs of molecular traits of a genes in each credible set, and then combine PIPs using the following formula: <span class="math inline">\(1 - \prod_k (1 - \text{PIP}_k)\)</span>, where <span class="math inline">\(\text{PIP}_k\)</span> is the summed PIP of the <span class="math inline">\(k\)</span>-th credible set of a gene.</li>
<li>“sum”: sum over PIPs of all molecular traits for the same gene. This summation is the expected number of causal molecular traits in this gene, and could be higher than 1.</li>
</ul>
<p>For example, when using gene expression (eQTL) data from multiple tissues, we can use the function to compute combined PIPs across contexts, using the “combine_cs” method. By default, we limit the results to molecular traits within credible sets (<code>filter_cs = TRUE</code>). We also report the credible set information of each molecular trait in the output (<code>include_cs_id = TRUE</code>), which allows us to understand how gene PIPs are obtained.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_pip_by_context</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_gene_pips.html">combine_gene_pips</a></span><span class="op">(</span><span class="va">susie_alpha_res</span>, </span>
<span>                                             group_by <span class="op">=</span> <span class="st">"molecular_id"</span>,</span>
<span>                                             by <span class="op">=</span> <span class="st">"context"</span>,</span>
<span>                                             method <span class="op">=</span> <span class="st">"combine_cs"</span>,</span>
<span>                                             filter_cs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                             include_cs_id <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">combined_pip_by_context</span>, <span class="va">combined_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         molecular_id          combined_cs_id combined_pip</span></span>
<span><span class="co">## 1  ENSG00000261701.6 16_71020125_72901251.L1    1.0000000</span></span>
<span><span class="co">## 2 ENSG00000087237.10 16_55869862_57630418.L1    0.9996853</span></span>
<span><span class="co">##               liver_cs_id liver_pip           adipose_cs_id adipose_pip</span></span>
<span><span class="co">## 1 16_71020125_72901251.L1         1                    &lt;NA&gt;          NA</span></span>
<span><span class="co">## 2                    &lt;NA&gt;        NA 16_55869862_57630418.L1   0.9996853</span></span></code></pre>
<p>In this example, we have gene IDs in the “molecular_id” column, so we could use that column to group genes. If the gene IDs are not provided, we will need to map molecular traits to their target genes in order to compute gene PIPs, as described below.</p>
</div>
<div class="section level3">
<h3 id="adding-gene-annotations-to-ctwas-results">Adding gene annotations to cTWAS results<a class="anchor" aria-label="anchor" href="#adding-gene-annotations-to-ctwas-results"></a>
</h3>
<p>In the example above, we have Ensembl gene IDs in cTWAS results. It is often helpful to add additional gene information such as gene names, gene types (protein coding, non-coding RNAs, etc.) and gene positions.</p>
<p>We have a helper function <code><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db()</a></code>, which extracts gene annotations from the Ensembl database for a list of genes. It takes an Ensembl database and a list of Ensembl gene IDs as input, and returns <code>gene_annot</code>, a data frame with gene IDs, gene names, gene types, and gene positions for the genes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">!=</span> <span class="st">"SNP"</span><span class="op">)</span></span>
<span><span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">molecular_id</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db</a></span><span class="op">(</span><span class="va">ens_db</span>, <span class="va">gene_ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_annot</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_annot</span><span class="op">)</span> <span class="op">==</span> <span class="st">"gene_id"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"molecular_id"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_annot</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         molecular_id gene_name      gene_type chrom    start      end</span></span>
<span><span class="co">## 1 ENSG00000103494.12  RPGRIP1L protein_coding    16 53597683 53703938</span></span>
<span><span class="co">## 2 ENSG00000177508.11      IRX3 protein_coding    16 54283304 54286763</span></span>
<span><span class="co">## 3 ENSG00000087245.12      MMP2 protein_coding    16 55389700 55506691</span></span>
<span><span class="co">## 4 ENSG00000087253.12    LPCAT2 protein_coding    16 55508998 55586670</span></span>
<span><span class="co">## 5 ENSG00000103546.18    SLC6A2 protein_coding    16 55655604 55706192</span></span>
<span><span class="co">## 6 ENSG00000198848.12      CES1 protein_coding    16 55802851 55833337</span></span></code></pre>
<p>Note: we used the gene annotations from <code>EnsDb.Hsapiens.v86</code> for the example data in hg38. You should use the Ensembl database for the genome build of your own data (e.g. <code>EnsDb.Hsapiens.v75</code> for hg19).</p>
<p>Basically, <code>gene_annot</code> serves as a map between Ensembl gene IDs and the corresponding genes. Using this map, we could use the function <code><a href="../reference/anno_finemap_res.html">anno_finemap_res()</a></code> to add additional columns of gene annotations and positions of genes and variants to the fine-mapping results.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>                                snp_map <span class="op">=</span> <span class="va">snp_map</span>,</span>
<span>                                mapping_table <span class="op">=</span> <span class="va">gene_annot</span>,</span>
<span>                                add_gene_annot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                map_by <span class="op">=</span> <span class="st">"molecular_id"</span>,</span>
<span>                                drop_unmapped <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                add_position <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-12-18 16:07:06 INFO::Annotating fine-mapping result ...</span></span>
<span><span class="co">## 2024-12-18 16:07:06 INFO::Map molecular traits to genes</span></span>
<span><span class="co">## 2024-12-18 16:07:07 INFO::Add gene positions</span></span>
<span><span class="co">## 2024-12-18 16:07:07 INFO::Add SNP positions</span></span></code></pre>
<p>If <code>add_gene_annot = TRUE</code>, it joins the fine-mapping result table (<code>finemap_res</code>) with the <code>gene_annot</code> table, by the “molecular ids” column (set in <code>map_by</code>), and adds additional columns of gene annotations to the fine-mapping results By default, we drop the genes not found in <code>gene_annot</code> (<code>drop_unmapped = TRUE</code>). If <code>add_position = TRUE</code>, it adds positional information to the fine-mapping results. The positional information of variants can be found in the data structure of reference SNPs: <code>snp_map</code>. The users need to provide positional information of molecular traits, through <code>mapping_table</code>. In our example, we use <code>gene_annot</code>, which has gene positions. By default, we will use the midpoint positions for genes. You could also use “start” or “end” positions by setting the <code>use_gene_pos</code> argument.</p>
<p>With the added gene information, we could limit results to protein coding genes, and view the prioritized genes with additional information:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">group</span> <span class="op">!=</span> <span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">gene_type</span> <span class="op">==</span> <span class="st">"protein_coding"</span> <span class="op">&amp;</span> <span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">cs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                        id       molecular_id       type context</span></span>
<span><span class="co">## 51  ENSG00000087237.10|adipose_expression ENSG00000087237.10 expression adipose</span></span>
<span><span class="co">## 111    ENSG00000261701.6|liver_expression  ENSG00000261701.6 expression   liver</span></span>
<span><span class="co">##                  group            region_id         z susie_pip      mu2 cs</span></span>
<span><span class="co">## 51  adipose|expression 16_55869862_57630418  13.74984 0.9996853 177.0699 L1</span></span>
<span><span class="co">## 111   liver|expression 16_71020125_72901251 -18.40315 1.0000000 430.9403 L1</span></span>
<span><span class="co">##             pval gene_name      gene_type chrom    start      end      pos</span></span>
<span><span class="co">## 51  5.104261e-43      CETP protein_coding    16 56961850 56983845 56972848</span></span>
<span><span class="co">## 111 1.239454e-75       HPR protein_coding    16 72063224 72077246 72070235</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="integrating-multiple-types-of-molecular-traits">Integrating multiple types of molecular traits<a class="anchor" aria-label="anchor" href="#integrating-multiple-types-of-molecular-traits"></a>
</h3>
<p>For more complex settings that involve different types of molecular traits, we will need to map molecular traits to their corresponding genes in order to compute gene PIPs.</p>
<p>To map molecular traits to their corresponding genes, we need a data frame <code>mapping_table</code>, with IDs of molecular traits (“molecular_id”), and the corresponding gene names (“gene_name”). It is helpful to include additional gene information, such as gene types and gene positions. We provide a <code>mapping_table</code> to map introns to genes for PredictDB expression and slicing data <a href="https://uchicago.box.com/s/gx7raqf5c9jp6tgvm5tpc2ej7zr6bnxz" class="external-link">here</a>. The mapping file is also available on UChicago RCC server: <code>/project2/xinhe/shared_data/multigroup_ctwas/weights/mapping_files/PredictDB_mapping.RDS</code> .</p>
<p>For example:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapping_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/weights/mapping_files/PredictDB_mapping.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mapping_table</span><span class="op">[</span><span class="va">mapping_table</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">==</span> <span class="st">"UBR1"</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       molecular_id gene_name      gene_type chrom    start</span></span>
<span><span class="co">## 310222          ENSG00000159459.11      UBR1 protein_coding    15 42942897</span></span>
<span><span class="co">## 173881 intron_15_43038232_43043215      UBR1 protein_coding    15 43038232</span></span>
<span><span class="co">## 173882 intron_15_43043395_43047161      UBR1 protein_coding    15 43043395</span></span>
<span><span class="co">## 173883 intron_15_43047289_43048392      UBR1 protein_coding    15 43047289</span></span>
<span><span class="co">## 173884 intron_15_43048491_43054742      UBR1 protein_coding    15 43048491</span></span>
<span><span class="co">## 173885 intron_15_43054899_43056344      UBR1 protein_coding    15 43054899</span></span>
<span><span class="co">##             end</span></span>
<span><span class="co">## 310222 43106113</span></span>
<span><span class="co">## 173881 43043215</span></span>
<span><span class="co">## 173882 43047161</span></span>
<span><span class="co">## 173883 43048392</span></span>
<span><span class="co">## 173884 43054742</span></span>
<span><span class="co">## 173885 43056344</span></span></code></pre>
<p>Using this mapping table, we can use the <code><a href="../reference/anno_finemap_res.html">anno_finemap_res()</a></code> function to add additional columns to the fine-mapping results. If a molecular trait is mapped to multiple genes, it splits the PIP of that molecular trait to its target genes. We use <code>gene_annot</code> below as the mapping table between expression traits and the corresponding gene names.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>                                snp_map <span class="op">=</span> <span class="va">snp_map</span>,</span>
<span>                                mapping_table <span class="op">=</span> <span class="va">gene_annot</span>,</span>
<span>                                add_gene_annot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                map_by <span class="op">=</span> <span class="st">"molecular_id"</span>,</span>
<span>                                drop_unmapped <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                add_position <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
<p>Similarly, we use the function <code><a href="../reference/anno_susie_alpha_res.html">anno_susie_alpha_res()</a></code> to add additional columns to <code>susie_alpha_res</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">susie_alpha_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_susie_alpha_res.html">anno_susie_alpha_res</a></span><span class="op">(</span><span class="va">susie_alpha_res</span>,</span>
<span>                                        mapping_table <span class="op">=</span> <span class="va">gene_annot</span>,</span>
<span>                                        map_by <span class="op">=</span> <span class="st">"molecular_id"</span>,</span>
<span>                                        drop_unmapped <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-12-18 16:07:10 INFO::Annotating susie alpha result ...</span></span>
<span><span class="co">## 2024-12-18 16:07:10 INFO::Map molecular traits to genes</span></span></code></pre>
<p>We could use the <code><a href="../reference/combine_gene_pips.html">combine_gene_pips()</a></code> function to compute gene PIPs across different types of molecular traits. Here we use <code>group_by = "gene_name"</code>, because molecular traits are mapped to their corresponding genes by the “gene_name” column.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_pip_by_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_gene_pips.html">combine_gene_pips</a></span><span class="op">(</span><span class="va">susie_alpha_res</span>,</span>
<span>                                          group_by <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>                                          by <span class="op">=</span> <span class="st">"type"</span>,</span>
<span>                                          method <span class="op">=</span> <span class="st">"combine_cs"</span>,</span>
<span>                                          filter_cs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                          include_cs_id <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">combined_pip_by_type</span>, <span class="va">combined_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   gene_name          combined_cs_id combined_pip        expression_cs_id</span></span>
<span><span class="co">## 1       HPR 16_71020125_72901251.L1    1.0000000 16_71020125_72901251.L1</span></span>
<span><span class="co">## 2      CETP 16_55869862_57630418.L1    0.9996853 16_55869862_57630418.L1</span></span>
<span><span class="co">##   expression_pip</span></span>
<span><span class="co">## 1      1.0000000</span></span>
<span><span class="co">## 2      0.9996853</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-the-ctwas-results">Visualizing the cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-the-ctwas-results"></a>
</h2>
<p>It is often useful to visualize the results of cTWAS. This would help one understand the rationale of how cTWAS chooses a particular molecular trait and identify issues when cTWAS doesn’t behave properly.</p>
<p>We make locus plots to visualize the association of variants and molecular traits, cTWAS PIP results, and other information, such as QTLs and gene annotations. We illustrate this with an example below.</p>
<p>We use the <code><a href="../reference/make_locusplot.html">make_locusplot()</a></code> function to make locus plots for regions of interest. We need the fine-mapping result with position information, region ID, Ensembl gene annotation database (<code>ens_db</code>) for plotting the gene track, and optionally, preprocessed weights (<code>weights</code>) for plotting the QTL track.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>               filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               filter_cs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               color_pval_by <span class="op">=</span> <span class="st">"cs"</span>,</span>
<span>               color_pip_by <span class="op">=</span> <span class="st">"cs"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-12-18 16:07:10 INFO::Limit to protein coding genes</span></span>
<span><span class="co">## 2024-12-18 16:07:10 INFO::focal id: ENSG00000261701.6|liver_expression</span></span>
<span><span class="co">## 2024-12-18 16:07:10 INFO::focal molecular trait: HPR liver expression</span></span>
<span><span class="co">## 2024-12-18 16:07:10 INFO::Range of locus: chr16:71020348-72900542</span></span>
<span><span class="co">## 2024-12-18 16:07:10 INFO::focal molecular trait QTL positions: 72063820,72063928</span></span>
<span><span class="co">## 2024-12-18 16:07:10 INFO::Limit PIPs to credible sets</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_results_files/figure-html/locus_plot_1-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
<p>The locus plot shows several tracks in the example region. The top one shows -log10(p-value) of the association of variants (from GWAS) and molecular traits (from the package computed z-scores) with the phenotype. The next track shows the PIPs of variants and molecular traits. By default, we limit PIP results to credible sets in the PIP track (<code>filter_cs = TRUE</code>). The next track shows the QTLs of the focal gene. By default, it chooses the molecular trait with the highest PIP (“HPR” in this case). We could specify the focal gene by setting the <code>focal_id</code> to the “ID” of interest, or <code>focal_gene</code> to the gene name of interest. If there are multiple molecular traits or contexts of the focal gene, it will automatically use the one with the highest PIP. The bottom is the gene track. We limit results to protein coding genes by default. We can draw a red dotted line to show the PIP cutoff of 0.8 by setting <code>highlight_pip = 0.8</code>.</p>
<p>There are a few ways to color the data points in the p-value and PIP tracks, by setting <code>color_pval_by</code> and <code>color_pip_by</code> options. “cs” would color the data points by credible sets (default option); “LD” would color data points by correlations with the focal gene; “none” would use the same colors for all data points, except for the focal gene.</p>
<p>To color data points by correlations with the focal gene, We need correlation matrices among molecular traits and SNPs. If we have saved correlation matrices in <code>cor_dir</code> when running cTWAS with LD, we could simply load those precomputed correlation matrices of the region using the function <code><a href="../reference/load_region_cor.html">load_region_cor()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_id</span> <span class="op">&lt;-</span> <span class="st">"16_71020125_72901251"</span></span>
<span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"cor_matrix"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">cor_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_region_cor.html">load_region_cor</a></span><span class="op">(</span><span class="va">region_id</span>, cor_dir <span class="op">=</span> <span class="va">cor_dir</span><span class="op">)</span></span></code></pre></div>
<p>If we don’t have precomputed correlation matrices, we could compute the correlation matrices of the region using the function <code><a href="../reference/get_region_cor.html">get_region_cor()</a></code>. We will need the region data with all SNPs (<code>screened_region_data</code>), <code>LD_map</code> and <code>weights</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">screen_res</span><span class="op">$</span><span class="va">screened_region_data</span></span>
<span><span class="va">region_id</span> <span class="op">&lt;-</span> <span class="st">"16_71020125_72901251"</span></span>
<span><span class="va">cor_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_region_cor.html">get_region_cor</a></span><span class="op">(</span><span class="va">region_id</span>,</span>
<span>                          sids <span class="op">=</span> <span class="va">screened_region_data</span><span class="op">[[</span><span class="va">region_id</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sid</span>,</span>
<span>                          gids <span class="op">=</span> <span class="va">screened_region_data</span><span class="op">[[</span><span class="va">region_id</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gid</span>,</span>
<span>                          LD_map <span class="op">=</span> <span class="va">LD_map</span>,</span>
<span>                          weights <span class="op">=</span> <span class="va">weights</span><span class="op">)</span></span></code></pre></div>
<p>The locus plot above shows the whole region. We could zoom in a region of interest by specifying the <code>locus_range</code> argument. In the example below, we zoom in the region, color the p-value track by correlations with the focal gene, and color the PIP track by credible sets. We could also highlight positions of interest by setting <code>highlight_pos</code>, for example, we highlight the position of HPR gene below.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>, </span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>               highlight_pos <span class="op">=</span> <span class="fl">72070235</span>,</span>
<span>               R_snp_gene <span class="op">=</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">R_snp_gene</span>,</span>
<span>               R_gene <span class="op">=</span> <span class="va">cor_res</span><span class="op">$</span><span class="va">R_gene</span>,</span>
<span>               filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               filter_cs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               color_pval_by <span class="op">=</span> <span class="st">"LD"</span>,</span>
<span>               color_pip_by <span class="op">=</span> <span class="st">"cs"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-12-18 16:07:16 INFO::Limit to protein coding genes</span></span>
<span><span class="co">## 2024-12-18 16:07:16 INFO::focal id: ENSG00000261701.6|liver_expression</span></span>
<span><span class="co">## 2024-12-18 16:07:16 INFO::focal molecular trait: HPR liver expression</span></span>
<span><span class="co">## 2024-12-18 16:07:16 INFO::Range of locus: chr16:71600000-72400000</span></span>
<span><span class="co">## 2024-12-18 16:07:16 INFO::focal molecular trait QTL positions: 72063820,72063928</span></span>
<span><span class="co">## 2024-12-18 16:07:16 INFO::Limit PIPs to credible sets</span></span>
<span><span class="co">## 2024-12-18 16:07:17 INFO::highlight positions: 72070235</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_results_files/figure-html/locus_plot_2-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
<p>In this plot, one can see that there are several genes with good associations with the phenotype (top panel), but only HPR in liver is prioritized by cTWAS (second panel). This suggests that other associations are likely due to their correlations with the prioritized genes rather than representing independent signals.</p>
<p>For the “no-LD” version, we could still make the locus plot. The difference is that it could not color the data points by correlations with the focal gene.</p>
</div>
<div class="section level2">
<h2 id="sample-report-of-the-ctwas-results">Sample report of the cTWAS results<a class="anchor" aria-label="anchor" href="#sample-report-of-the-ctwas-results"></a>
</h2>
<p>We present a <a href="https://xinhe-lab.github.io/multigroup_ctwas/articles/sample_report.html" class="external-link">sample cTWAS report</a> based on real data analysis. The analyzed trait is LDL cholesterol, the prediction models are liver gene expression and splicing models trained on GTEx v8 in the PredictDB format.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
