<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing cTWAS result • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing cTWAS result">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="summarizing_ctwas_results_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing cTWAS result</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-05-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/summarizing_ctwas_results.Rmd" class="external-link"><code>vignettes/summarizing_ctwas_results.Rmd</code></a></small>
      <div class="hidden name"><code>summarizing_ctwas_results.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="summarizing-ctwas-result">Summarizing cTWAS result<a class="anchor" aria-label="anchor" href="#summarizing-ctwas-result"></a>
</h2>
<p>Load the packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
<p>Load the cTWAS results from the <code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> function</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># output directory</span></span>
<span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/tutorial/LDL_multitissue_tutorial/output"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"LDL.Liver_Adipose"</span></span>
<span></span>
<span><span class="co"># directory storing computed correlation matrices</span></span>
<span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GWAS sample size</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span>
<span></span>
<span><span class="co"># Load region info</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"region_info.RDS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># # Load z_snp</span></span>
<span><span class="co"># z_snp &lt;- readRDS(file.path(outputdir, paste0(outname, ".preprocessed.z_snp.RDS")))</span></span>
<span></span>
<span><span class="co"># Load weights</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".preprocessed.weights.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the cTWAS results </span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_sumstats_res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span></code></pre></div>
<div class="section level3">
<h3 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h3>
<p>We could use the <code><a href="../reference/summarize_param.html">summarize_param()</a></code> function to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ctwas_parameters</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_parameters.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Estimated prior inclusion probability:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##       0.0144522860       0.0021325754       0.0002151146</span></span></code></pre>
<p>Estimated prior effect size:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##          24.524181          24.524181           9.749157</span></span></code></pre>
<p>Estimated enrichment of molecular traits over variants:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">enrichment</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose </span></span>
<span><span class="co">##          67.184133           9.913673</span></span></code></pre>
<p>PVE explained by molecular traits and variants:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##         0.00905104         0.00160390         0.03927540</span></span></code></pre>
<p>Total heritability (sum of PVE)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">total_pve</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.04993034</span></span></code></pre>
<p>Attributable heritability</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">attributable_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##         0.18127336         0.03212276         0.78660388</span></span></code></pre>
<p>Plot convergence</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">convergence_plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="summarizing_ctwas_results_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="840"><p class="caption">
 
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="interpreting-ctwas-results">Interpreting cTWAS results<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-results"></a>
</h2>
<p>To interpret and visualize the results, we need to add gene names to the finemapping result.</p>
<p>We use gene annotations from the Ensembl database (EnsDb.Hsapiens.v86). Please choose the Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load gene info database</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span></code></pre></div>
<p>We then add gene names and use gene mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract gene IDs from finemapping result</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">id</span>, split <span class="op">=</span> <span class="st">"[|]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">gene_id</span>, split <span class="op">=</span> <span class="st">"[.]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="va">gene_id</span></span>
<span><span class="va">finemap_SNP_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># get gene position info</span></span>
<span><span class="va">gene_annot_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">ens_db</span>, filter <span class="op">=</span> <span class="fu">GeneIdFilter</span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">start_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">gene_annot_gr</span><span class="op">)</span></span>
<span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">end_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">gene_annot_gr</span><span class="op">)</span></span>
<span><span class="co"># use the midpoint as gene position</span></span>
<span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_mid_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="op">(</span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">start_pos</span><span class="op">+</span><span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">end_pos</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># get gene_name and gene position</span></span>
<span><span class="va">gene_annot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene_id <span class="op">=</span> <span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_id</span>, </span>
<span>                            gene_name <span class="op">=</span> <span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_name</span>, </span>
<span>                            gene_pos <span class="op">=</span> <span class="va">gene_annot_gr</span><span class="op">$</span><span class="va">gene_mid_pos</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add gene name to finemapping result and update gene positions</span></span>
<span><span class="va">gene_annot_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span>, <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span>
<span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_pos</span><span class="op">[</span><span class="va">gene_annot_idx</span><span class="op">]</span></span>
<span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="va">gene_annot_idx</span><span class="op">]</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">finemap_gene_res</span>, <span class="va">finemap_SNP_res</span><span class="op">)</span></span></code></pre></div>
<p>We list all molecular traits with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select PIP &gt; 0.8</span></span>
<span><span class="va">sig_finemap_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sig_finemap_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">sig_finemap_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>When we have multiple contexts (tissues), it is useful to evaluate the combined PIP for each molecular trait over different contexts:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># combine PIPs across contexts (tissues)</span></span>
<span><span class="va">gene_pip_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">susie_pip</span>, by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>, FUN<span class="op">=</span><span class="va">sum</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_pip_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"combined_pip"</span><span class="op">)</span></span>
<span><span class="va">gene_annot_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">gene_pip_df</span><span class="op">$</span><span class="va">gene_id</span>, <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span>
<span><span class="va">gene_pip_df</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="va">gene_annot_df</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="va">gene_annot_idx</span><span class="op">]</span></span>
<span><span class="va">gene_pip_df</span> <span class="op">&lt;-</span> <span class="va">gene_pip_df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>, <span class="st">"combined_pip"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># PIPs for each context</span></span>
<span><span class="va">contexts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">context</span><span class="op">)</span></span>
<span><span class="va">gene_pips_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="cn">NA</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gene_pip_df</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">contexts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_pips_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">contexts</span>, <span class="st">"_pip"</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gene_pips_df</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="va">gene_pip_df</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="va">finemap_gene_res_subset</span> <span class="op">&lt;-</span> <span class="va">finemap_gene_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">gene_id</span><span class="op">==</span><span class="va">gene_id</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="va">gene_contexts</span> <span class="op">&lt;-</span> <span class="va">finemap_gene_res_subset</span><span class="op">$</span><span class="va">context</span></span>
<span>  <span class="va">gene_pips_df</span><span class="op">[</span><span class="va">i</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gene_contexts</span>, <span class="st">"_pip"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">finemap_gene_res_subset</span><span class="op">$</span><span class="va">susie_pip</span></span>
<span><span class="op">}</span></span>
<span><span class="va">gene_pips_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">gene_pip_df</span>, <span class="va">gene_pips_df</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># sort by combined PIP</span></span>
<span><span class="va">gene_pips_df</span> <span class="op">&lt;-</span> <span class="va">gene_pips_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">gene_pips_df</span><span class="op">$</span><span class="va">combined_pip</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co"># select combined PIP &gt; 0.8</span></span>
<span><span class="va">sig_gene_pips_df</span> <span class="op">&lt;-</span> <span class="va">gene_pips_df</span><span class="op">[</span><span class="va">gene_pips_df</span><span class="op">$</span><span class="va">combined_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># list genes with combined PIP &gt; 0.8 or top 20 genes with highest combined PIPs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_pips_df</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sig_gene_pips_df</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              gene_id gene_name combined_pip    liver_pip  adipose_pip</span></span>
<span><span class="co">## 2785 ENSG00000153786    ZDHHC7    1.1885530 2.051826e-01 9.833704e-01</span></span>
<span><span class="co">## 1645 ENSG00000123612    ACVR1C    1.0640076 9.381139e-01 1.258937e-01</span></span>
<span><span class="co">## 2095 ENSG00000134222     PSRC1    1.0193856 1.000000e+00 1.938563e-02</span></span>
<span><span class="co">## 3595 ENSG00000168899     VAMP5    1.0007488 5.018790e-03 9.957300e-01</span></span>
<span><span class="co">## 1711 ENSG00000125629    INSIG2    1.0005182 9.999974e-01 5.207537e-04</span></span>
<span><span class="co">## 906  ENSG00000104870     FCGRT    1.0000531 6.890371e-05 9.999842e-01</span></span>
<span><span class="co">## 5355 ENSG00000261701       HPR    1.0000006 1.000000e+00 6.028600e-07</span></span>
<span><span class="co">## 1081 ENSG00000107443      CCNJ    1.0000000           NA 1.000000e+00</span></span>
<span><span class="co">## 4696 ENSG00000197114     ZGPAT    0.9996843 9.996843e-01           NA</span></span>
<span><span class="co">## 4929 ENSG00000204599    TRIM39    0.9996749 9.987271e-01 9.477794e-04</span></span>
<span><span class="co">## 2486 ENSG00000142534     RPS11    0.9984900 8.700934e-01 1.283966e-01</span></span>
<span><span class="co">## 3307 ENSG00000165029     ABCA1    0.9947563 9.947563e-01           NA</span></span>
<span><span class="co">## 2768 ENSG00000152684      PELO    0.9918423 9.139214e-01 7.792089e-02</span></span>
<span><span class="co">## 4307 ENSG00000183087      GAS6    0.9903482 9.871194e-01 3.228812e-03</span></span>
<span><span class="co">## 440  ENSG00000082701     GSK3B    0.9883018 9.883018e-01           NA</span></span>
<span><span class="co">## 762  ENSG00000100979      PLTP    0.9835695 9.821665e-01 1.402914e-03</span></span>
<span><span class="co">## 3171 ENSG00000162836      ACP6    0.9803660 9.763189e-01 4.047100e-03</span></span>
<span><span class="co">## 3181 ENSG00000163083     INHBB    0.9793823 9.793823e-01           NA</span></span>
<span><span class="co">## 1520 ENSG00000119537      KDSR    0.9782196 8.724115e-01 1.058081e-01</span></span>
<span><span class="co">## 2172 ENSG00000136271     DDX56    0.9755742 9.717791e-01 3.795073e-03</span></span>
<span><span class="co">## 2753 ENSG00000151292   CSNK1G3    0.9747330 9.697076e-01 5.025412e-03</span></span>
<span><span class="co">## 955  ENSG00000105287     PRKD2    0.9739308 9.735878e-01 3.429443e-04</span></span>
<span><span class="co">## 5307 ENSG00000255974    CYP2A6    0.9726368 9.703135e-01 2.323308e-03</span></span>
<span><span class="co">## 131  ENSG00000015520    NPC1L1    0.9715337 8.454926e-01 1.260411e-01</span></span>
<span><span class="co">## 3883 ENSG00000173757    STAT5B    0.9705069 9.681830e-01 2.323851e-03</span></span>
<span><span class="co">## 1025 ENSG00000105866       SP4    0.9698799 9.683790e-01 1.500853e-03</span></span>
<span><span class="co">## 2560 ENSG00000143771     CNIH4    0.9688407 9.643197e-01 4.520982e-03</span></span>
<span><span class="co">## 4035 ENSG00000176920      FUT2    0.9681595 9.681595e-01           NA</span></span>
<span><span class="co">## 2695 ENSG00000149485     FADS1    0.9681429 9.681429e-01           NA</span></span>
<span><span class="co">## 485  ENSG00000087087      SRRT    0.9567536 9.283432e-01 2.841042e-02</span></span>
<span><span class="co">## 2288 ENSG00000138439   FAM117B    0.9489060           NA 9.489060e-01</span></span>
<span><span class="co">## 2809 ENSG00000155158    TTC39B    0.9453617 9.424667e-01 2.894983e-03</span></span>
<span><span class="co">## 3614 ENSG00000169174     PCSK9    0.9437763           NA 9.437763e-01</span></span>
<span><span class="co">## 3861 ENSG00000173273      TNKS    0.9405702 9.405702e-01           NA</span></span>
<span><span class="co">## 1549 ENSG00000119965  C10orf88    0.9367484 8.050771e-01 1.316713e-01</span></span>
<span><span class="co">## 2170 ENSG00000136111    TBC1D4    0.9248189 7.443954e-01 1.804235e-01</span></span>
<span><span class="co">## 490  ENSG00000087237      CETP    0.9138415 4.339903e-02 8.704425e-01</span></span>
<span><span class="co">## 2293 ENSG00000138496     PARP9    0.9106863 9.106863e-01           NA</span></span>
<span><span class="co">## 3157 ENSG00000162607      USP1    0.8987965 8.923145e-01 6.482049e-03</span></span>
<span><span class="co">## 4133 ENSG00000179023   KLHDC7A    0.8967349 7.799812e-01 1.167536e-01</span></span>
<span><span class="co">## 2618 ENSG00000145390     USP53    0.8917811 8.666812e-01 2.509988e-02</span></span>
<span><span class="co">## 2754 ENSG00000151360      ALLC    0.8899287 8.899287e-01           NA</span></span>
<span><span class="co">## 4142 ENSG00000179119   SPTY2D1    0.8797328 7.652548e-01 1.144780e-01</span></span>
<span><span class="co">## 1508 ENSG00000118971     CCND2    0.8730757 7.596135e-01 1.134622e-01</span></span>
<span><span class="co">## 586  ENSG00000095485   CWF19L1    0.8725110 7.003567e-01 1.721543e-01</span></span>
<span><span class="co">## 4983 ENSG00000205978    NYNRIN    0.8697279 8.681045e-01 1.623422e-03</span></span>
<span><span class="co">## 2505 ENSG00000142765     SYTL1    0.8650402 8.626569e-01 2.383332e-03</span></span>
<span><span class="co">## 911  ENSG00000104888   SLC17A7    0.8482341 8.482290e-01 5.113841e-06</span></span>
<span><span class="co">## 3804 ENSG00000172336      POP7    0.8438463 8.438463e-01           NA</span></span>
<span><span class="co">## 392  ENSG00000076554     TPD52    0.8166550 1.205181e-02 8.046032e-01</span></span>
<span><span class="co">## 3290 ENSG00000164713      BRI3    0.8111217 8.074609e-01 3.660857e-03</span></span>
<span><span class="co">## 1204 ENSG00000110243     APOA5    0.8085898 8.085898e-01           NA</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualizing-ctwas-results">Visualizing cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-ctwas-results"></a>
</h2>
<p>We could make locus plots for regions of interest, and zoom in the region by specifying the <code>locus_range</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"1:107867043-109761309"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               cor_dir <span class="op">=</span> <span class="va">cor_dir</span>, </span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">109e6</span>, <span class="fl">109.6e6</span><span class="op">)</span>,</span>
<span>               gene_biotype<span class="op">=</span><span class="st">"protein_coding"</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-05-10 21:53:20 INFO::focus gene: PSRC1</span></span>
<span><span class="co">## 2024-05-10 21:53:20 INFO::focus id: ENSG00000134222.16|mashr_Liver</span></span>
<span><span class="co">## 2024-05-10 21:53:20 INFO::locus_range: 1.09e+08,109600000</span></span>
<span><span class="co">## 2024-05-10 21:53:21 INFO::QTL positions: 109275684</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_ctwas_results_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
<p>If we do not have precomputed correlation matrices, we could compute those by providing <code>region_info</code> and <code>region_data</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"1:107867043-109761309"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               region_info <span class="op">=</span> <span class="va">region_info</span>, </span>
<span>               region_data <span class="op">=</span> <span class="va">region_data</span>,</span>
<span>               cor_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">109e6</span>, <span class="fl">109.6e6</span><span class="op">)</span>,</span>
<span>               gene_biotype<span class="op">=</span><span class="st">"protein_coding"</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
