<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Converts PLINK genotype data to LD matrices and SNP info files, saves LD matrices as .RDS files and SNP info as .Rvar files — convert_geno_to_LD_matrix • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Converts PLINK genotype data to LD matrices and SNP info files, saves LD matrices as .RDS files and SNP info as .Rvar files — convert_geno_to_LD_matrix"><meta property="og:description" content="Converts PLINK genotype data to LD matrices and SNP info files,
saves LD matrices as .RDS files and SNP info as .Rvar files"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.38</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Converts PLINK genotype data to LD matrices and SNP info files, saves LD matrices as .RDS files and SNP info as .Rvar files</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/R/ctwas_convert_geno_to_LD_matrix.R" class="external-link"><code>R/ctwas_convert_geno_to_LD_matrix.R</code></a></small>
    <div class="hidden name"><code>convert_geno_to_LD_matrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Converts PLINK genotype data to LD matrices and SNP info files,
saves LD matrices as .RDS files and SNP info as .Rvar files</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">convert_geno_to_LD_matrix</span><span class="op">(</span></span>
<span>  <span class="va">region_info</span>,</span>
<span>  <span class="va">genotype_files</span>,</span>
<span>  <span class="va">varinfo_files</span>,</span>
<span>  chrom <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>  outputdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  outname <span class="op">=</span> <span class="st">""</span>,</span>
<span>  include_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  include_allele_freq <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_progress_bar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-region-info">region_info<a class="anchor" aria-label="anchor" href="#arg-region-info"></a></dt>
<dd><p>a data frame of region definitions, with columns:
chrom, start, stop, and region_id.</p></dd>


<dt id="arg-genotype-files">genotype_files<a class="anchor" aria-label="anchor" href="#arg-genotype-files"></a></dt>
<dd><p>Reference genotype files in PLINK binary genotype data
in .pgen or .bed format. It should contain files for all chromosomes
(from 1 to 22), one file per chromosome.</p></dd>


<dt id="arg-varinfo-files">varinfo_files<a class="anchor" aria-label="anchor" href="#arg-varinfo-files"></a></dt>
<dd><p>Reference variant information files in PLINK
.pvar or .bim format. It could have one file per chromosome or
have one big file for all chromosomes.
The output will use the genome positions in <code>varinfo_files</code>.</p></dd>


<dt id="arg-chrom">chrom<a class="anchor" aria-label="anchor" href="#arg-chrom"></a></dt>
<dd><p>a vector of chromosome numbers to process genotype data.</p></dd>


<dt id="arg-outputdir">outputdir<a class="anchor" aria-label="anchor" href="#arg-outputdir"></a></dt>
<dd><p>Output directory.</p></dd>


<dt id="arg-outname">outname<a class="anchor" aria-label="anchor" href="#arg-outname"></a></dt>
<dd><p>Output filestem.</p></dd>


<dt id="arg-include-variance">include_variance<a class="anchor" aria-label="anchor" href="#arg-include-variance"></a></dt>
<dd><p>If TRUE, include variance in .Rvar output.</p></dd>


<dt id="arg-include-allele-freq">include_allele_freq<a class="anchor" aria-label="anchor" href="#arg-include-allele-freq"></a></dt>
<dd><p>If TRUE, include allele frequency in .Rvar output.</p></dd>


<dt id="arg-show-progress-bar">show_progress_bar<a class="anchor" aria-label="anchor" href="#arg-show-progress-bar"></a></dt>
<dd><p>If TRUE, print progress bar.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>If TRUE, print detail messages.</p></dd>


<dt id="arg-logfile">logfile<a class="anchor" aria-label="anchor" href="#arg-logfile"></a></dt>
<dd><p>The log filename. If NULL, print log info on screen.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a data frame of region_metatable, with region definitions and
filenames of LD matrices and variant information.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

