<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>cTWAS analysis using summary statistics with "no LD" version — ctwas_sumstats_noLD • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="cTWAS analysis using summary statistics with " no ld version ctwas_sumstats_nold><meta property="og:description" content='cTWAS analysis using summary statistics with "no LD" version'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.19</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>cTWAS analysis using summary statistics with "no LD" version</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/R/ctwas_sumstats_noLD.R" class="external-link"><code>R/ctwas_sumstats_noLD.R</code></a></small>
    <div class="hidden name"><code>ctwas_sumstats_noLD.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>cTWAS analysis using summary statistics with "no LD" version</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ctwas_sumstats_noLD</span><span class="op">(</span></span>
<span>  <span class="va">z_snp</span>,</span>
<span>  <span class="va">weights</span>,</span>
<span>  <span class="va">region_info</span>,</span>
<span>  <span class="va">snp_map</span>,</span>
<span>  z_gene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thin <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  niter_prefit <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  niter <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  init_group_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  init_group_prior_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior_var_structure <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"shared_all"</span>, <span class="st">"shared_type"</span>, <span class="st">"shared_context"</span>,</span>
<span>    <span class="st">"shared_nonSNP"</span>, <span class="st">"independent"</span><span class="op">)</span>,</span>
<span>  maxSNP <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  min_var <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min_gene <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  min_group_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  min_nonSNP_PIP <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  min_pval <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  min_p_single_effect <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  null_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctwas"</span>, <span class="st">"susie"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  EM_tol <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  include_prior <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  include_susie_result <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outputdir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outname <span class="op">=</span> <span class="st">"ctwas_noLD"</span>,</span>
<span>  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">99</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>z_snp</dt>
<dd><p>A data frame with four columns: "id", "A1", "A2", "z".
giving the z scores for SNPs. "A1" is effect allele. "A2" is the other allele.</p></dd>


<dt>weights</dt>
<dd><p>a list of pre-processed prediction weights.</p></dd>


<dt>region_info</dt>
<dd><p>a data frame of region definitions.</p></dd>


<dt>snp_map</dt>
<dd><p>a list of data frames with SNP-to-region map for the reference.</p></dd>


<dt>z_gene</dt>
<dd><p>A data frame with columns: "id", "z", giving the z-scores for genes.</p></dd>


<dt>thin</dt>
<dd><p>The proportion of SNPs to be used for estimating parameters and screening regions.</p></dd>


<dt>niter_prefit</dt>
<dd><p>the number of iterations of the E-M algorithm to perform during the initial parameter estimation step.</p></dd>


<dt>niter</dt>
<dd><p>the maximum number of iterations of the E-M algorithm to perform during the complete parameter estimation step.</p></dd>


<dt>init_group_prior</dt>
<dd><p>a vector of initial values of prior inclusion probabilities for different groups.</p></dd>


<dt>init_group_prior_var</dt>
<dd><p>a vector of initial values of prior variances for different groups.</p></dd>


<dt>group_prior_var_structure</dt>
<dd><p>a string indicating the structure to put on the prior variance parameters.
"shared_all" allows all groups to share the same variance parameter.
"shared_type" allows all groups in one molecular QTL type to share the same variance parameter.
"shared_context" allows all groups in one context (tissue, cell type, condition) to share the same variance parameter.
"shared_nonSNP" allows all non-SNP groups to share the same variance parameter.
"independent" allows all groups to have their own separate variance parameters.
"fixed" sets prior variance parameters to values in <code>init_group_prior_var</code>.</p></dd>


<dt>maxSNP</dt>
<dd><p>Inf or integer. Maximum number of SNPs in a region. Default is
Inf, no limit. This can be useful if there are many SNPs in a region and you don't
have enough memory to run the program.</p></dd>


<dt>min_var</dt>
<dd><p>minimum number of variables (SNPs and genes) in a region
when estimating paramters and screening regions.</p></dd>


<dt>min_gene</dt>
<dd><p>minimum number of genes in a region
when estimating paramters and screening regions.</p></dd>


<dt>min_group_size</dt>
<dd><p>Minimum number of genes in a group.
Groups with number of genes &lt; <code>min_group_size</code> will be removed for the analysis.</p></dd>


<dt>min_nonSNP_PIP</dt>
<dd><p>Regions with non-SNP PIP &gt;= <code>min_nonSNP_PIP</code>
will be selected to run finemapping using full SNPs.</p></dd>


<dt>min_pval</dt>
<dd><p>Keep regions with minimum p-values from z_snp and z_gene &lt; <code>min_pval</code>.</p></dd>


<dt>min_p_single_effect</dt>
<dd><p>Regions with probability greater than <code>min_p_single_effect</code> of
having 1 or fewer effects will be used for parameter estimation.</p></dd>


<dt>null_method</dt>
<dd><p>Method to compute null model, options: "ctwas", "susie" or "none".</p></dd>


<dt>EM_tol</dt>
<dd><p>A small, non-negative number specifying the convergence
tolerance of log-likelihood for the EM iterations.</p></dd>


<dt>coverage</dt>
<dd><p>A number between 0 and 1 specifying the “coverage” of
the estimated confidence sets.</p></dd>


<dt>include_prior</dt>
<dd><p>If TRUE, include priors in finemapping results.</p></dd>


<dt>include_susie_result</dt>
<dd><p>If TRUE, include the "susie" result object in finemapping results.</p></dd>


<dt>outputdir</dt>
<dd><p>The directory to store output. If specified, save outputs to the directory.</p></dd>


<dt>outname</dt>
<dd><p>The output name.</p></dd>


<dt>ncore</dt>
<dd><p>The number of cores used to parallelize susie over regions.</p></dd>


<dt>seed</dt>
<dd><p>seed for random sampling when thinning the SNPs in region data.</p></dd>


<dt>logfile</dt>
<dd><p>The log filename. If NULL, print log info on screen.</p></dd>


<dt>verbose</dt>
<dd><p>If TRUE, print detailed messages.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments of <code>susie_rss</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list, including z_gene, estimated parameters, region_data,
cross-boundary genes, screening region results, and fine-mapping results.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

