<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Causal inference for TWAS — ctwas • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Causal inference for TWAS — ctwas"><meta property="og:description" content="Causal inference for TWAS"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.38</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Causal inference for TWAS</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/R/ctwas.R" class="external-link"><code>R/ctwas.R</code></a></small>
    <div class="hidden name"><code>ctwas.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Causal inference for TWAS</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ctwas</span><span class="op">(</span></span>
<span>  <span class="va">pgenfs</span>,</span>
<span>  <span class="va">exprfs</span>,</span>
<span>  <span class="va">Y</span>,</span>
<span>  ld_regions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR"</span>, <span class="st">"ASN"</span>, <span class="st">"AFR"</span><span class="op">)</span>,</span>
<span>  ld_regions_version <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"b37"</span>, <span class="st">"b38"</span><span class="op">)</span>,</span>
<span>  ld_regions_custom <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thin <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  prob_single <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  max_snp_region <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  rerun_gene_PIP <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  niter1 <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  niter2 <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  group_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  estimate_group_prior <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  estimate_group_prior_var <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  use_null_weight <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  outputdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  outname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-pgenfs">pgenfs<a class="anchor" aria-label="anchor" href="#arg-pgenfs"></a></dt>
<dd><p>A character vector of .pgen or .bed files. One file for one
chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22. If .pgen files are given, then .pvar and .psam are assumed
to present in the same directory. If .bed files are given, then .bim and
.fam files are assumed to present in the same directory.</p></dd>


<dt id="arg-exprfs">exprfs<a class="anchor" aria-label="anchor" href="#arg-exprfs"></a></dt>
<dd><p>A character vector of .`expr` or `.expr.gz` files. One file for
one chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22.  `.expr.gz` file is gzip compressed `.expr` files. `.expr` is
a matrix of imputed expression values, row is for each sample, column is for
each gene. Its sample order is same as in files provided by `.pgenfs`. We also
assume corresponding `.exprvar` files are present in the same directory.
`.exprvar` files are just tab delimited text files, with columns:</p><dl><dt>chrom</dt>
<dd><p>chromosome number, numeric</p></dd>

  <dt>p0</dt>
<dd><p>gene boundary position, the smaller value</p></dd>

  <dt>p1</dt>
<dd><p>gene boundary position, the larger value</p></dd>

  <dt>id</dt>
<dd><p>gene id</p></dd>


</dl><p>Its rows should be in the same order as the columns for corresponding `.expr`
files.</p></dd>


<dt id="arg-y">Y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>a vector of length n, phenotype, the same order as provided
by `.pgenfs` (defined in .psam or .fam files).</p></dd>


<dt id="arg-ld-regions">ld_regions<a class="anchor" aria-label="anchor" href="#arg-ld-regions"></a></dt>
<dd><p>A string representing the population to use for defining
LD regions. These LD regions were previously defined by ldetect. The user can also
provide custom LD regions matching genotype data, see
<code>ld_regions_custom</code>.</p></dd>


<dt id="arg-ld-regions-version">ld_regions_version<a class="anchor" aria-label="anchor" href="#arg-ld-regions-version"></a></dt>
<dd><p>A string representing the genome reference build ("b37", "b38") to use for defining
LD regions. See <code>ld_regions</code>.</p></dd>


<dt id="arg-ld-regions-custom">ld_regions_custom<a class="anchor" aria-label="anchor" href="#arg-ld-regions-custom"></a></dt>
<dd><p>A bed format file defining LD regions. The default
is <code>NULL</code>; when specified, <code>ld_regions</code> and <code>ld_regions_version</code> will be ignored.</p></dd>


<dt id="arg-thin">thin<a class="anchor" aria-label="anchor" href="#arg-thin"></a></dt>
<dd><p>The proportion of SNPs to be used for the parameter estimation and initial fine
mapping steps. Smaller <code>thin</code> parameters reduce runtime at the expense of accuracy. The fine mapping step is rerun using full SNPs
for regions with strong gene signals; see <code>rerun_gene_PIP</code>.</p></dd>


<dt id="arg-prob-single">prob_single<a class="anchor" aria-label="anchor" href="#arg-prob-single"></a></dt>
<dd><p>Blocks with probability greater than <code>prob_single</code> of having 1 or fewer effects will be
used for parameter estimation</p></dd>


<dt id="arg-max-snp-region">max_snp_region<a class="anchor" aria-label="anchor" href="#arg-max-snp-region"></a></dt>
<dd><p>Inf or integer. Maximum number of SNPs in a region. Default is
Inf, no limit. This can be useful if there are many SNPs in a region and you don't
have enough memory to run the program. This applies to the last rerun step
(using full SNPs and rerun susie for regions with strong gene signals) only.</p></dd>


<dt id="arg-rerun-gene-pip">rerun_gene_PIP<a class="anchor" aria-label="anchor" href="#arg-rerun-gene-pip"></a></dt>
<dd><p>if thin &lt;1, will rerun blocks with the max gene PIP
&gt; <code>rerun_gene_PIP</code> using full SNPs. if <code>rerun_gene_PIP</code> is 0, then
all blocks will rerun with full SNPs</p></dd>


<dt id="arg-niter-">niter1<a class="anchor" aria-label="anchor" href="#arg-niter-"></a></dt>
<dd><p>the number of iterations of the E-M algorithm to perform during the initial parameter estimation step</p></dd>


<dt id="arg-niter-">niter2<a class="anchor" aria-label="anchor" href="#arg-niter-"></a></dt>
<dd><p>the number of iterations of the E-M algorithm to perform during the complete parameter estimation step</p></dd>


<dt id="arg-l">L<a class="anchor" aria-label="anchor" href="#arg-l"></a></dt>
<dd><p>the number of effects for susie during the fine mapping steps</p></dd>


<dt id="arg-group-prior">group_prior<a class="anchor" aria-label="anchor" href="#arg-group-prior"></a></dt>
<dd><p>a vector of two prior inclusion probabilities for SNPs and genes. This is ignored
if <code>estimate_group_prior = T</code></p></dd>


<dt id="arg-group-prior-var">group_prior_var<a class="anchor" aria-label="anchor" href="#arg-group-prior-var"></a></dt>
<dd><p>a vector of two prior variances for SNPs and gene effects. This is ignored
if <code>estimate_group_prior_var = T</code></p></dd>


<dt id="arg-estimate-group-prior">estimate_group_prior<a class="anchor" aria-label="anchor" href="#arg-estimate-group-prior"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, the prior inclusion probabilities for SNPs and genes are estimated
using the data. If FALSE, <code>group_prior</code> must be specified</p></dd>


<dt id="arg-estimate-group-prior-var">estimate_group_prior_var<a class="anchor" aria-label="anchor" href="#arg-estimate-group-prior-var"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, the prior variances for SNPs and genes are estimated
using the data. If FALSE, <code>group_prior_var</code> must be specified</p></dd>


<dt id="arg-use-null-weight">use_null_weight<a class="anchor" aria-label="anchor" href="#arg-use-null-weight"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, allow for a probability of no effect in susie</p></dd>


<dt id="arg-coverage">coverage<a class="anchor" aria-label="anchor" href="#arg-coverage"></a></dt>
<dd><p>A number between 0 and 1 specifying the “coverage” of the estimated confidence sets</p></dd>


<dt id="arg-standardize">standardize<a class="anchor" aria-label="anchor" href="#arg-standardize"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, all variables are standardized to unit variance</p></dd>


<dt id="arg-ncore">ncore<a class="anchor" aria-label="anchor" href="#arg-ncore"></a></dt>
<dd><p>The number of cores used to parallelize susie over regions</p></dd>


<dt id="arg-outputdir">outputdir<a class="anchor" aria-label="anchor" href="#arg-outputdir"></a></dt>
<dd><p>a string, the directory to store output</p></dd>


<dt id="arg-outname">outname<a class="anchor" aria-label="anchor" href="#arg-outname"></a></dt>
<dd><p>a string, the output name</p></dd>


<dt id="arg-logfile">logfile<a class="anchor" aria-label="anchor" href="#arg-logfile"></a></dt>
<dd><p>the log file, if NULL will print log info on screen</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

