<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Harmonize z scores from GWAS to match ld reference genotypes. Flip signs when reverse complement matches, remove/recover strand ambiguous SNPs — harmonize_wgt_ld • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Harmonize z scores from GWAS to match ld reference genotypes. Flip signs when reverse complement matches, remove/recover strand ambiguous SNPs — harmonize_wgt_ld"><meta property="og:description" content="Harmonize z scores from GWAS to match ld reference genotypes.
Flip signs when reverse complement matches, remove/recover strand ambiguous SNPs"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.38</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Harmonize z scores from GWAS to match ld reference genotypes. Flip signs when reverse complement matches, remove/recover strand ambiguous SNPs</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/R/ctwas_harmonize_data.R" class="external-link"><code>R/ctwas_harmonize_data.R</code></a></small>
    <div class="hidden name"><code>harmonize_wgt_ld.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Harmonize z scores from GWAS to match ld reference genotypes.
Flip signs when reverse complement matches, remove/recover strand ambiguous SNPs</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">harmonize_wgt_ld</span><span class="op">(</span></span>
<span>  <span class="va">wgt.matrix</span>,</span>
<span>  <span class="va">snps</span>,</span>
<span>  <span class="va">ld_snpinfo</span>,</span>
<span>  recover_strand_ambig <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  ld_pgenfs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ld_Rinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_wgt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  wgt <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-wgt-matrix">wgt.matrix<a class="anchor" aria-label="anchor" href="#arg-wgt-matrix"></a></dt>
<dd><p>from FUSION weight .Rdat file</p></dd>


<dt id="arg-snps">snps<a class="anchor" aria-label="anchor" href="#arg-snps"></a></dt>
<dd><p>from FUSION weight .Rdat file
with columns "chrom", "id", "pos", "alt", "ref". The effect allele
for FUSION is alt.</p></dd>


<dt id="arg-ld-snpinfo">ld_snpinfo<a class="anchor" aria-label="anchor" href="#arg-ld-snpinfo"></a></dt>
<dd><p>a data frame, snp info for LD reference,
with columns "chrom", "id", "pos", "alt", "ref".</p></dd>


<dt id="arg-recover-strand-ambig">recover_strand_ambig<a class="anchor" aria-label="anchor" href="#arg-recover-strand-ambig"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, a procedure is used to recover strand ambiguous variants. If FALSE,
these variants are dropped from the prediction model. This procedure compares correlations between variants in the the
LD reference and prediction models, and it can only be used with PredictDB format prediction models, which include this
information.</p></dd>


<dt id="arg-ld-pgenfs">ld_pgenfs<a class="anchor" aria-label="anchor" href="#arg-ld-pgenfs"></a></dt>
<dd><p>a character vector of .pgen or .bed files. One file for one
chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22. If .pgen files are given, then .pvar and .psam are assumed
to present in the same directory. If .bed files are given, then .bim and
.fam files are assumed to present in the same directory.</p></dd>


<dt id="arg-ld-rinfo">ld_Rinfo<a class="anchor" aria-label="anchor" href="#arg-ld-rinfo"></a></dt>
<dd><p>a vector of paths to the variant information for all LD matrices</p></dd>


<dt id="arg-r-wgt">R_wgt<a class="anchor" aria-label="anchor" href="#arg-r-wgt"></a></dt>
<dd><p>the LD matrix for the variants in <code>wgt.matrix</code></p></dd>


<dt id="arg-wgt">wgt<a class="anchor" aria-label="anchor" href="#arg-wgt"></a></dt>
<dd><p>allele information from the weights</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>wgt.matrix and snps with alleles flipped to match</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

