<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get gene and SNP index for each region — index_regions • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get gene and SNP index for each region — index_regions"><meta property="og:description" content="For each region, get the index for snp and gene (index
is location/column number in .pgen file or .expr file) located within
this region."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.38</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get gene and SNP index for each region</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/R/ctwas_process_regions.R" class="external-link"><code>R/ctwas_process_regions.R</code></a></small>
    <div class="hidden name"><code>index_regions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For each region, get the index for snp and gene (index
is location/column number in .pgen file or .expr file) located within
this region.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">index_regions</span><span class="op">(</span></span>
<span>  <span class="va">regionfile</span>,</span>
<span>  <span class="va">exprvarfs</span>,</span>
<span>  pvarfs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ld_Rfs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  select <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thin <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  maxSNP <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  minvar <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  merge <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  outname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outputdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-regionfile">regionfile<a class="anchor" aria-label="anchor" href="#arg-regionfile"></a></dt>
<dd><p>regions file. Has three columns: chr, start, end. The regions file
should provide non overlaping regions defining LD blocks. currently does not support
chromsome X/Y etc.</p></dd>


<dt id="arg-select">select<a class="anchor" aria-label="anchor" href="#arg-select"></a></dt>
<dd><p>Default is NULL, all variants will be selected. Or a vector of variant IDs,
or a data frame with columns id and z (id is for gene or SNP id, z is for z scores).
z will be used for remove SNPs if the total number of SNPs exceeds limit. See
parameter `maxSNP` for more information.</p></dd>


<dt id="arg-thin">thin<a class="anchor" aria-label="anchor" href="#arg-thin"></a></dt>
<dd><p>A scalar in (0,1]. The proportion of SNPs
left after down sampling. Only applied on SNPs after selecting variants.</p></dd>


<dt id="arg-maxsnp">maxSNP<a class="anchor" aria-label="anchor" href="#arg-maxsnp"></a></dt>
<dd><p>Default is Inf, no limit for the maximum number of SNPs in a region. Or an
integer indicating the maximum number of SNPs allowed in a region. This
parameter is useful when a region contains many SNPs and you don't have enough memory to
run the program. In this case, you can put a limit on the number of SNPs in the region.
If z scores are given in the parameter `select`, i.e. a data frame with columns id and z is
provided, SNPs are ranked based on |z| from high to low and only the top `maxSNP` SNPs
are kept. If only variant ids are provided, then `maxSNP` number of SNPs will be chosen
randomly.</p></dd>


<dt id="arg-minvar">minvar<a class="anchor" aria-label="anchor" href="#arg-minvar"></a></dt>
<dd><p>minimum number of variatns in a region</p></dd>


<dt id="arg-merge">merge<a class="anchor" aria-label="anchor" href="#arg-merge"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, merge regions when a gene spans a region boundary (i.e. belongs to multiple regions.)</p></dd>


<dt id="arg-outname">outname<a class="anchor" aria-label="anchor" href="#arg-outname"></a></dt>
<dd><p>a string, the output name</p></dd>


<dt id="arg-outputdir">outputdir<a class="anchor" aria-label="anchor" href="#arg-outputdir"></a></dt>
<dd><p>a string, the directory to store output</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list. Items correspond to each pvarf/exprvarf. Each Item is
 also a list, the items in this list are for each region.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

