<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.25</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/simple_ctwas_tutorial.html">Simple cTWAS tutorial</a>
    </li>
    <li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="N2finemapping.html">N2finemapping</a></code> </p>
        </td>
        <td><p>Simulated Fine-mapping Data with Two Effect Variables</p></td>
      </tr><tr><td>
          <p><code><a href="anno_finemap_res.html">anno_finemap_res()</a></code> </p>
        </td>
        <td><p>Annotates cTWAS finemapping result with ensembl gene annotation database</p></td>
      </tr><tr><td>
          <p><code><a href="assemble_region_data.html">assemble_region_data()</a></code> </p>
        </td>
        <td><p>Assembles data for all the regions</p></td>
      </tr><tr><td>
          <p><code><a href="coef.susie.html">coef(<i>&lt;susie&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extract regression coefficients from susie fit</p></td>
      </tr><tr><td>
          <p><code><a href="compute_gene_z.html">compute_gene_z()</a></code> </p>
        </td>
        <td><p>Computes z-scores of molecular traits</p></td>
      </tr><tr><td>
          <p><code><a href="compute_region_nonSNP_PIPs.html">compute_region_nonSNP_PIPs()</a></code> </p>
        </td>
        <td><p>Computes non-SNP PIPs for regions</p></td>
      </tr><tr><td>
          <p><code><a href="compute_ss.html">compute_ss()</a></code> </p>
        </td>
        <td><p>Compute sufficient statistics for susie.</p></td>
      </tr><tr><td>
          <p><code><a href="convert_geno_to_LD_matrix.html">convert_geno_to_LD_matrix()</a></code> </p>
        </td>
        <td><p>Converts PLINK genotype data to LD matrices and SNP info files, saves LD matrices as .RDS files and SNP info as .Rvar files</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_sumstats.html">ctwas_sumstats()</a></code> </p>
        </td>
        <td><p>Causal inference for TWAS using summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_sumstats_noLD.html">ctwas_sumstats_noLD()</a></code> </p>
        </td>
        <td><p>Runs cTWAS analysis with "no LD" version</p></td>
      </tr><tr><td>
          <p><code><a href="diagnose_ld_mismatch_susie.html">diagnose_ld_mismatch_susie()</a></code> </p>
        </td>
        <td><p>Diagnose LD mismatches using SuSiE RSS</p></td>
      </tr><tr><td>
          <p><code><a href="est_param.html">est_param()</a></code> </p>
        </td>
        <td><p>Estimates cTWAS parameters using EM</p></td>
      </tr><tr><td>
          <p><code><a href="estimate_s_rss.html">estimate_s_rss()</a></code> </p>
        </td>
        <td><p>Estimate s in <code>susie_rss</code> Model Using Regularized LD</p></td>
      </tr><tr><td>
          <p><code><a href="expand_region_data.html">expand_region_data()</a></code> </p>
        </td>
        <td><p>Expands region_data with full SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_region.html">finemap_region()</a></code> </p>
        </td>
        <td><p>Runs cTWAS finemapping for a single region</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_regions.html">finemap_regions()</a></code> </p>
        </td>
        <td><p>Runs cTWAS finemapping for multiple regions</p></td>
      </tr><tr><td>
          <p><code><a href="get_problematic_genes.html">get_problematic_genes()</a></code> </p>
        </td>
        <td><p>Gets problematic genes from problematic SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="kriging_rss.html">kriging_rss()</a></code> </p>
        </td>
        <td><p>Compute Distribution of z-scores of Variant j Given Other z-scores, and Detect Possible Allele Switch Issue</p></td>
      </tr><tr><td>
          <p><code><a href="load_weights.html">load_weights()</a></code> </p>
        </td>
        <td><p>Loads PredictDB or FUSION weights</p></td>
      </tr><tr><td>
          <p><code><a href="make_locusplot.html">make_locusplot()</a></code> </p>
        </td>
        <td><p>Plots the cTWAS result for a single locus</p></td>
      </tr><tr><td>
          <p><code><a href="map_snp_info_regions.html">map_snp_info_regions()</a></code> </p>
        </td>
        <td><p>Reads SNP info from all regions and map SNPs to regions.</p></td>
      </tr><tr><td>
          <p><code><a href="merge_finemap_regions.html">merge_finemap_regions()</a></code> </p>
        </td>
        <td><p>Merges regions with boundary genes and finemaps merged regions</p></td>
      </tr><tr><td>
          <p><code><a href="merge_region_data.html">merge_region_data()</a></code> </p>
        </td>
        <td><p>Merges region data for cross-boundary genes</p></td>
      </tr><tr><td>
          <p><code><a href="predict.susie.html">predict(<i>&lt;susie&gt;</i>)</a></code> </p>
        </td>
        <td><p>Predict outcomes or extract coefficients from susie fit.</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_region_LD_snp_info.html">preprocess_region_LD_snp_info()</a></code> </p>
        </td>
        <td><p>Preprocess region info, LD info, and SNP info from all regions and map SNPs to regions.</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_weights.html">preprocess_weights()</a></code> </p>
        </td>
        <td><p>Preprocess PredictDB/FUSION weights and harmonize with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_z_snp.html">preprocess_z_snp()</a></code> </p>
        </td>
        <td><p>Preprocess GWAS z-scores, harmonize GWAS z-scores with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="screen_regions.html">screen_regions()</a></code> </p>
        </td>
        <td><p>Screens regions with strong non-SNP PIPs to be used in finemapping step</p></td>
      </tr><tr><td>
          <p><code><a href="sum_pip_across_contexts.html">sum_pip_across_contexts()</a></code> </p>
        </td>
        <td><p>Combines PIPs across contexts (tissues)</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_param.html">summarize_param()</a></code> </p>
        </td>
        <td><p>Summarizes and plots cTWAS parameter estimates</p></td>
      </tr><tr><td>
          <p><code><a href="susie.html">susie()</a></code> <code><a href="susie.html">susie_suff_stat()</a></code> </p>
        </td>
        <td><p>Sum of Single Effects (SuSiE) Regression</p></td>
      </tr><tr><td>
          <p><code><a href="susie_get_methods.html">susie_get_objective()</a></code> <code><a href="susie_get_methods.html">susie_get_posterior_mean()</a></code> <code><a href="susie_get_methods.html">susie_get_posterior_sd()</a></code> <code><a href="susie_get_methods.html">susie_get_niter()</a></code> <code><a href="susie_get_methods.html">susie_get_prior_variance()</a></code> <code><a href="susie_get_methods.html">susie_get_residual_variance()</a></code> <code><a href="susie_get_methods.html">susie_get_lfsr()</a></code> <code><a href="susie_get_methods.html">susie_get_posterior_samples()</a></code> <code><a href="susie_get_methods.html">susie_get_cs()</a></code> <code><a href="susie_get_methods.html">susie_get_pip()</a></code> </p>
        </td>
        <td><p>Inferences From Fitted SuSiE Model</p></td>
      </tr><tr><td>
          <p><code><a href="susie_init_coef.html">susie_init_coef()</a></code> </p>
        </td>
        <td><p>Initialize a susie object using regression coefficients</p></td>
      </tr><tr><td>
          <p><code><a href="susie_rss.html">susie_rss()</a></code> </p>
        </td>
        <td><p>Sum of Single Effects (SuSiE) Regression using summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="univariate_regression.html">univariate_regression()</a></code> </p>
        </td>
        <td><p>Perform Univariate Linear Regression Separately for Columns of X</p></td>
      </tr><tr><td>
          <p><code><a href="update_finemap_res.html">update_finemap_res()</a></code> </p>
        </td>
        <td><p>Updates finemapping result</p></td>
      </tr><tr><td>
          <p><code><a href="update_merged_regions_finemap_res.html">update_merged_regions_finemap_res()</a></code> </p>
        </td>
        <td><p>Updates finemapping result for merged regions</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

