<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.25</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_pos.html">add_pos()</a></code> </p>
        </td>
        <td><p>Add SNP and gene positions to cTWAS finemapping result.</p></td>
      </tr><tr><td>
          <p><code><a href="anno_finemap_res.html">anno_finemap_res()</a></code> </p>
        </td>
        <td><p>Map finemapping result of molecular traits to genes.</p></td>
      </tr><tr><td>
          <p><code><a href="anno_susie_alpha_res.html">anno_susie_alpha_res()</a></code> </p>
        </td>
        <td><p>Map molecular traits to genes in susie alpha result.</p></td>
      </tr><tr><td>
          <p><code><a href="assemble_region_data.html">assemble_region_data()</a></code> </p>
        </td>
        <td><p>Assembles data for all the regions</p></td>
      </tr><tr><td>
          <p><code><a href="combine_gene_pips.html">combine_gene_pips()</a></code> </p>
        </td>
        <td><p>Combines gene PIPs by context, type or group.</p></td>
      </tr><tr><td>
          <p><code><a href="compute_enrichment_test.html">compute_enrichment_test()</a></code> </p>
        </td>
        <td><p>Computes enrichment (log-scale), standard error and p-value</p></td>
      </tr><tr><td>
          <p><code><a href="compute_gene_z.html">compute_gene_z()</a></code> </p>
        </td>
        <td><p>Computes z-scores of molecular traits</p></td>
      </tr><tr><td>
          <p><code><a href="compute_region_cor.html">compute_region_cor()</a></code> </p>
        </td>
        <td><p>Computes correlation matrices for a single region</p></td>
      </tr><tr><td>
          <p><code><a href="compute_region_nonSNP_PIPs.html">compute_region_nonSNP_PIPs()</a></code> </p>
        </td>
        <td><p>Computes non-SNP PIPs for all regions from finemapping result</p></td>
      </tr><tr><td>
          <p><code><a href="compute_weight_LD_from_ref.html">compute_weight_LD_from_ref()</a></code> </p>
        </td>
        <td><p>Computes LD for weight variants using reference LD</p></td>
      </tr><tr><td>
          <p><code><a href="convert_fusion_to_predictdb.html">convert_fusion_to_predictdb()</a></code> </p>
        </td>
        <td><p>Converts fusion weights to predictDB format</p></td>
      </tr><tr><td>
          <p><code><a href="convert_geno_to_LD_matrix.html">convert_geno_to_LD_matrix()</a></code> </p>
        </td>
        <td><p>Converts PLINK genotype data to LD matrices and SNP info files,
saves LD matrices as .RDS files and SNP info as .Rvar files</p></td>
      </tr><tr><td>
          <p><code><a href="convert_to_ukb_varIDs.html">convert_to_ukb_varIDs()</a></code> </p>
        </td>
        <td><p>Convert variant IDs from Open GWAS format or PredictDB weight format</p></td>
      </tr><tr><td>
          <p><code><a href="count_n_snps.html">count_n_snps()</a></code> </p>
        </td>
        <td><p>Count the numbers of SNPs in snp_map, z_snp and weights</p></td>
      </tr><tr><td>
          <p><code><a href="create_predictdb_from_QTLs.html">create_predictdb_from_QTLs()</a></code> </p>
        </td>
        <td><p>Creates weight files in PredictDB format from QTL data</p></td>
      </tr><tr><td>
          <p><code><a href="create_snp_LD_map.html">create_snp_LD_map()</a></code> </p>
        </td>
        <td><p>Map SNPs to regions using region meta table.</p></td>
      </tr><tr><td>
          <p><code><a href="create_snp_map.html">create_snp_map()</a></code> </p>
        </td>
        <td><p>Map SNPs to regions using all the variants in the LD reference.</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_sumstats.html">ctwas_sumstats()</a></code> </p>
        </td>
        <td><p>cTWAS analysis using summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_sumstats_noLD.html">ctwas_sumstats_noLD()</a></code> </p>
        </td>
        <td><p>cTWAS analysis using summary statistics with "no LD" version</p></td>
      </tr><tr><td>
          <p><code><a href="diagnose_LD_mismatch_susie.html">diagnose_LD_mismatch_susie()</a></code> </p>
        </td>
        <td><p>Diagnose LD mismatch using SuSiE RSS</p></td>
      </tr><tr><td>
          <p><code><a href="est_param.html">est_param()</a></code> </p>
        </td>
        <td><p>Estimates cTWAS parameters using EM with cTWAS SER model</p></td>
      </tr><tr><td>
          <p><code><a href="estimate_region_L.html">estimate_region_L()</a></code> </p>
        </td>
        <td><p>Estimate L for all regions by running finemapping with uniform prior</p></td>
      </tr><tr><td>
          <p><code><a href="expand_region_data.html">expand_region_data()</a></code> </p>
        </td>
        <td><p>Expands region_data with all SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="filter_z_gene_by_group_size.html">filter_z_gene_by_group_size()</a></code> </p>
        </td>
        <td><p>Filter z_gene by group size</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_regions.html">finemap_regions()</a></code> </p>
        </td>
        <td><p>Runs cTWAS fine-mapping for regions</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_regions_noLD.html">finemap_regions_noLD()</a></code> </p>
        </td>
        <td><p>Runs cTWAS fine-mapping for regions without LD (L = 1)</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db()</a></code> </p>
        </td>
        <td><p>Get gene annotation table from Ensembl database</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_info.html">get_gene_info()</a></code> </p>
        </td>
        <td><p>Get gene info from weights</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_regions.html">get_gene_regions()</a></code> </p>
        </td>
        <td><p>Get regions for each gene</p></td>
      </tr><tr><td>
          <p><code><a href="get_molecular_ids.html">get_molecular_ids()</a></code> </p>
        </td>
        <td><p>Get original molecular IDs</p></td>
      </tr><tr><td>
          <p><code><a href="get_predictdb_genome_build.html">get_predictdb_genome_build()</a></code> </p>
        </td>
        <td><p>Get genome build of PredictDB weight</p></td>
      </tr><tr><td>
          <p><code><a href="get_problematic_genes.html">get_problematic_genes()</a></code> </p>
        </td>
        <td><p>Gets problematic genes from problematic SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="get_region_cor.html">get_region_cor()</a></code> </p>
        </td>
        <td><p>Gets correlation matrices for a single region.</p></td>
      </tr><tr><td>
          <p><code><a href="load_LD.html">load_LD()</a></code> </p>
        </td>
        <td><p>Load LD matrix</p></td>
      </tr><tr><td>
          <p><code><a href="load_fusion_weights.html">load_fusion_weights()</a></code> </p>
        </td>
        <td><p>Loads weights in FUSION format</p></td>
      </tr><tr><td>
          <p><code><a href="load_predictdb_weights.html">load_predictdb_weights()</a></code> </p>
        </td>
        <td><p>Loads weights in PredictDB format</p></td>
      </tr><tr><td>
          <p><code><a href="load_region_cor.html">load_region_cor()</a></code> </p>
        </td>
        <td><p>Loads precomputed correlation matrices for a single region.</p></td>
      </tr><tr><td>
          <p><code><a href="load_weights.html">load_weights()</a></code> </p>
        </td>
        <td><p>Loads weights in PredictDB or FUSION format</p></td>
      </tr><tr><td>
          <p><code><a href="make_convergence_plots.html">make_convergence_plots()</a></code> </p>
        </td>
        <td><p>Make convergence plots for the estimated parameters</p></td>
      </tr><tr><td>
          <p><code><a href="make_locusplot.html">make_locusplot()</a></code> </p>
        </td>
        <td><p>Plots the cTWAS result for a single locus</p></td>
      </tr><tr><td>
          <p><code><a href="merge_region_data.html">merge_region_data()</a></code> </p>
        </td>
        <td><p>Merges region data for cross-boundary genes</p></td>
      </tr><tr><td>
          <p><code><a href="merge_region_data_noLD.html">merge_region_data_noLD()</a></code> </p>
        </td>
        <td><p>Merges region data for cross-boundary genes without using LD</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_weights.html">preprocess_weights()</a></code> </p>
        </td>
        <td><p>Preprocess PredictDB/FUSION weights and harmonize with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_z_snp.html">preprocess_z_snp()</a></code> </p>
        </td>
        <td><p>Preprocess GWAS z-scores, harmonize GWAS z-scores with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="read_exprvar.html">read_exprvar()</a></code> </p>
        </td>
        <td><p>Read .exprvar file into R</p></td>
      </tr><tr><td>
          <p><code><a href="read_gwas.html">read_gwas()</a></code> </p>
        </td>
        <td><p>Read GWAS summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="read_snp_info_file.html">read_snp_info_file()</a></code> </p>
        </td>
        <td><p>Read a single SNP info file as a data frame</p></td>
      </tr><tr><td>
          <p><code><a href="read_snp_info_files.html">read_snp_info_files()</a></code> </p>
        </td>
        <td><p>Read multiple single SNP info files as a data frame</p></td>
      </tr><tr><td>
          <p><code><a href="screen_regions.html">screen_regions()</a></code> </p>
        </td>
        <td><p>Screens regions with strong signals</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_param.html">summarize_param()</a></code> </p>
        </td>
        <td><p>Summarizes estimated parameters</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_region_signals.html">summarize_region_signals()</a></code> </p>
        </td>
        <td><p>Gets a basic summary of region signals</p></td>
      </tr><tr><td>
          <p><code><a href="trim_weights.html">trim_weights()</a></code> </p>
        </td>
        <td><p>Trim weights by keeping top number of SNPs and/or filtering by minimum absolute weights.</p></td>
      </tr><tr><td>
          <p><code><a href="update_finemap_res.html">update_finemap_res()</a></code> </p>
        </td>
        <td><p>Updates cTWAS finemapping result for selected regions</p></td>
      </tr><tr><td>
          <p><code><a href="update_merged_region_data.html">update_merged_region_data()</a></code> </p>
        </td>
        <td><p>Updates cTWAS input data with merged region data</p></td>
      </tr><tr><td>
          <p><code><a href="update_merged_region_data_noLD.html">update_merged_region_data_noLD()</a></code> </p>
        </td>
        <td><p>Updates cTWAS input data without LD with merged region data</p></td>
      </tr><tr><td>
          <p><code><a href="update_merged_region_finemap_res.html">update_merged_region_finemap_res()</a></code> </p>
        </td>
        <td><p>Updates cTWAS finemapping result for merged regions</p></td>
      </tr><tr><td>
          <p><code><a href="update_region_z.html">update_region_z()</a></code> </p>
        </td>
        <td><p>Adds or updates z-scores in region_data based on z_snp and z_gene.
this will also update sid and gid based on z_snp and z_gene.</p></td>
      </tr><tr><td>
          <p><code><a href="z2p.html">z2p()</a></code> </p>
        </td>
        <td><p>convert z-scores to p-values (two-sided test)</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

