<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plots the cTWAS result for a single locus — make_locusplot • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plots the cTWAS result for a single locus — make_locusplot"><meta property="og:description" content="Plots the cTWAS result for a single locus"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.19</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plots the cTWAS result for a single locus</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/R/ctwas_plots.R" class="external-link"><code>R/ctwas_plots.R</code></a></small>
    <div class="hidden name"><code>make_locusplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots the cTWAS result for a single locus</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">make_locusplot</span><span class="op">(</span></span>
<span>  <span class="va">finemap_res</span>,</span>
<span>  <span class="va">region_id</span>,</span>
<span>  <span class="va">ens_db</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_snp_gene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_gene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  locus_range <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  focal_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  focal_gene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_cs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  color_pval_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cs"</span>, <span class="st">"LD"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  color_pip_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cs"</span>, <span class="st">"LD"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  LD.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"blue"</span>, <span class="st">"purple"</span>, <span class="st">"salmon"</span><span class="op">)</span>,</span>
<span>  cs.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firebrick"</span>, <span class="st">"dodgerblue"</span>, <span class="st">"forestgreen"</span>, <span class="st">"darkmagenta"</span>, <span class="st">"darkorange"</span><span class="op">)</span>,</span>
<span>  focal.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"salmon"</span><span class="op">)</span>,</span>
<span>  label_QTLs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlight_pval <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlight_pip <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  highlight_pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlight.color <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  highlight.pos.color <span class="op">=</span> <span class="st">"cyan2"</span>,</span>
<span>  point.sizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  point.alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.6</span><span class="op">)</span>,</span>
<span>  point.shapes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">15</span>, <span class="fl">18</span>, <span class="fl">17</span>, <span class="fl">10</span>, <span class="fl">12</span>, <span class="fl">14</span>, <span class="fl">11</span><span class="op">)</span>,</span>
<span>  label.text.size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>  max.overlaps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  axis.text.size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  axis.title.size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legend.text.size <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>  genelabel.cex.text <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  panel.heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>finemap_res</dt>
<dd><p>a data frame of cTWAS finemapping result</p></dd>


<dt>region_id</dt>
<dd><p>region ID to be plotted</p></dd>


<dt>ens_db</dt>
<dd><p>Ensembl database</p></dd>


<dt>weights</dt>
<dd><p>a list of proprocessed weights, used to plot QTL track.
If NULL, will not plot the QTL track.</p></dd>


<dt>R_snp_gene</dt>
<dd><p>SNP-gene correlation matrix of the region.
If both R_snp_gene and R_gene are available,
color data points with correlations with the focal gene.</p></dd>


<dt>R_gene</dt>
<dd><p>gene-gene correlation matrix of the region</p></dd>


<dt>locus_range</dt>
<dd><p>a vector of start and end positions to define the region boundary to be plotted.
If NULL, the entire region will be plotted (with 100 bp flanks on both sides).</p></dd>


<dt>focal_id</dt>
<dd><p>the focal ID.</p></dd>


<dt>focal_gene</dt>
<dd><p>the focal gene name. By default, use the gene with the highest PIP.</p></dd>


<dt>filter_protein_coding_genes</dt>
<dd><p>If TRUE, limits to protein coding genes only.</p></dd>


<dt>filter_cs</dt>
<dd><p>If TRUE, limits to credible sets.</p></dd>


<dt>color_pval_by</dt>
<dd><p>Options to color the p-value track.
"LD": colors the p-value track by the correlations to the focal gene.
"cs": colors the p-value track by the credible sets.
"none": uses the same color for non-focal genes.</p></dd>


<dt>color_pip_by</dt>
<dd><p>Options to color the PIP track.
"LD": colors the PIP track by the correlations to the focal gene.
"cs": colors the PIP track by the credible sets.
"none": uses the same color for non-focal genes.</p></dd>


<dt>LD.colors</dt>
<dd><p>Colors for correlation levels.</p></dd>


<dt>cs.colors</dt>
<dd><p>Colors for credible sets.</p></dd>


<dt>focal.colors</dt>
<dd><p>Colors for non-focal and focal gene.</p></dd>


<dt>label_QTLs</dt>
<dd><p>If TRUE, label SNP IDs in the QTL panel.</p></dd>


<dt>highlight_pval</dt>
<dd><p>p-value to highlight with a horizontal line</p></dd>


<dt>highlight_pip</dt>
<dd><p>PIP to highlight with a horizontal line</p></dd>


<dt>highlight_pos</dt>
<dd><p>genomic positions to highlight with vertical lines</p></dd>


<dt>highlight.color</dt>
<dd><p>color for horizontal lines highlighting p-values and PIPs.</p></dd>


<dt>highlight.pos.color</dt>
<dd><p>color for vertical lines highlighting genomic positions.</p></dd>


<dt>point.sizes</dt>
<dd><p>size values for SNP and non-SNP data points in the scatter plots</p></dd>


<dt>point.alpha</dt>
<dd><p>alpha values for SNP and non-SNP data points in the scatter plots</p></dd>


<dt>point.shapes</dt>
<dd><p>shapes values for data points of different types in the scatter plots</p></dd>


<dt>label.text.size</dt>
<dd><p>Font size for gene and SNP label text</p></dd>


<dt>max.overlaps</dt>
<dd><p>Setting for geom_text_repel() function to label texts.
Exclude text labels when they overlap too many other things.</p></dd>


<dt>axis.text.size</dt>
<dd><p>Font size for axis label text.</p></dd>


<dt>axis.title.size</dt>
<dd><p>Font size for axis title text.</p></dd>


<dt>legend.text.size</dt>
<dd><p>Font size for legend text.</p></dd>


<dt>legend.position</dt>
<dd><p>position to put legends. If "none", no legends will be shown.</p></dd>


<dt>genelabel.cex.text</dt>
<dd><p>Size for gene label text.</p></dd>


<dt>panel.heights</dt>
<dd><p>Relative heights of the panels.</p></dd>


<dt>verbose</dt>
<dd><p>If TRUE, print detail messages.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

