<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Preprocess PredictDB/FUSION weights and harmonize with LD reference — preprocess_weights • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Preprocess PredictDB/FUSION weights and harmonize with LD reference — preprocess_weights"><meta property="og:description" content="Preprocess PredictDB/FUSION weights and harmonize with LD reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Preprocess PredictDB/FUSION weights and harmonize with LD reference</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/R/ctwas_preprocess_weights.R" class="external-link"><code>R/ctwas_preprocess_weights.R</code></a></small>
    <div class="hidden name"><code>preprocess_weights.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Preprocess PredictDB/FUSION weights and harmonize with LD reference</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">preprocess_weights</span><span class="op">(</span></span>
<span>  <span class="va">weight_file</span>,</span>
<span>  <span class="va">region_info</span>,</span>
<span>  <span class="va">gwas_snp_ids</span>,</span>
<span>  <span class="va">snp_map</span>,</span>
<span>  LD_map <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  <span class="va">context</span>,</span>
<span>  weight_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">context</span>, <span class="st">"_"</span>, <span class="va">type</span><span class="op">)</span>,</span>
<span>  weight_format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PredictDB"</span>, <span class="st">"FUSION"</span><span class="op">)</span>,</span>
<span>  drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale_predictdb_weights <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  load_predictdb_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  include_weight_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fusion_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lasso"</span>, <span class="st">"enet"</span>, <span class="st">"top1"</span>, <span class="st">"blup"</span>, <span class="st">"bslmm"</span>, <span class="st">"best.cv"</span><span class="op">)</span>,</span>
<span>  fusion_genome_version <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  top_n_snps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  LD_format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rds"</span>, <span class="st">"rdata"</span>, <span class="st">"csv"</span>, <span class="st">"txt"</span>, <span class="st">"custom"</span><span class="op">)</span>,</span>
<span>  LD_loader_fun <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  snpinfo_loader_fun <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  varID_converter_fun <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>weight_file</dt>
<dd><p>filename of the '.db' file for PredictDB weights;
or the directory containing '.wgt.RDat' files for FUSION weights.</p></dd>


<dt>region_info</dt>
<dd><p>a data frame of region definitions.</p></dd>


<dt>gwas_snp_ids</dt>
<dd><p>a vector of SNP IDs in GWAS summary statistics (z_snp$id).</p></dd>


<dt>snp_map</dt>
<dd><p>a list of SNP-to-region map for the reference.</p></dd>


<dt>LD_map</dt>
<dd><p>a data frame with filenames of LD matrices and SNP information for the regions.
Required when <code>load_predictdb_LD = FALSE</code>.</p></dd>


<dt>type</dt>
<dd><p>a string, specifying QTL type of the weight file, e.g. expression, splicing, protein.</p></dd>


<dt>context</dt>
<dd><p>a string, specifying context (tissue/cell type) of the weight file, e.g. Liver, Lung, Brain.</p></dd>


<dt>weight_name</dt>
<dd><p>a string, specifying name of the weight file.
By default, it is <code>weight_name = paste0(context, "_", type)</code></p></dd>


<dt>weight_format</dt>
<dd><p>a string, specifying format of the weight file, e.g. PredictDB, FUSION.</p></dd>


<dt>drop_strand_ambig</dt>
<dd><p>If TRUE remove strand ambiguous variants (A/T, G/C).</p></dd>


<dt>filter_protein_coding_genes</dt>
<dd><p>If TRUE, keep protein coding genes only.
This option is only for PredictDB weights.</p></dd>


<dt>scale_predictdb_weights</dt>
<dd><p>If TRUE, scale PredictDB weights by the variance.
This is because PredictDB weights assume that variant genotypes are not
standardized, but our implementation assumes standardized variant genotypes.
This option is only for PredictDB weights.</p></dd>


<dt>load_predictdb_LD</dt>
<dd><p>If TRUE, load pre-computed LD among weight SNPs.
This option is only for PredictDB weights.</p></dd>


<dt>include_weight_LD</dt>
<dd><p>If TRUE, include LD of variants in weights (R_wgt) in the weights object.
R_wgt is used for computing gene Z-scores.
If FALSE, will skip computing R_wgt. This could save running time
when using precomputed gene Z-scores.</p></dd>


<dt>fusion_method</dt>
<dd><p>a string, specifying the method to choose in FUSION models.
"best.cv" option will use the best model (smallest p-value) under cross-validation.</p></dd>


<dt>fusion_genome_version</dt>
<dd><p>a string, specifying the genome version of FUSION models</p></dd>


<dt>top_n_snps</dt>
<dd><p>a number, specifying the top n SNPs included in weight models.
By default, use all SNPs in weights.</p></dd>


<dt>LD_format</dt>
<dd><p>file format for LD matrix. If "custom", use a user defined
<code>LD_loader_fun()</code> function to load LD matrix.</p></dd>


<dt>LD_loader_fun</dt>
<dd><p>a user defined function to load LD matrix when <code>LD_format = "custom"</code>.</p></dd>


<dt>snpinfo_loader_fun</dt>
<dd><p>a user defined function to load SNP information file,
if SNP information files are not in standard cTWAS reference format.</p></dd>


<dt>varID_converter_fun</dt>
<dd><p>a user defined function to convert
weight variant IDs to the reference variant format.</p></dd>


<dt>ncore</dt>
<dd><p>The number of cores used to parallelize computation.</p></dd>


<dt>logfile</dt>
<dd><p>The log filename. If NULL, will print log info on screen.</p></dd>


<dt>verbose</dt>
<dd><p>If TRUE, print detail messages.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list of processed weights</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

