<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Sum of Single Effects (SuSiE) Regression using summary statistics — susie_rss • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Sum of Single Effects (SuSiE) Regression using summary statistics — susie_rss"><meta property="og:description" content="susie_rss performs sum of single-effect linear
  regression with z scores; all posterior calculations are for
  z-scores. This function fits the regression model \(z = \sum_l
  R*b_l + e\), where e is \(N(0,R)\) and \(\sum_l b_l\) is a
  p-vector of effects to be estimated. The required summary data are
  the p by p correlation matrix, R, and the p-vector z."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Sum of Single Effects (SuSiE) Regression using summary statistics</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/R/susie_susie_rss.R" class="external-link"><code>R/susie_susie_rss.R</code></a></small>
    <div class="hidden name"><code>susie_rss.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>susie_rss</code> performs sum of single-effect linear
  regression with z scores; all posterior calculations are for
  z-scores. This function fits the regression model \(z = \sum_l
  R*b_l + e\), where e is \(N(0,R)\) and \(\sum_l b_l\) is a
  p-vector of effects to be estimated. The required summary data are
  the p by p correlation matrix, <code>R</code>, and the p-vector <code>z</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">susie_rss</span><span class="op">(</span></span>
<span>  <span class="va">z</span>,</span>
<span>  <span class="va">R</span>,</span>
<span>  maf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maf_thresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  z_ld_weight <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  L <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  prior_variance <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  residual_variance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior_weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  null_weight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  estimate_residual_variance <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  estimate_prior_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_prior_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"optim"</span>, <span class="st">"EM"</span>, <span class="st">"simple"</span><span class="op">)</span>,</span>
<span>  check_null_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  prior_tol <span class="op">=</span> <span class="fl">1e-09</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  s_init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept_value <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  min_abs_corr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  track_fit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  check_R <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  r_tol <span class="op">=</span> <span class="fl">1e-08</span>,</span>
<span>  refine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  warn_converge_fail <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>z</dt>
<dd><p>A p-vector of z scores.</p></dd>


<dt>R</dt>
<dd><p>A p by p symmetric, positive semidefinite correlation
matrix.</p></dd>


<dt>maf</dt>
<dd><p>Minor allele frequency; to be used along with
<code>maf_thresh</code> to filter input summary statistics.</p></dd>


<dt>maf_thresh</dt>
<dd><p>Variants having a minor allele frequency smaller
than this threshold are not used.</p></dd>


<dt>z_ld_weight</dt>
<dd><p>This feature is not recommended. The weights
assigned to the z scores in the LD matrix. If <code>z_ld_weight &gt;
0</code>, the LD matrix used in the model is <code>cov2cor((1-w)*R +
w*tcrossprod(z))</code>, where <code>w = z_ld_weight</code>.</p></dd>


<dt>L</dt>
<dd><p>Number of components (nonzero coefficients) in the susie
regression model. If L is larger than the number of covariates, p,
L is set to p.</p></dd>


<dt>prior_variance</dt>
<dd><p>The prior variance. It is either a scalar or
a vector of length L.</p></dd>


<dt>residual_variance</dt>
<dd><p>Variance of the residual.
If it is not specified, we set it to 1.</p></dd>


<dt>prior_weights</dt>
<dd><p>A vector of length p, in which each entry
gives the prior probability that SNP j has non-zero effect.</p></dd>


<dt>null_weight</dt>
<dd><p>Prior probability of no effect (a number between
0 and 1, and cannot be exactly 1).</p></dd>


<dt>estimate_residual_variance</dt>
<dd><p>The residual variance is
fixed to the value supplied by <code>residual_variance</code>. We don't
estimate residual variance in susie_rss.</p></dd>


<dt>estimate_prior_variance</dt>
<dd><p>If <code>estimate_prior_variance =
TRUE</code>, the prior variance is estimated (this is a separate
parameter for each of the L effects). If provided,
<code>prior_variance</code> is then used as an initial value for
the optimization. When <code>estimate_prior_variance = FALSE</code>, the
prior variance for each of the L effects is determined by the
value supplied to <code>prior_variance</code>.</p></dd>


<dt>estimate_prior_method</dt>
<dd><p>The method used for estimating prior
variance. When <code>estimate_prior_method = "simple"</code> is used, the
likelihood at the specified prior variance is compared to the
likelihood at a variance of zero, and the setting with the larger
likelihood is retained.</p></dd>


<dt>check_null_threshold</dt>
<dd><p>When the prior variance is estimated,
compare the estimate with the null, and set the prior variance to
zero unless the log-likelihood using the estimate is larger by this
threshold amount. For example, if you set
<code>check_null_threshold = 0.1</code>, this will "nudge" the estimate
towards zero when the difference in log-likelihoods is small. A
note of caution that setting this to a value greater than zero may
lead the IBSS fitting procedure to occasionally decrease the ELBO.</p></dd>


<dt>prior_tol</dt>
<dd><p>When the prior variance is estimated, compare the
estimated value to <code>prior_tol</code> at the end of the computation,
and exclude a single effect from PIP computation if the estimated
prior variance is smaller than this tolerance value.</p></dd>


<dt>max_iter</dt>
<dd><p>Maximum number of IBSS iterations to perform.</p></dd>


<dt>s_init</dt>
<dd><p>A previous susie fit with which to initialize.</p></dd>


<dt>intercept_value</dt>
<dd><p>The intercept. (The intercept cannot be
estimated from centered summary data.) This setting will be used by
<code>coef</code> to assign an intercept value, mainly for consistency
with <code>susie</code>. Set to <code>NULL</code> if you want <code>coef</code> not
to include an intercept term (and so only a p-vector is returned).</p></dd>


<dt>coverage</dt>
<dd><p>A number between 0 and 1 specifying the
“coverage” of the estimated confidence sets.</p></dd>


<dt>min_abs_corr</dt>
<dd><p>Minimum absolute correlation allowed in a
credible set. The default, 0.5, corresponds to a squared
correlation of 0.25, which is a commonly used threshold for
genotype data in genetic studies.</p></dd>


<dt>tol</dt>
<dd><p>A small, non-negative number specifying the convergence
tolerance for the IBSS fitting procedure. The fitting procedure
will halt when the difference in the variational lower bound, or
“ELBO” (the objective function to be maximized), is
less than <code>tol</code>.</p></dd>


<dt>verbose</dt>
<dd><p>If <code>verbose = TRUE</code>, the algorithm's progress,
and a summary of the optimization settings, are printed to the
console.</p></dd>


<dt>track_fit</dt>
<dd><p>If <code>track_fit = TRUE</code>, <code>trace</code>
is also returned containing detailed information about the
estimates at each iteration of the IBSS fitting procedure.</p></dd>


<dt>check_R</dt>
<dd><p>If <code>check_R = TRUE</code>, check that <code>R</code> is
positive semidefinite.</p></dd>


<dt>r_tol</dt>
<dd><p>Tolerance level for eigenvalue check of positive
semidefinite matrix of R.</p></dd>


<dt>refine</dt>
<dd><p>If <code>refine = TRUE</code>, we use a procedure to help
SuSiE get out of local optimum.</p></dd>


<dt>warn_converge_fail</dt>
<dd><p>If <code>warn_converge_fail = TRUE</code>,
prints a warning message when IBSS algorithm does not converge.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>"susie"</code> object with some or all of the following
  elements:</p>
<dl><dt>alpha</dt>
<dd><p>An L by p matrix of posterior inclusion probabilites.</p></dd>


<dt>mu</dt>
<dd><p>An L by p matrix of posterior means, conditional on
  inclusion.</p></dd>


<dt>mu2</dt>
<dd><p>An L by p matrix of posterior second moments,
  conditional on inclusion.</p></dd>


<dt>lbf</dt>
<dd><p>log-Bayes Factor for each single effect.</p></dd>


<dt>lbf_variable</dt>
<dd><p>log-Bayes Factor for each variable and single effect.</p></dd>


<dt>intercept</dt>
<dd><p>Fixed Intercept.</p></dd>


<dt>sigma2</dt>
<dd><p>Fixed Residual variance.</p></dd>


<dt>V</dt>
<dd><p>Prior variance of the non-zero elements of b, equal to
  <code>scaled_prior_variance * var(Y)</code>.</p></dd>


<dt>elbo</dt>
<dd><p>The value of the variational lower bound, or
  “ELBO” (objective function to be maximized), achieved at
  each iteration of the IBSS fitting procedure.</p></dd>


<dt>fitted</dt>
<dd><p>Vector of length n containing the fitted values of
  the outcome.</p></dd>


<dt>sets</dt>
<dd><p>Credible sets estimated from model fit; see
  <code><a href="susie_get_methods.html">susie_get_cs</a></code> for details.</p></dd>


<dt>pip</dt>
<dd><p>A vector of length p giving the (marginal) posterior
  inclusion probabilities for all p covariates.</p></dd>


<dt>niter</dt>
<dd><p>Number of IBSS iterations that were performed.</p></dd>


<dt>converged</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> indicating whether
  the IBSS converged to a solution within the chosen tolerance
  level.</p></dd>


<dt>Rr</dt>
<dd><p>An p-vector of <code>t(X)</code> times fitted values, <code>X
  %*% colSums(alpha*mu)</code>.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

