<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Iteratively run susie and estimate parameters — susieI • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Iteratively run susie and estimate parameters — susieI"><meta property="og:description" content="Iteratively run susie and estimate parameters"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.38</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Iteratively run susie and estimate parameters</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/R/ctwas_susieI.R" class="external-link"><code>R/ctwas_susieI.R</code></a></small>
    <div class="hidden name"><code>susieI.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Iteratively run susie and estimate parameters</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">susieI</span><span class="op">(</span></span>
<span>  <span class="va">pgenfs</span>,</span>
<span>  <span class="va">exprfs</span>,</span>
<span>  <span class="va">Y</span>,</span>
<span>  <span class="va">regionlist</span>,</span>
<span>  niter <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  L <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  group_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  estimate_group_prior <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  estimate_group_prior_var <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  use_null_weight <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  outputdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  outname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  report_parameters <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-pgenfs">pgenfs<a class="anchor" aria-label="anchor" href="#arg-pgenfs"></a></dt>
<dd><p>A character vector of .pgen or .bed files. One file for one
chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22. If .pgen files are given, then .pvar and .psam are assumed
to present in the same directory. If .bed files are given, then .bim and
.fam files are assumed to present in the same directory.</p></dd>


<dt id="arg-y">Y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>a vector of length n, phenotype, the same order as provided
by `.pgenfs` (defined in .psam or .fam files).</p></dd>


<dt id="arg-regionlist">regionlist<a class="anchor" aria-label="anchor" href="#arg-regionlist"></a></dt>
<dd><p>a list object indexing regions, variants and genes. The output of
<code>index_regions</code></p></dd>


<dt id="arg-niter">niter<a class="anchor" aria-label="anchor" href="#arg-niter"></a></dt>
<dd><p>the number of iterations of the E-M algorithm to perform</p></dd>


<dt id="arg-l">L<a class="anchor" aria-label="anchor" href="#arg-l"></a></dt>
<dd><p>the number of effects for susie</p></dd>


<dt id="arg-group-prior">group_prior<a class="anchor" aria-label="anchor" href="#arg-group-prior"></a></dt>
<dd><p>a vector of two prior inclusion probabilities for SNPs and genes. This is ignored
if <code>estimate_group_prior = T</code></p></dd>


<dt id="arg-group-prior-var">group_prior_var<a class="anchor" aria-label="anchor" href="#arg-group-prior-var"></a></dt>
<dd><p>a vector of two prior variances for SNPs and gene effects. This is ignored
if <code>estimate_group_prior_var = T</code></p></dd>


<dt id="arg-estimate-group-prior">estimate_group_prior<a class="anchor" aria-label="anchor" href="#arg-estimate-group-prior"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, the prior inclusion probabilities for SNPs and genes are estimated
using the data. If FALSE, <code>group_prior</code> must be specified</p></dd>


<dt id="arg-estimate-group-prior-var">estimate_group_prior_var<a class="anchor" aria-label="anchor" href="#arg-estimate-group-prior-var"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, the prior variances for SNPs and genes are estimated
using the data. If FALSE, <code>group_prior_var</code> must be specified</p></dd>


<dt id="arg-use-null-weight">use_null_weight<a class="anchor" aria-label="anchor" href="#arg-use-null-weight"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, allow for a probability of no effect in susie</p></dd>


<dt id="arg-coverage">coverage<a class="anchor" aria-label="anchor" href="#arg-coverage"></a></dt>
<dd><p>A number between 0 and 1 specifying the “coverage” of the estimated confidence sets</p></dd>


<dt id="arg-standardize">standardize<a class="anchor" aria-label="anchor" href="#arg-standardize"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, all variables are standardized to unit variance</p></dd>


<dt id="arg-ncore">ncore<a class="anchor" aria-label="anchor" href="#arg-ncore"></a></dt>
<dd><p>The number of cores used to parallelize susie over regions</p></dd>


<dt id="arg-outputdir">outputdir<a class="anchor" aria-label="anchor" href="#arg-outputdir"></a></dt>
<dd><p>a string, the directory to store output</p></dd>


<dt id="arg-outname">outname<a class="anchor" aria-label="anchor" href="#arg-outname"></a></dt>
<dd><p>a string, the output name</p></dd>


<dt id="arg-report-parameters">report_parameters<a class="anchor" aria-label="anchor" href="#arg-report-parameters"></a></dt>
<dd><p>TRUE/FALSE. If TRUE, estimated parameters are reported at the end of iteration</p></dd>


<dt id="arg-a">A<a class="anchor" aria-label="anchor" href="#arg-a"></a></dt>
<dd><p>character vector of .`expr` or `.expr.gz` files. One file for
one chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22.  `.expr.gz` file is gzip compressed `.expr` files. `.expr` is
a matrix of imputed expression values, row is for each sample, column is for
each gene. Its sample order is same as in files provided by `.pgenfs`. We also
assume corresponding `.exprvar` files are present in the same directory.
`.exprvar` files are just tab delimited text files, with columns:</p><dl><dt>chrom</dt>
<dd><p>chromosome number, numeric</p></dd>

  <dt>p0</dt>
<dd><p>gene boundary position, the smaller value</p></dd>

  <dt>p1</dt>
<dd><p>gene boundary position, the larger value</p></dd>

  <dt>id</dt>
<dd><p>gene id</p></dd>


</dl><p>Its rows should be in the same order as the columns for corresponding `.expr`
files.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

