<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Sum of Single Effects (SuSiE) Regression — susie • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Sum of Single Effects (SuSiE) Regression — susie"><meta property="og:description" content="Performs Bayesian multiple linear regression of Y on
  X; that is, this function fits the regression model \(Y = \sum_l
  X b_{l=1}^L + e\), where elements of e are i.i.d. normal with
  zero mean and variance residual_variance, and
  \(\sum_{l=1}^L b_l\) is a vector of length p representing the
  effects to be estimated. The &amp;#8220;susie assumption&amp;#8221; is that each
  \(b_l\) has exactly one non-zero element. The prior on the
  non-zero element is normal with zero mean and variance var(Y)
  * scaled_prior_variance. The model is fitted using the
  &amp;#8220;Iterative Bayesian Stepwise Selection&amp;#8221; (IBSS) algorithm."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/minimal_tutorial.html">A minimal tutorial of running cTWAS without LD</a>
    </li>
    <li>
      <a href="../articles/preparing_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/running_ctwas_analysis.html">Running cTWAS analysis</a>
    </li>
    <li>
      <a href="../articles/summarizing_results.html">Summarizing and visualizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/sample_report.html">Sample report of cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/utility_functions.html">cTWAS utility functions</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/software_updates.html">Software updates</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Sum of Single Effects (SuSiE) Regression</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/R/susie_susie.R" class="external-link"><code>R/susie_susie.R</code></a>, <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup/R/susie_susie_ss.R" class="external-link"><code>R/susie_susie_ss.R</code></a></small>
    <div class="hidden name"><code>susie.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs Bayesian multiple linear regression of Y on
  X; that is, this function fits the regression model \(Y = \sum_l
  X b_{l=1}^L + e\), where elements of e are <em>i.i.d.</em> normal with
  zero mean and variance <code>residual_variance</code>, and
  \(\sum_{l=1}^L b_l\) is a vector of length p representing the
  effects to be estimated. The “susie assumption” is that each
  \(b_l\) has exactly one non-zero element. The prior on the
  non-zero element is normal with zero mean and variance <code>var(Y)
  * scaled_prior_variance</code>. The model is fitted using the
  “Iterative Bayesian Stepwise Selection” (IBSS) algorithm.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">susie</span><span class="op">(</span></span>
<span>  <span class="va">X</span>,</span>
<span>  <span class="va">Y</span>,</span>
<span>  L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  scaled_prior_variance <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  residual_variance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior_weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  null_weight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  intercept <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_residual_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_prior_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_prior_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"optim"</span>, <span class="st">"EM"</span>, <span class="st">"simple"</span><span class="op">)</span>,</span>
<span>  check_null_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  prior_tol <span class="op">=</span> <span class="fl">1e-09</span>,</span>
<span>  residual_variance_upperbound <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  s_init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  min_abs_corr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  compute_univariate_zscore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  na.rm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  track_fit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  residual_variance_lowerbound <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/drop.html" class="external-link">drop</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">10000</span>,</span>
<span>  refine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  warn_converge_fail <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">susie_suff_stat</span><span class="op">(</span></span>
<span>  <span class="va">bhat</span>,</span>
<span>  <span class="va">shat</span>,</span>
<span>  <span class="va">R</span>,</span>
<span>  <span class="va">n</span>,</span>
<span>  <span class="va">var_y</span>,</span>
<span>  <span class="va">XtX</span>,</span>
<span>  <span class="va">Xty</span>,</span>
<span>  <span class="va">yty</span>,</span>
<span>  maf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maf_thresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  L <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  scaled_prior_variance <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  residual_variance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  estimate_residual_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_prior_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_prior_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"optim"</span>, <span class="st">"EM"</span>, <span class="st">"simple"</span><span class="op">)</span>,</span>
<span>  check_null_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  prior_tol <span class="op">=</span> <span class="fl">1e-09</span>,</span>
<span>  r_tol <span class="op">=</span> <span class="fl">1e-08</span>,</span>
<span>  prior_weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  null_weight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  s_init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept_value <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  min_abs_corr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  track_fit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  check_input <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  refine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  warn_converge_fail <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>X</dt>
<dd><p>An n by p matrix of covariates.</p></dd>


<dt>Y</dt>
<dd><p>The observed responses, a vector of length n.</p></dd>


<dt>L</dt>
<dd><p>Number of components (nonzero coefficients) in the susie
regression model. If L is larger than the number of covariates, p,
L is set to p.</p></dd>


<dt>scaled_prior_variance</dt>
<dd><p>The scaled prior variance. This is
either a scalar or a vector of length <code>L</code>. The prior variance
of each non-zero element of b is set to <code>var(Y) *
scaled_prior_variance</code>. If <code>estimate_prior_variance = TRUE</code>,
this provides initial estimates of the prior variances.</p></dd>


<dt>residual_variance</dt>
<dd><p>Variance of the residual. If
<code>estimate_residual_variance = TRUE</code>, this value provides the
initial estimate of the residual variance. By default, it is
<code>var(Y)</code>.</p></dd>


<dt>prior_weights</dt>
<dd><p>A vector of length p, in which each entry
gives the prior probability that corresponding column of X has a
nonzero effect on the outcome, Y.</p></dd>


<dt>null_weight</dt>
<dd><p>Prior probability of no effect (a number between
0 and 1, and cannot be exactly 1).</p></dd>


<dt>standardize</dt>
<dd><p>If <code>standardize = TRUE</code>, standardize the
columns of X (or XtX and Xty) to unit variance prior to
fitting. Note that <code>scaled_prior_variance</code> specifies the prior
on the coefficients of X <em>after</em> standardization (if it is
performed). If you do not standardize, you may need to think more
carefully about specifying <code>scaled_prior_variance</code>. Whatever
your choice, the coefficients returned by <code>coef</code> are given for
<code>X</code> on the original input scale. Any column of <code>X</code> that
has zero variance is not standardized.</p></dd>


<dt>intercept</dt>
<dd><p>If <code>intercept = TRUE</code>, the intercept is
fitted; it <code>intercept = FALSE</code>, the intercept is set to
zero. Setting <code>intercept = FALSE</code> is generally not
recommended.</p></dd>


<dt>estimate_residual_variance</dt>
<dd><p>If
<code>estimate_residual_variance = TRUE</code>, the residual variance is
estimated, using <code>residual_variance</code> as an initial value. If
<code>estimate_residual_variance = FALSE</code>, the residual variance is
fixed to the value supplied by <code>residual_variance</code>.</p></dd>


<dt>estimate_prior_variance</dt>
<dd><p>If <code>estimate_prior_variance =
TRUE</code>, the prior variance is estimated (this is a separate
parameter for each of the L effects). If provided,
<code>scaled_prior_variance</code> is then used as an initial value for
the optimization. When <code>estimate_prior_variance = FALSE</code>, the
prior variance for each of the L effects is determined by the
value supplied to <code>scaled_prior_variance</code>.</p></dd>


<dt>estimate_prior_method</dt>
<dd><p>The method used for estimating prior
variance. When <code>estimate_prior_method = "simple"</code> is used, the
likelihood at the specified prior variance is compared to the
likelihood at a variance of zero, and the setting with the larger
likelihood is retained.</p></dd>


<dt>check_null_threshold</dt>
<dd><p>When the prior variance is estimated,
compare the estimate with the null, and set the prior variance to
zero unless the log-likelihood using the estimate is larger by this
threshold amount. For example, if you set
<code>check_null_threshold = 0.1</code>, this will "nudge" the estimate
towards zero when the difference in log-likelihoods is small. A
note of caution that setting this to a value greater than zero may
lead the IBSS fitting procedure to occasionally decrease the ELBO.</p></dd>


<dt>prior_tol</dt>
<dd><p>When the prior variance is estimated, compare the
estimated value to <code>prior_tol</code> at the end of the computation,
and exclude a single effect from PIP computation if the estimated
prior variance is smaller than this tolerance value.</p></dd>


<dt>residual_variance_upperbound</dt>
<dd><p>Upper limit on the estimated
residual variance. It is only relevant when
<code>estimate_residual_variance = TRUE</code>.</p></dd>


<dt>s_init</dt>
<dd><p>A previous susie fit with which to initialize.</p></dd>


<dt>coverage</dt>
<dd><p>A number between 0 and 1 specifying the
“coverage” of the estimated confidence sets.</p></dd>


<dt>min_abs_corr</dt>
<dd><p>Minimum absolute correlation allowed in a
credible set. The default, 0.5, corresponds to a squared
correlation of 0.25, which is a commonly used threshold for
genotype data in genetic studies.</p></dd>


<dt>compute_univariate_zscore</dt>
<dd><p>If <code>compute_univariate_zscore
= TRUE</code>, the univariate regression z-scores are outputted for each
variable.</p></dd>


<dt>na.rm</dt>
<dd><p>Drop any missing values in Y from both X and Y.</p></dd>


<dt>max_iter</dt>
<dd><p>Maximum number of IBSS iterations to perform.</p></dd>


<dt>tol</dt>
<dd><p>A small, non-negative number specifying the convergence
tolerance for the IBSS fitting procedure. The fitting procedure
will halt when the difference in the variational lower bound, or
“ELBO” (the objective function to be maximized), is
less than <code>tol</code>.</p></dd>


<dt>verbose</dt>
<dd><p>If <code>verbose = TRUE</code>, the algorithm's progress,
and a summary of the optimization settings, are printed to the
console.</p></dd>


<dt>track_fit</dt>
<dd><p>If <code>track_fit = TRUE</code>, <code>trace</code>
is also returned containing detailed information about the
estimates at each iteration of the IBSS fitting procedure.</p></dd>


<dt>residual_variance_lowerbound</dt>
<dd><p>Lower limit on the estimated
residual variance. It is only relevant when
<code>estimate_residual_variance = TRUE</code>.</p></dd>


<dt>refine</dt>
<dd><p>If <code>refine = TRUE</code>, we use a procedure to help
SuSiE get out of local optimum.</p></dd>


<dt>warn_converge_fail</dt>
<dd><p>If <code>warn_converge_fail = TRUE</code>,
prints a warning message when IBSS algorithm does not converge.</p></dd>


<dt>bhat</dt>
<dd><p>A p-vector of estimated effects.</p></dd>


<dt>shat</dt>
<dd><p>A p-vector of standard errors.</p></dd>


<dt>R</dt>
<dd><p>A p by p symmetric, positive semidefinite matrix. It
can be \(X'X\), the covariance matrix \(X'X/(n-1)\), or a
correlation matrix. It should be estimated from the same samples
used to compute <code>bhat</code> and <code>shat</code>. Using an out-of-sample
matrix may produce unreliable results.</p></dd>


<dt>n</dt>
<dd><p>The sample size.</p></dd>


<dt>var_y</dt>
<dd><p>The sample variance of y, defined as \(y'y/(n-1)\).
When the sample variance cannot be provided, the coefficients
(returned from <code>coef</code>) are computed on the "standardized" X, y
scale.</p></dd>


<dt>XtX</dt>
<dd><p>A p by p matrix \(X'X\) in which the columns of X
are centered to have mean zero.</p></dd>


<dt>Xty</dt>
<dd><p>A p-vector \(X'y\) in which y and the columns of X are
centered to have mean zero.</p></dd>


<dt>yty</dt>
<dd><p>A scalar \(y'y\) in which y is centered to have mean
zero.</p></dd>


<dt>maf</dt>
<dd><p>Minor allele frequency; to be used along with
<code>maf_thresh</code> to filter input summary statistics.</p></dd>


<dt>maf_thresh</dt>
<dd><p>Variants having a minor allele frequency smaller
than this threshold are not used.</p></dd>


<dt>r_tol</dt>
<dd><p>Tolerance level for eigenvalue check of positive
semidefinite matrix of R.</p></dd>


<dt>intercept_value</dt>
<dd><p>The intercept. (The intercept cannot be
estimated from centered summary data.) This setting will be used by
<code>coef</code> to assign an intercept value, mainly for consistency
with <code>susie</code>. Set to <code>NULL</code> if you want <code>coef</code> not
to include an intercept term (and so only a p-vector is returned).</p></dd>


<dt>check_input</dt>
<dd><p>If <code>check_input = TRUE</code>,
<code>susie_suff_stat</code> performs additional checks on <code>XtX</code> and
<code>Xty</code>. The checks are: (1) check that <code>XtX</code> is positive
semidefinite; (2) check that <code>Xty</code> is in the space spanned by
the non-zero eigenvectors of <code>XtX</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>"susie"</code> object with some or all of the following
  elements:</p>
<dl><dt>alpha</dt>
<dd><p>An L by p matrix of posterior inclusion probabilites.</p></dd>


<dt>mu</dt>
<dd><p>An L by p matrix of posterior means, conditional on
  inclusion.</p></dd>


<dt>mu2</dt>
<dd><p>An L by p matrix of posterior second moments,
  conditional on inclusion.</p></dd>


<dt>Xr</dt>
<dd><p>A vector of length n, equal to <code>X %*% colSums(alpha
  * mu)</code>.</p></dd>


<dt>lbf</dt>
<dd><p>log-Bayes Factor for each single effect.</p></dd>


<dt>lbf_variable</dt>
<dd><p>log-Bayes Factor for each variable and single effect.</p></dd>


<dt>intercept</dt>
<dd><p>Intercept (fixed or estimated).</p></dd>


<dt>sigma2</dt>
<dd><p>Residual variance (fixed or estimated).</p></dd>


<dt>V</dt>
<dd><p>Prior variance of the non-zero elements of b, equal to
  <code>scaled_prior_variance * var(Y)</code>.</p></dd>


<dt>elbo</dt>
<dd><p>The value of the variational lower bound, or
  “ELBO” (objective function to be maximized), achieved at
  each iteration of the IBSS fitting procedure.</p></dd>


<dt>fitted</dt>
<dd><p>Vector of length n containing the fitted values of
  the outcome.</p></dd>


<dt>sets</dt>
<dd><p>Credible sets estimated from model fit; see
  <code><a href="susie_get_methods.html">susie_get_cs</a></code> for details.</p></dd>


<dt>pip</dt>
<dd><p>A vector of length p giving the (marginal) posterior
  inclusion probabilities for all p covariates.</p></dd>


<dt>z</dt>
<dd><p>A vector of univariate z-scores.</p></dd>


<dt>niter</dt>
<dd><p>Number of IBSS iterations that were performed.</p></dd>


<dt>converged</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> indicating whether
  the IBSS converged to a solution within the chosen tolerance
  level.</p></dd>


</dl><p><code>susie_suff_stat</code> returns also outputs:</p>
<dl><dt>XtXr</dt>
<dd><p>A p-vector of <code>t(X)</code> times the fitted values,
  <code>X %*% colSums(alpha*mu)</code>.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p><code>susie_suff_stat</code> performs sum of single-effect
linear regression with summary statistics. The required summary
data are either: <code>bhat</code>, <code>shat</code>, the p by p symmetric,
positive semidefinite correlation (or covariance) matrix <code>R</code>,
the sample size <code>n</code>, and the variance of y; or the p by p
matrix \(X'X\), the p-vector \(X'y\), the sum of squares
\(y'y\), and the sample size <code>n</code>. The summary statistics
should come from the same individuals. Both the columns of X and
the vector y should be centered to have mean zero before computing
these summary statistics; you may also want to scale each column of
X and y to have variance 1 (see examples).</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>G. Wang, A. Sarkar, P. Carbonetto and M. Stephens (2020). A simple
  new approach to variable selection in regression, with application
  to genetic fine-mapping. <em>Journal of the Royal Statistical
  Society, Series B</em> <a href="https://doi.org/10.1101/501114" class="external-link">doi:10.1101/501114</a>
.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="susie_rss.html">susie_rss</a></code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

